--- hibernate-core-5.3.6.Final-sources/org/hibernate/cfg/OneToOneSecondPass.java	2018-08-13 12:16:20.000000000 -0700
+++ hibernate-core-5.3.6.Final-sources-patched/org/hibernate/cfg/OneToOneSecondPass.java.new	2018-09-13 14:12:26.359531715 -0700
@@ -109,6 +109,7 @@
 		binder.setValue( value );
 		binder.setCascade( cascadeStrategy );
 		binder.setAccessType( inferredData.getDefaultAccess() );
+		binder.setProperty( inferredData.getProperty() );
 
 		final LazyGroup lazyGroupAnnotation = inferredData.getProperty().getAnnotation( LazyGroup.class );
 		if ( lazyGroupAnnotation != null ) {
