*** Configuration.java2	Wed Dec 17 15:07:12 2003
--- Configuration.java	Mon Jan 19 12:12:45 2004
***************
*** 286,295 ****
--- 286,301 ----
  			throw me;
  		}
  		catch (Exception e) {
  			log.error("Could not configure datastore from input stream", e);
  			throw new MappingException(e);
+ 		} finally{
+ 			try{
+ 			     xmlInputStream.close();
+ 			} catch(IOException e){
+ 				log.error("Could not close InputStream", e);				
+ 			}
  		}
  	}
  
  	/**
  	 * Read mappings from an application resource
***************
*** 902,911 ****
--- 908,924 ----
  			if ( errors.size()!=0 ) throw new MappingException( "invalid configuration", (Throwable) errors.get(0) );
  		}
  		catch (Exception e) {
  			log.error("problem parsing configuration" + resourceName, e);
  			throw new HibernateException("problem parsing configuration" + resourceName, e);
+ 		} finally{
+ 			try{
+ 			     stream.close();
+ 			} catch(IOException ie){
+ 				log.error("could not close input stream on: " + resourceName, ie);
+ 			}
+ 				     
  		}
  
  		return doConfigure(doc);
  
  	}
