*** Environment.java2	Wed Dec 17 15:07:12 2003
--- Environment.java	Mon Jan 19 11:52:45 2004
***************
*** 1,9 ****
--- 1,10 ----
  //$Id: Environment.java,v 1.18.2.19 2003/12/13 16:15:01 oneovthafew Exp $
  package net.sf.hibernate.cfg;
  
  import java.io.InputStream;
+ import java.io.IOException;
  import java.sql.Connection;
  import java.sql.Timestamp;
  import java.util.HashMap;
  import java.util.Iterator;
  import java.util.Map;
***************
*** 465,474 ****
--- 466,481 ----
  				GLOBAL_PROPERTIES.load(stream);
  				log.info("loaded properties from resource hibernate.properties: " + GLOBAL_PROPERTIES);
  			}
  			catch (Exception e) {
  				log.error("problem loading properties from hibernate.properties");
+ 			} finally{
+ 				try{
+ 					stream.close();
+ 				} catch(IOException e){
+ 					log.error("could not close stream on hibernate.properties");	
+ 				}
  			}
  		}
  		
  		GLOBAL_PROPERTIES.putAll( System.getProperties() );
  		
