Index: IncrementGenerator.java
===================================================================
--- IncrementGenerator.java	(revision 21080)
+++ IncrementGenerator.java	(working copy)
@@ -102,7 +102,7 @@
 		for ( int i=0; i < tables.length; i++ ) {
 			final String tableName = dialect.quote( normalizer.normalizeIdentifierQuoting( tables[i] ) );
 			if ( tables.length > 1 ) {
-				buf.append( "select " ).append( column ).append( " from " );
+				buf.append( "select max(" ).append( column ).append( ") as mx from " );
 			}
 			buf.append( Table.qualify( catalog, schema, tableName ) );
 			if ( i < tables.length-1 ) {
@@ -111,10 +111,9 @@
 		}
 		if ( tables.length > 1 ) {
 			buf.insert( 0, "( " ).append( " ) ids_" );
-			column = "ids_." + column;
 		}
 		
-		sql = "select max(" + column + ") from " + buf.toString();
+		sql = "select max(ids_.mx) from " + buf.toString();
 	}
 
 	private void initializePreviousValueHolder(SessionImplementor session) {
