Index: src/org/hibernate/hql/ast/exec/AbstractStatementExecutor.java
===================================================================
--- src/org/hibernate/hql/ast/exec/AbstractStatementExecutor.java	(revision 11270)
+++ src/org/hibernate/hql/ast/exec/AbstractStatementExecutor.java	(working copy)
@@ -92,7 +92,7 @@
 		select.setWhereClause( whereJoinFragment + userWhereClause );
 
 		InsertSelect insert = new InsertSelect( getFactory().getDialect() );
-		if ( getFactory().getSettings().isCommentsEnabled() && getFactory().getDialect().supportsCommentOn() ) {
+		if ( getFactory().getSettings().isCommentsEnabled() ) {
 			insert.setComment( "insert-select for " + persister.getEntityName() + " ids" );
 		}
 		insert.setTableName( persister.getTemporaryIdTableName() );
Index: src/org/hibernate/hql/ast/exec/MultiTableDeleteExecutor.java
===================================================================
--- src/org/hibernate/hql/ast/exec/MultiTableDeleteExecutor.java	(revision 11270)
+++ src/org/hibernate/hql/ast/exec/MultiTableDeleteExecutor.java	(working copy)
@@ -60,7 +60,7 @@
 			final Delete delete = new Delete()
 					.setTableName( tableNames[i] )
 					.setWhere( "(" + StringHelper.join( ", ", columnNames[i] ) + ") IN (" + idSubselect + ")" );
-			if ( getFactory().getSettings().isCommentsEnabled() && getFactory().getDialect().supportsCommentOn() ) {
+			if ( getFactory().getSettings().isCommentsEnabled() ) {
 				delete.setComment( "bulk delete" );
 			}
 
Index: src/org/hibernate/hql/ast/exec/MultiTableUpdateExecutor.java
===================================================================
--- src/org/hibernate/hql/ast/exec/MultiTableUpdateExecutor.java	(revision 11270)
+++ src/org/hibernate/hql/ast/exec/MultiTableUpdateExecutor.java	(working copy)
@@ -65,7 +65,7 @@
 			Update update = new Update( getFactory().getDialect() )
 					.setTableName( tableNames[tableIndex] )
 					.setWhere( "(" + StringHelper.join( ", ", columnNames[tableIndex] ) + ") IN (" + idSubselect + ")" );
-			if ( getFactory().getSettings().isCommentsEnabled() && getFactory().getDialect().supportsCommentOn() ) {
+			if ( getFactory().getSettings().isCommentsEnabled() ) {
 				update.setComment( "bulk update" );
 			}
 			final Iterator itr = assignmentSpecifications.iterator();
