*** daohome.vm.orig     2005-09-25 22:27:08.000000000 +0200
--- daohome.vm  2005-11-17 09:07:55.000000000 +0100
***************
*** 217,223 ****
--- 217,227 ----
              ${pojo.importType("java.util.List")} results = sessionFactory.getCurrentSession()
  #end
                      .createCriteria("$clazz.entityName")
+ #if($generics)
                      .add( $pojo.staticImport("org.hibernate.criterion.Example", "create")(instance) )
+ #else
+                     .add(${pojo.importType("org.hibernate.criterion.Example")}.create(instance))
+ #endif
                      .list();
              log.debug("find by example successful, result size: " + results.size());
              return results;
***************
*** 263,266 ****
  #end
  #end
 
! }
\ Kein Zeilenumbruch am Dateiende.
--- 267,270 ----
  #end
  #end
 
! }
