Index: EntityPOJOClass.java
===================================================================
--- EntityPOJOClass.java	(revision 17090)
+++ EntityPOJOClass.java	(working copy)
@@ -628,8 +628,24 @@
 					annotation.append( " }");
 					annotation.append(")");
 				}
-
+				
 			}
+			if (collection instanceof org.hibernate.mapping.List &&
+					((org.hibernate.mapping.List) collection).getIndex() != null) {
+				
+				org.hibernate.mapping.List list = (org.hibernate.mapping.List) collection;
+				StringBuffer indexColumnAnnotation = new StringBuffer();
+				indexColumnAnnotation.append("@");
+				indexColumnAnnotation.append(importType("org.hibernate.annotations.IndexColumn"));
+				indexColumnAnnotation.append("(name=\"");
+				Value indexValue = list.getIndex();
+				Column column = (Column) indexValue.getColumnIterator().next();
+				indexColumnAnnotation.append(column.getName());
+				indexColumnAnnotation.append("\")");
+				
+				annotation.append("\n    ").append(indexColumnAnnotation.toString());
+			}
+			
 			String hibernateCascade = getHibernateCascadeTypeAnnotation( property );
 			if (hibernateCascade.length() > 0) annotation.append("\n    ").append(hibernateCascade);
 		}
