Index: Hbm2JavaEjb3Test.java
===================================================================
--- Hbm2JavaEjb3Test.java	(revision 17090)
+++ Hbm2JavaEjb3Test.java	(working copy)
@@ -137,6 +137,20 @@
 		
 	}
 	
+	
+	public void testIndexColumnAnnotation() {
+		PersistentClass classMapping = getCfg().getClassMapping("org.hibernate.tool.hbm2x.Article");
+		Cfg2JavaTool cfg2java = new Cfg2JavaTool();
+		EntityPOJOClass clazz = (EntityPOJOClass) cfg2java.getPOJOClass(classMapping);
+		
+		Property p = classMapping.getProperty("aList");
+		
+		String string = clazz.generateCollectionAnnotation(p, getCfg() );
+		assertTrue(string.indexOf("@IndexColumn")>0);
+		
+	}
+	
+	
 	public void testEmptyCascade() {
 		PersistentClass classMapping = getCfg().getClassMapping("org.hibernate.tool.hbm2x.Article");
 		
