Index: src/net/sf/hibernate/collection/Map.java
===================================================================
--- src/net/sf/hibernate/collection/Map.java	(revision 4777)
+++ src/net/sf/hibernate/collection/Map.java	(revision 4778)
@@ -336,7 +336,7 @@
 	public boolean needsInserting(Object entry, int i, Type elemType) throws HibernateException {
 		final java.util.Map sn = (java.util.Map) getSnapshot();
 		java.util.Map.Entry e = (java.util.Map.Entry) entry;
-		return e.getValue()!=null && sn.get( e.getKey() )==null;
+		return e.getValue()!=null && !sn.containsKey( e.getKey() );
 	}
 	
 	public boolean needsUpdating(Object entry, int i, Type elemType) throws HibernateException {
@@ -344,7 +344,7 @@
 		java.util.Map.Entry e = (java.util.Map.Entry) entry;
 		Object snValue = sn.get( e.getKey() );
 		return e.getValue()!=null && 
-			snValue!=null && 
+			sn.containsKey(e.getKey()) &&
 			elemType.isDirty( snValue, e.getValue(), getSession() );
 	}
 	
