Index: src/net/sf/hibernate/mapping/Table.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/mapping/Table.java,v
retrieving revision 1.12
diff -u -r1.12 Table.java
--- src/net/sf/hibernate/mapping/Table.java	16 Jun 2003 03:30:32 -0000	1.12
+++ src/net/sf/hibernate/mapping/Table.java	20 Jun 2003 07:37:36 -0000
@@ -115,6 +115,7 @@
 			if (columnInfo==null) {
 				// the column doesnt exist at all.
 				StringBuffer alter = new StringBuffer( root.toString() )
+                                       .append(' ')
 					.append( col.getQuotedName(dialect) )
 					.append(' ')
 					.append( col.getSqlType(dialect, p) );
