Index: envers/src/main/java/org/hibernate/envers/configuration/metadata/MetadataTools.java
===================================================================
--- envers/src/main/java/org/hibernate/envers/configuration/metadata/MetadataTools.java	(revision 20267)
+++ envers/src/main/java/org/hibernate/envers/configuration/metadata/MetadataTools.java	(working copy)
@@ -32,6 +32,8 @@
 import org.hibernate.envers.tools.StringTools;
 
 import org.hibernate.mapping.Column;
+import org.hibernate.mapping.Formula;
+import org.hibernate.mapping.Selectable;
 
 /**
  * @author Adam Warski (adam at warski dot org)
@@ -174,9 +176,13 @@
 
     public static void addColumns(Element any_mapping, Iterator<Column> columns) {
         while (columns.hasNext()) {
-            Column column = columns.next();
-            addColumn(any_mapping, column.getName(), column.getLength(), column.getScale(), column.getPrecision(),
-					column.getSqlType());
+            Selectable selectable = (Selectable)columns.next();
+            if( !selectable.isFormula() ) {
+                Column column = (Column)selectable;
+                addColumn(any_mapping, column.getName(), column.getLength(), column.getScale(), 
+                          column.getPrecision(), column.getSqlType());
+            }
+
         }
     }
 
