Index: src/test/java/org/hibernate/test/idgen/enhanced/SequenceStyleConfigUnitTest.java
===================================================================
--- src/test/java/org/hibernate/test/idgen/enhanced/SequenceStyleConfigUnitTest.java	(revision 18789)
+++ src/test/java/org/hibernate/test/idgen/enhanced/SequenceStyleConfigUnitTest.java	(working copy)
@@ -5,17 +5,18 @@
 import junit.framework.Test;
 import junit.framework.TestSuite;
 
-import org.hibernate.junit.UnitTestCase;
+import org.hibernate.Hibernate;
+import org.hibernate.MappingException;
+import org.hibernate.cfg.NamingStrategy;
+import org.hibernate.cfg.ObjectNameNormalizer;
 import org.hibernate.dialect.Dialect;
+import org.hibernate.id.GeneratorProperties;
+import org.hibernate.id.PersistentIdentifierGenerator;
+import org.hibernate.id.enhanced.OptimizerFactory;
+import org.hibernate.id.enhanced.SequenceStructure;
 import org.hibernate.id.enhanced.SequenceStyleGenerator;
-import org.hibernate.id.enhanced.SequenceStructure;
-import org.hibernate.id.enhanced.OptimizerFactory;
 import org.hibernate.id.enhanced.TableStructure;
-import org.hibernate.id.PersistentIdentifierGenerator;
-import org.hibernate.Hibernate;
-import org.hibernate.MappingException;
-import org.hibernate.cfg.ObjectNameNormalizer;
-import org.hibernate.cfg.NamingStrategy;
+import org.hibernate.junit.UnitTestCase;
 
 /**
  * Tests that SequenceStyleGenerator configures itself as expected
@@ -47,9 +48,8 @@
 	}
 
 	private Properties buildGeneratorPropertiesBase() {
-		Properties props = new Properties();
-		props.put(
-				PersistentIdentifierGenerator.IDENTIFIER_NORMALIZER,
+		GeneratorProperties props = new GeneratorProperties();
+		props.setNormalizer(
 				new ObjectNameNormalizer() {
 					protected boolean isUseQuotedIdentifiersGlobally() {
 						return false;
