Index: src/net/sf/hibernate/persister/EntityPersister.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/persister/EntityPersister.java,v
retrieving revision 1.38.2.43
diff -u -r1.38.2.43 EntityPersister.java
--- src/net/sf/hibernate/persister/EntityPersister.java	1 May 2004 09:01:51 -0000	1.38.2.43
+++ src/net/sf/hibernate/persister/EntityPersister.java	3 Aug 2004 16:13:16 -0000
@@ -654,7 +654,7 @@
 		try {
 		
 			final PreparedStatement statement;
-			if ( isVersioned() ) {
+			if ( isVersioned() && optimisticLockMode() >= Versioning.OPTIMISTIC_LOCK_VERSION ) {
 				statement = session.getBatcher().prepareStatement(sql);
 			}
 			else {
@@ -700,7 +700,7 @@
 				throw sqle;
 			}
 			finally {
-				if ( isVersioned() ) session.getBatcher().closeStatement(statement);
+				if ( isVersioned() && optimisticLockMode() >= Versioning.OPTIMISTIC_LOCK_VERSION ) session.getBatcher().closeStatement(statement);
 			}
 		}
 		catch (SQLException sqle) {
