Index: C:/workspace/Hibernate3/src/org/hibernate/event/def/AbstractSaveEventListener.java
===================================================================
--- C:/workspace/Hibernate3/src/org/hibernate/event/def/AbstractSaveEventListener.java	(revision 10138)
+++ C:/workspace/Hibernate3/src/org/hibernate/event/def/AbstractSaveEventListener.java	(working copy)
@@ -164,6 +164,17 @@
 					source.forceFlush( source.getPersistenceContext().getEntry(old) );
 				}
 				else {
+					log
+					.error("Two instances of different classes found for id ["
+							+ id
+							+ "]: cached object ["
+							+ old
+							+ "] is instanceof "
+							+ old.getClass().getCanonicalName()
+							+ " while object to be serialized ["
+							+ entity
+							+ "] is instanceof "
+							+ entity.getClass().getCanonicalName());
 					throw new NonUniqueObjectException( id, persister.getEntityName() );
 				}
 			}
Index: C:/workspace/Hibernate3/src/org/hibernate/loader/Loader.java
===================================================================
--- C:/workspace/Hibernate3/src/org/hibernate/loader/Loader.java	(revision 10138)
+++ C:/workspace/Hibernate3/src/org/hibernate/loader/Loader.java	(working copy)
@@ -1232,7 +1232,7 @@
 
 		if ( !persister.isInstance( object, session.getEntityMode() ) ) {
 			throw new WrongClassException( 
-					"loaded object was of wrong class", 
+					"loaded object was of wrong class " + object.getClass(), 
 					key.getIdentifier(), 
 					persister.getEntityName() 
 				);
