Index: core/src/main/java/org/hibernate/criterion/Restrictions.java
===================================================================
--- core/src/main/java/org/hibernate/criterion/Restrictions.java	(revision 16257)
+++ core/src/main/java/org/hibernate/criterion/Restrictions.java	(working copy)
@@ -62,6 +62,7 @@
 	 * @return Criterion
 	 */
 	public static SimpleExpression eq(String propertyName, Object value) {
+		if (value == null) return isNull(propertyName);
 		return new SimpleExpression(propertyName, value, "=");
 	}
 	/**
