Index: core/src/main/java/org/hibernate/criterion/Restrictions.java
===================================================================
--- core/src/main/java/org/hibernate/criterion/Restrictions.java	(revision 16257)
+++ core/src/main/java/org/hibernate/criterion/Restrictions.java	(working copy)
@@ -61,7 +61,8 @@
 	 * @param value
 	 * @return Criterion
 	 */
-	public static SimpleExpression eq(String propertyName, Object value) {
+	public static Criterion eq(String propertyName, Object value) {
+		if (value == null) return isNull(propertyName);
 		return new SimpleExpression(propertyName, value, "=");
 	}
 	/**
