--- src\org\hibernate\criterion\DetachedCriteria.java	Mon Dec 05 21:35:50 2005
+++ src\org\hibernate\criterion\DetachedCriteria.java	Sat Aug 25 16:20:45 2007
@@ -87,7 +87,7 @@ public class DetachedCriteria implements
 
 	public DetachedCriteria createCriteria(String associationPath, String alias)
 	throws HibernateException {
-		return new DetachedCriteria( impl, criteria.createCriteria(associationPath) );
+		return new DetachedCriteria( impl, criteria.createCriteria(associationPath, alias) );
 	}
 
 	public DetachedCriteria createCriteria(String associationPath)
