Index: ejb/src/java/org/hibernate/ejb/Ejb3Configuration.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/ejb/src/java/org/hibernate/ejb/Ejb3Configuration.java,v
retrieving revision 1.29
diff -u -r1.29 Ejb3Configuration.java
--- ejb/src/java/org/hibernate/ejb/Ejb3Configuration.java	13 Jan 2006 01:58:39 -0000	1.29
+++ ejb/src/java/org/hibernate/ejb/Ejb3Configuration.java	16 Jan 2006 18:11:16 -0000
@@ -215,7 +215,7 @@
 		Map workingVars = new HashMap();
 		workingVars.put( HibernatePersistence.TRANSACTION_TYPE, info.getTransactionType() );
 		boolean[] detectArtifact = getDetectedArtifacts( info.getProperties(), null );
-		for ( URL jar : info.getJarFiles() ) {
+		for ( URL jar : info.getJarFileUrls() ) {
 			if ( detectArtifact[0] ) scanForClasses( jar, packages, entities );
 			if ( detectArtifact[1] ) scanForHbmXmlFiles( jar, hbmFiles );
 		}
Index: ejb/src/java/org/hibernate/ejb/util/LogHelper.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/ejb/src/java/org/hibernate/ejb/util/LogHelper.java,v
retrieving revision 1.1
diff -u -r1.1 LogHelper.java
--- ejb/src/java/org/hibernate/ejb/util/LogHelper.java	30 Dec 2005 09:49:47 -0000	1.1
+++ ejb/src/java/org/hibernate/ejb/util/LogHelper.java	16 Jan 2006 18:11:16 -0000
@@ -26,7 +26,7 @@
 				.append("Transaction type: ").append( unitInfo.getTransactionType() ).append("\n\t")
 				.append("PU root URL: ").append( unitInfo.getPersistenceUnitRootUrl() ).append("\n\t");
 		sb.append("Jar files URLs [");
-		List<URL> jarFileUrls = unitInfo.getJarFiles();
+		List<URL> jarFileUrls = unitInfo.getJarFileUrls();
 		for(URL url : jarFileUrls) {
 			sb.append("\n\t\t").append( url );
 		}
Index: ejb-api/src/javax/persistence/spi/PersistenceUnitInfo.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/ejb-api/src/javax/persistence/spi/PersistenceUnitInfo.java,v
retrieving revision 1.1
diff -u -r1.1 PersistenceUnitInfo.java
--- ejb-api/src/javax/persistence/spi/PersistenceUnitInfo.java	29 Dec 2005 02:43:37 -0000	1.1
+++ ejb-api/src/javax/persistence/spi/PersistenceUnitInfo.java	16 Jan 2006 18:11:18 -0000
@@ -70,7 +70,7 @@
     * will always be the last entry in the list.
     * Each jar file URL corresponds to a named <jar-file> element in persistence.xml
     */
-   public List<URL> getJarFiles();
+   public List<URL> getJarFileUrls();
 
    /**
     * @return the URL or directory that is the root of the persistence unit.
