Index: core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaExportTask.java
===================================================================
--- core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaExportTask.java	(revision 20876)
+++ core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaExportTask.java	(working copy)
@@ -238,10 +238,13 @@
 			properties.load( new FileInputStream(propertiesFile) );
 		}
 		cfg.setProperties(properties);
-		return new SchemaExport(cfg)
-				.setHaltOnError(haltOnError)
-				.setOutputFile( outputFile.getPath() )
-				.setDelimiter(delimiter);
+		SchemaExport se = new SchemaExport(cfg)
+			.setHaltOnError(haltOnError)
+			.setDelimiter(delimiter);
+		if (outputFile != null){
+			se.setOutputFile( outputFile.getPath() );
+		}
+		return se;
 	}
 
 	public void setNamingStrategy(String namingStrategy) {
Index: core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaUpdateTask.java
===================================================================
--- core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaUpdateTask.java	(revision 20876)
+++ core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaUpdateTask.java	(working copy)
@@ -202,7 +202,9 @@
 		}
 		cfg.setProperties(properties);
 		SchemaUpdate su = new SchemaUpdate(cfg);
-		su.setOutputFile( outputFile.getPath() );
+		if (outputFile != null){
+			su.setOutputFile( outputFile.getPath() );
+		}
 		su.setDelimiter(delimiter);
 		su.setHaltOnError(haltOnError);
 		return su;
