Index: src/org/hibernate/event/def/AbstractSaveEventListener.java
===================================================================
--- src/org/hibernate/event/def/AbstractSaveEventListener.java	(revision 4318)
+++ src/org/hibernate/event/def/AbstractSaveEventListener.java	(revision 4749)
@@ -21,6 +22,7 @@
 import org.hibernate.engine.Nullability;
 import org.hibernate.engine.SessionImplementor;
 import org.hibernate.engine.Status;
+import org.hibernate.engine.ValueInclusion;
 import org.hibernate.engine.Versioning;
 import org.hibernate.event.EventSource;
 import org.hibernate.id.IdentifierGenerationException;
@@ -329,6 +340,27 @@
 			source.getActionQueue().addAction(
 					new EntityInsertAction( id, values, entity, version, persister, source )
 			);
+
+	        /*
+	         * Bugfix for HHH-2588 start: if the entity has "generated at insert time" properties
+	         * we need to insert this record immediately. The insert will cause the properties that
+	         * are generated to be read back and pushed into the entity before follow-up updates
+	         * are done which need those generated properties.
+	         */
+	        ValueInclusion[]  viar = persister.getPropertyInsertGenerationInclusions();
+	        if(viar != null && viar.length > 0) {
+	        	boolean found = false;
+	        	for(int i = viar.length; --i >= 0;) {
+	        		if(viar[i] != ValueInclusion.NONE) {
+	        			found = true;
+	        			break;
+	        		}
+	        	}
+	        	if(found) {
+		            log.trace( "executing insertions" );
+		            source.getActionQueue().executeInserts();
+	        	}
+	        }
 		}
 
 		cascadeAfterSave( source, persister, entity, anything );
