Index: Category.java
===================================================================
retrieving revision 1.1.2.1
diff -u -r1.1.2.1 Category.java
--- Category.java	8 Nov 2003 02:25:29 -0000	1.1.2.1
+++ Category.java	2 Feb 2004 11:33:45 -0000
@@ -7,7 +7,7 @@
 /**
  *
  */
-public class Category {
+public class Category extends PolymorphCriteriaBase {
 	
 	public static final String ROOT_CATEGORY = "/";
 	
Index: MasterDetailTest.java
===================================================================
retrieving revision 1.1.2.9
diff -u -r1.1.2.9 MasterDetailTest.java
--- MasterDetailTest.java	19 Jan 2004 18:00:32 -0000	1.1.2.9
+++ MasterDetailTest.java	2 Feb 2004 11:33:50 -0000
@@ -60,6 +60,13 @@
 		assertTrue( "Eye Eye".equals( jay.getEye().getName() ) );
 		s.connection().commit();
 		s.close();
+		
+		// cleanup
+		s = openSession();
+		s.delete(jay);
+		s.delete(e);
+		s.connection().commit();
+		s.close();
 	}
 	
 	public void testMeta() throws Exception {
@@ -71,21 +78,21 @@
 	}
 	
 	public void testPolymorphicCriteria() throws Exception {
-		Session s = openSession();
-		Category f = new Category();
-		Single b = new Single();
-		b.setId("asdfa");
-		b.setString("asdfasdf");
-		s.save(f);
-		s.save(b);
-		List list = s.createCriteria(Object.class).list();
-		assertTrue( list.size()==2 );
-		assertTrue( list.contains(f) && list.contains(b) );
-		s.delete(f);
-		s.delete(b);
-		s.flush();
-		s.connection().commit();
-		s.close();
+        Session s = openSession();
+        Category f = new Category();
+        Single b = new Single();
+        b.setId("asdfa");
+        b.setString("asdfasdf");
+        s.save(f);
+        s.save(b);
+        List list = s.createCriteria(PolymorphCriteriaBase.class).list();
+        assertTrue( list.size()==2 );
+        assertTrue( list.contains(f) && list.contains(b) );
+        s.delete(f);
+        s.delete(b);
+        s.flush();
+        s.connection().commit();
+        s.close();
 	}
 	
 	public void testCopy() throws Exception {
Index: Single.java
===================================================================
retrieving revision 1.1.2.2
diff -u -r1.1.2.2 Single.java
--- Single.java	15 Jan 2004 07:53:31 -0000	1.1.2.2
+++ Single.java	2 Feb 2004 11:33:51 -0000
@@ -6,7 +6,7 @@
 import java.util.HashSet;
 
 
-public class Single implements Serializable {
+public class Single extends PolymorphCriteriaBase implements Serializable {
 	private String id;
 	private String prop;
 	private String string;
Index: test/org/hibernate/test/PolymorphCriteriaBase.java
===================================================================
RCS file: test/org/hibernate/test/PolymorphCriteriaBase.java
diff -N test/org/hibernate/test/PolymorphCriteriaBase.java
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ test/org/hibernate/test/PolymorphCriteriaBase.java	1 Jan 1970 00:00:00 -0000
@@ -0,0 +1,9 @@
+package org.hibernate.test;
+
+/**
+ * Common base class for persistent objects used in the PolymorphicCriteria test
+ * */
+public class PolymorphCriteriaBase {
+
+
+}
