Index: src/main/java/org/hibernate/envers/entities/mapper/relation/ToOneIdMapper.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/ToOneIdMapper.java	(revision 17249)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/ToOneIdMapper.java	Thu Sep 17 16:22:49 EEST 2009
@@ -82,7 +82,7 @@
                 Class<?> entityClass = ReflectionTools.loadClass(referencedEntityName);
 
                 value = versionsReader.getSessionImplementor().getFactory().getEntityPersister(referencedEntityName).
-                        createProxy(null, new ToOneDelegateSessionImplementor(versionsReader, entityClass, entityId, revision, verCfg));
+                        createProxy((Serializable)entityId, new ToOneDelegateSessionImplementor(versionsReader, entityClass, entityId, revision, verCfg));
             }
         }
 
