--- //depot/ThirdParty/main/hibernate4/project/hibernate-entitymanager/src/main/java/org/hibernate/engine/spi/EJB3CascadingAction.java	2012-04-06 20:11:16.000000000 -0700
+++ c:\commander\ThirdParty\main\hibernate4\project\hibernate-entitymanager\src\main\java\org\hibernate\engine\spi\EJB3CascadingAction.java	2012-04-06 20:11:16.000000000 -0700
--- C:\cygwin\tmp\t9628t61.tmp	2012-06-26 18:31:24.643088900 -0700
+++ c:\commander\ThirdParty\main\hibernate4\project\hibernate-entitymanager\src\main\java\org\hibernate\engine\spi\EJB3CascadingAction.java	2012-06-26 18:21:11.557786500 -0700
@@ -46,7 +46,7 @@
 	 */
 	public static final CascadingAction PERSIST_SKIPLAZY = new CascadingAction() {
 		@Override
-        public void cascade(EventSource session, Object child, String entityName, Object anything, boolean isCascadeDeleteEnabled)
+        public void cascade(EventSource session, Object parent, Object child, String entityName, Object anything, boolean isCascadeDeleteEnabled)
 		throws HibernateException {
             LOG.trace("Cascading to persist: " + entityName);
 			session.persist( entityName, child, (Map) anything );
