Index: testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/Employee.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/Employee.java	Thu Jan 14 23:26:45 CST 2010
+++ testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/Employee.java	Thu Jan 14 23:26:45 CST 2010
@@ -0,0 +1,50 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * Copyright (c) 2010, Red Hat Inc. or third-party contributors as
+ * indicated by the @author tags or express copyright attribution
+ * statements applied by the authors.  All third-party contributions are
+ * distributed under license by Red Hat Inc.
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.hibernate.test.orphan.one2one.fk.component;
+
+/**
+ * TODO : javadoc
+ *
+ * @author Steve Ebersole
+ */
+public class Employee {
+	private Long id;
+	private EmployeeSubData subData;
+
+	public Long getId() {
+		return id;
+	}
+
+	public void setId(Long id) {
+		this.id = id;
+	}
+
+	public EmployeeSubData getSubData() {
+		return subData;
+	}
+
+	public void setSubData(EmployeeSubData subData) {
+		this.subData = subData;
+	}
+}
\ No newline at end of file
Index: testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/Mapping.hbm.xml
===================================================================
--- testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/Mapping.hbm.xml	Thu Jan 14 23:26:45 CST 2010
+++ testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/Mapping.hbm.xml	Thu Jan 14 23:26:45 CST 2010
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Hibernate, Relational Persistence for Idiomatic Java
+  ~
+  ~ Copyright (c) 2010, Red Hat Inc. or third-party contributors as
+  ~ indicated by the @author tags or express copyright attribution
+  ~ statements applied by the authors.  All third-party contributions are
+  ~ distributed under license by Red Hat Inc.
+  ~
+  ~ This copyrighted material is made available to anyone wishing to use, modify,
+  ~ copy, or redistribute it subject to the terms and conditions of the GNU
+  ~ Lesser General Public License, as published by the Free Software Foundation.
+  ~
+  ~ This program is distributed in the hope that it will be useful,
+  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+  ~ or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+  ~ for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public License
+  ~ along with this distribution; if not, write to:
+  ~ Free Software Foundation, Inc.
+  ~ 51 Franklin Street, Fifth Floor
+  ~ Boston, MA  02110-1301  USA
+  -->
+
+<!DOCTYPE hibernate-mapping PUBLIC
+        '-//Hibernate/Hibernate Mapping DTD 3.0//EN'
+        'http://hibernate.sourceforge.net/hibernate-mapping-3.0.dtd'>
+
+<hibernate-mapping package="org.hibernate.test.orphan.one2one.fk.component" >
+
+    <class name="Employee">
+        <id name="id" type="long" column="id">
+            <generator class="increment" />
+        </id>
+        <component name="subData" class="EmployeeSubData">
+            <one-to-one name="info"
+                        property-ref="employee"
+                        class="EmployeeInfo"
+                        cascade="all,delete-orphan"
+                        constrained="false" />
+        </component>
+    </class>
+
+    <class name="EmployeeInfo">
+        <id name="id" type="long" column="id">
+            <generator class="increment" />
+        </id>
+        <many-to-one name="employee"
+                     column="employee_id"
+                     unique="true"
+                     not-null="true" />
+    </class>
+
+</hibernate-mapping>
Index: testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/EmployeeSubData.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/EmployeeSubData.java	Thu Jan 14 23:28:57 CST 2010
+++ testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/EmployeeSubData.java	Thu Jan 14 23:28:57 CST 2010
@@ -0,0 +1,48 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * Copyright (c) 2010, Red Hat Inc. or third-party contributors as
+ * indicated by the @author tags or express copyright attribution
+ * statements applied by the authors.  All third-party contributions are
+ * distributed under license by Red Hat Inc.
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.hibernate.test.orphan.one2one.fk.component;
+
+/**
+ * TODO : javadoc
+ *
+ * @author Steve Ebersole
+ */
+public class EmployeeSubData {
+	private EmployeeInfo info;
+
+	public EmployeeSubData() {
+	}
+
+	public EmployeeSubData(EmployeeInfo info) {
+		this.info = info;
+	}
+
+	public EmployeeInfo getInfo() {
+		return info;
+	}
+
+	public void setInfo(EmployeeInfo info) {
+		this.info = info;
+	}
+}
Index: testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/DeleteOneToOneOrphansTest.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/DeleteOneToOneOrphansTest.java	Sat Jan 16 13:15:34 CST 2010
+++ testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/DeleteOneToOneOrphansTest.java	Sat Jan 16 13:15:34 CST 2010
@@ -0,0 +1,93 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * Copyright (c) 2010, Red Hat Inc. or third-party contributors as
+ * indicated by the @author tags or express copyright attribution
+ * statements applied by the authors.  All third-party contributions are
+ * distributed under license by Red Hat Inc.
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.hibernate.test.orphan.one2one.fk.component;
+
+import java.util.List;
+
+import org.hibernate.Session;
+import org.hibernate.junit.functional.FunctionalTestCase;
+
+/**
+ * TODO : javadoc
+ *
+ * @author Steve Ebersole
+ */
+public class DeleteOneToOneOrphansTest extends FunctionalTestCase {
+	public DeleteOneToOneOrphansTest(String string) {
+		super( string );
+	}
+
+	public String[] getMappings() {
+		return new String[] { "orphan/one2one/fk/component/Mapping.hbm.xml" };
+	}
+
+	private void createData() {
+		Session session = openSession();
+		session.beginTransaction();
+		Employee emp = new Employee();
+		emp.setSubData( new EmployeeSubData( new EmployeeInfo( emp ) ) );
+		session.save( emp );
+		session.getTransaction().commit();
+		session.close();
+	}
+
+	private void cleanupData() {
+		Session session = openSession();
+		session.beginTransaction();
+		session.createQuery( "delete EmployeeInfo" ).executeUpdate();
+		session.createQuery( "delete Employee" ).executeUpdate();
+		session.getTransaction().commit();
+		session.close();
+	}
+
+	public void testOrphanedWhileManagedFailureExpected() {
+		createData();
+
+		Session session = openSession();
+		session.beginTransaction();
+		List results = session.createQuery( "from EmployeeInfo" ).list();
+		assertEquals( 1, results.size() );
+		results = session.createQuery( "from Employee" ).list();
+		assertEquals( 1, results.size() );
+		Employee emp = ( Employee ) results.get( 0 );
+		assertNotNull( emp.getSubData() );
+		assertNotNull( emp.getSubData().getInfo() );
+		emp.getSubData().setInfo( null );
+		session.getTransaction().commit();
+		session.close();
+
+		session = openSession();
+		session.beginTransaction();
+		emp = ( Employee ) session.get( Employee.class, emp.getId() );
+		assertNull( emp.getSubData() );
+		results = session.createQuery( "from EmployeeInfo" ).list();
+		assertEquals( 0, results.size() );
+		results = session.createQuery( "from Employee" ).list();
+		assertEquals( 1, results.size() );
+		session.getTransaction().commit();
+		session.close();
+
+		cleanupData();
+	}
+}
\ No newline at end of file
Index: testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/EmployeeInfo.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/EmployeeInfo.java	Thu Jan 14 23:25:23 CST 2010
+++ testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/component/EmployeeInfo.java	Thu Jan 14 23:25:23 CST 2010
@@ -0,0 +1,57 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * Copyright (c) 2010, Red Hat Inc. or third-party contributors as
+ * indicated by the @author tags or express copyright attribution
+ * statements applied by the authors.  All third-party contributions are
+ * distributed under license by Red Hat Inc.
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.hibernate.test.orphan.one2one.fk.component;
+
+/**
+ * TODO : javadoc
+ *
+ * @author Steve Ebersole
+ */
+public class EmployeeInfo {
+	private Long id;
+	private Employee employee;
+
+	public EmployeeInfo() {
+	}
+
+	public EmployeeInfo(Employee employee) {
+		this.employee = employee;
+	}
+
+	public Long getId() {
+		return id;
+	}
+
+	public void setId(Long id) {
+		this.id = id;
+	}
+
+	public Employee getEmployee() {
+		return employee;
+	}
+
+	public void setEmployee(Employee employee) {
+		this.employee = employee;
+	}
+}
\ No newline at end of file
