Index: CollectionType.java
===================================================================
--- CollectionType.java	(revision 10633)
+++ CollectionType.java	(working copy)
@@ -404,8 +404,23 @@
 
 		// TODO: does not work for EntityMode.DOM4J yet!
 
-		java.util.Collection result = (java.util.Collection) target;
+        java.util.Collection result = (java.util.Collection) target;
 
+        /*
+         * PATCH for JIRA HHH-2162.
+         * If the target collection is the same as the original there's
+         * no need to perform the copy. Avoiding the copy in this apparently no-op
+         * case avoids the collection being marked dirty when it doesn't in fact
+         * changed. This in turn avoids the entity's version being spuriously
+         * increased leading to frequent unwarranted OptimisticLockException failures,
+         * even in single user interactions.
+         */
+        if ( ((java.util.Collection)original).size() == result.size() &&
+                result.containsAll((java.util.Collection) original )) {
+            return result;
+        }
+        /* END PATCH */
+
 		result.clear();
 
 		// copy elements into newly empty target collection
