--- EnumType.java.orig	2010-04-22 11:32:22.343750000 +0200
+++ EnumType.java	2010-04-22 11:45:16.187000000 +0200
@@ -24,7 +24,6 @@
 package org.hibernate.type;
 
 import java.io.Serializable;
-import java.lang.reflect.Method;
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
@@ -203,8 +203,8 @@
 		catch (ClassNotFoundException exception) {
 			throw new HibernateException( "Enum class not found", exception );
 		}
-		// is might be good to call it here, to see a possible error immediatelly
-		// initEnumValue();
+		// it may be good to call it here, to see a possible error immediately
+		// initEnumValues();
 		
 		//nullify unnullified properties yuck!
 		schema = parameters.getProperty( SCHEMA );
@@ -224,13 +224,10 @@
 	 * Lazy init of {@link #enumValues}.
 	 */
 	private void initEnumValues() {
-		if ( enumValues == null ) {
-			try {
-				Method method = enumClass.getDeclaredMethod( "values" );
-				enumValues = (Object[]) method.invoke( null );
-			}
-			catch (Exception e) {
-				throw new HibernateException( "Error while accessing enum.values(): " + enumClass, e );
+		if (enumValues == null) {
+			this.enumValues = enumClass.getEnumConstants();
+			if (enumValues == null) {
+				throw new NullPointerException("Failed to init enumValues");
 			}
 		}
 	}
