--- testsuite/src/test/java/org/hibernate/test/cfg/CacheableFileTest.java	(revision 13962)
+++ testsuite/src/test/java/org/hibernate/test/cfg/CacheableFileTest.java	Thu Aug 30 11:17:29 EDT 2007
@@ -22,7 +22,7 @@
 
 	protected void setUp() throws Exception {
 		super.setUp();
-		mappingFile = new File( getClass().getClassLoader().getResource( MAPPING ).getFile() );
+		mappingFile = new File( getClass().getClassLoader().getResource( MAPPING ).toURI() );
 		assertTrue( mappingFile.exists() );
 		File cached = new File( mappingFile.getParentFile(), mappingFile.getName() + ".bin" );
 		if ( cached.exists() ) {
