diff -ru hibernate-3.5.2-Final/org/hibernate/persister/entity/AbstractEntityPersister.java oss-workspace/Hibernate-3.5.2-CDM/src/org/hibernate/persister/entity/AbstractEntityPersister.java
--- hibernate-3.5.2-Final/org/hibernate/persister/entity/AbstractEntityPersister.java	2010-05-11 19:11:28.000000000 -0700
+++ oss-workspace/Hibernate-3.5.2-CDM/src/org/hibernate/persister/entity/AbstractEntityPersister.java	2010-05-18 16:01:45.000000000 -0700
@@ -1338,7 +1338,7 @@
 			update.setComment( "forced version increment" );
 		}
 		update.addColumn( getVersionColumnName() );
-		update.addPrimaryKeyColumns( getIdentifierColumnNames() );
+		update.addPrimaryKeyColumns( getRootTableIdentifierColumnNames() );
 		update.setVersionColumnName( getVersionColumnName() );
 		return update.toStatementString();
 	}
