Index: src/org/hibernate/tool/hbm2ddl/DatabaseMetadata.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/tool/hbm2ddl/DatabaseMetadata.java,v
retrieving revision 1.14
diff -u -r1.14 DatabaseMetadata.java
--- src/org/hibernate/tool/hbm2ddl/DatabaseMetadata.java	21 Dec 2005 22:24:41 -0000	1.14
+++ src/org/hibernate/tool/hbm2ddl/DatabaseMetadata.java	28 Dec 2005 22:25:06 -0000
@@ -142,9 +142,9 @@
 			} else {
 				String[] strings = StringHelper.split(".", (String) key);
 				if(strings.length==3) {
-					return getTableMetadata(strings[2], strings[1], strings[2]) != null;
+					return getTableMetadata(strings[2], strings[1], strings[0]) != null;
 				} else if (strings.length==2) {
-					return getTableMetadata(strings[2], strings[1], null) != null;
+					return getTableMetadata(strings[1], strings[0], null) != null;
 				}
 			}
 		}
