Index: src/org/hibernate/tuple/PojoTuplizer.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/tuple/PojoTuplizer.java,v
retrieving revision 1.13
diff -u -r1.13 PojoTuplizer.java
--- src/org/hibernate/tuple/PojoTuplizer.java	28 Feb 2005 20:24:14 -0000	1.13
+++ src/org/hibernate/tuple/PojoTuplizer.java	19 Mar 2005 17:09:27 -0000
@@ -115,7 +115,7 @@
 		Iterator iter = persistentClass.getSubclassIterator();
 		while ( iter.hasNext() ) {
 			Subclass subclass = ( Subclass ) iter.next();
-			if ( !subclass.getEntityName().equals( subclass.getProxyInterfaceName() ) ) {
+			if ( !subclass.getClassName().equals( subclass.getProxyInterfaceName() ) ) {
 				Class subclassProxy = subclass.getProxyInterface();
 				if ( subclassProxy == null ) {
 					throw new MappingException( "All subclasses must also have proxies: " + getEntityName() );
