Index: src/org/hibernate/hql/ast/tree/DotNode.java
===================================================================
--- src/org/hibernate/hql/ast/tree/DotNode.java	(revision 11694)
+++ src/org/hibernate/hql/ast/tree/DotNode.java	(working copy)
@@ -434,7 +434,8 @@
 			);
 		}
 		else {
-			currentFromClause.addDuplicateAlias(classAlias, elem);
+			if (classAlias != null)
+				currentFromClause.addDuplicateAlias(classAlias, elem);
 		}
 		setImpliedJoin( elem );
 		getWalker().addQuerySpaces( elem.getEntityPersister().getQuerySpaces() );
