diff -ur orig/org/hibernate/cfg/Configuration.java new/org/hibernate/cfg/Configuration.java
--- orig/org/hibernate/cfg/Configuration.java	2008-10-29 14:27:00.000000000 -0500
+++ new/org/hibernate/cfg/Configuration.java	2008-10-29 14:29:38.000000000 -0500
@@ -1193,6 +1193,14 @@
 			}
 		}
 
+		log.debug( "updating table unique integers" );
+
+		int uniqueInteger = 0;
+		iter = getTableMappings();
+		while ( iter.hasNext() ) {
+			( (Table) iter.next() ).setUniqueInteger(uniqueInteger++);
+		}
+
 		//TODO: Somehow add the newly created foreign keys to the internal collection
 
 		log.debug( "processing foreign key constraints" );
diff -ur orig/org/hibernate/mapping/Table.java new/org/hibernate/mapping/Table.java
--- orig/org/hibernate/mapping/Table.java	2008-10-29 14:26:55.000000000 -0500
+++ new/org/hibernate/mapping/Table.java	2008-10-29 14:28:24.000000000 -0500
@@ -58,7 +58,7 @@
 	private Map indexes = new HashMap();
 	private Map foreignKeys = new HashMap();
 	private Map uniqueKeys = new HashMap();
-	private final int uniqueInteger;
+	private int uniqueInteger;
 	private boolean quoted;
 	private boolean schemaQuoted;
 	private static int tableCounter = 0;
@@ -660,6 +660,10 @@
 		return uniqueInteger;
 	}
 
+	public void setUniqueInteger(int uniqueInteger) {
+		this.uniqueInteger = uniqueInteger;
+	}
+
 	public void setIdentifierValue(KeyValue idValue) {
 		this.idValue = idValue;
 	}
