Index: src/org/hibernate/tuple/entity/EntityMetamodel.java
===================================================================
--- src/org/hibernate/tuple/entity/EntityMetamodel.java (revision 10964)
+++ src/org/hibernate/tuple/entity/EntityMetamodel.java (working copy)
@@ -116,10 +116,14 @@
                rootName = persistentClass.getRootClass().getEntityName();
                entityType = TypeFactory.manyToOne( name );
 
+               try {
                identifierProperty = PropertyFactory.buildIdentifierProperty(
                        persistentClass,
                        sessionFactory.getIdentifierGenerator( rootName )
                        );
+               } catch (UnsupportedOperationException ex) {
+                       throw new UnsupportedOperationException("Couldn't get identifier for: " + rootName, ex);
+               }
 
                versioned = persistentClass.isVersioned();

