Index: PostgreSQLDialect.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/dialect/PostgreSQLDialect.java,v
retrieving revision 1.22
diff -u -r1.22 PostgreSQLDialect.java
--- PostgreSQLDialect.java	8 Aug 2004 08:23:16 -0000	1.22
+++ PostgreSQLDialect.java	17 Jan 2005 23:50:10 -0000
@@ -111,6 +111,10 @@
 		return "";//" cascade";
 	}
 
+    public String getQuerySequencesString() {
+        return "select relname as \"SEQUENCE_NAME\" from pg_class where relkind = 'S'";
+    }
+
 	public boolean supportsSequences() {
 		return true;
 	}
