--- C:/Devel/projects/hibernate-core/hibernate-core/src/main/java/org/hibernate/cache/QueryKey.java	Sun Jan 30 12:31:39 2011
+++ C:/Devel/projects/hibernate-core/hibernate-core/src/main/java/org/hibernate/cache/QueryKey2.java	Sun Jan 30 12:25:26 2011
@@ -181,6 +181,10 @@
 		return customTransformer;
 	}
 
+	public Map getNamedParameters() {
+		return namedParameters;
+	}
+
 	/**
 	 * Deserialization hook used to re-init the cached hashcode which is needed for proper clustering support.
 	 *
