Index: src/org/hibernate/transform/ToListResultTransformer.java
===================================================================
--- src/org/hibernate/transform/ToListResultTransformer.java	(Revision 9893)
+++ src/org/hibernate/transform/ToListResultTransformer.java	(Arbeitskopie)
@@ -17,4 +17,11 @@
 		return collection;
 	}
 
+        public boolean equals(Object obj) {
+                return obj != null && obj.getClass().equals(getClass());
+        }
+
+        public int hashCode() {
+                return getClass().hashCode();
+        }
 }
Index: src/org/hibernate/transform/RootEntityResultTransformer.java
===================================================================
--- src/org/hibernate/transform/RootEntityResultTransformer.java	(Revision 9893)
+++ src/org/hibernate/transform/RootEntityResultTransformer.java	(Arbeitskopie)
@@ -16,4 +16,11 @@
 		return collection;
 	}
 
+	public boolean equals(Object obj) {
+		return obj != null && obj.getClass().equals(getClass());
+	}
+
+	public int hashCode() {
+		return getClass().hashCode();
+	}
 }
Index: src/org/hibernate/transform/AliasToBeanResultTransformer.java
===================================================================
--- src/org/hibernate/transform/AliasToBeanResultTransformer.java	(Revision 9893)
+++ src/org/hibernate/transform/AliasToBeanResultTransformer.java	(Arbeitskopie)
@@ -78,4 +78,12 @@
 		return collection;
 	}
 
+        public boolean equals(Object obj) {
+                return obj != null && obj.getClass().equals(getClass()) &&
+			resultClass.equals(((AliasToBeanResultTransformer) obj).resultClass);
+        }
+
+        public int hashCode() {
+                return getClass().hashCode() * 37 + resultClass.hashCode();
+        }
 }
Index: src/org/hibernate/transform/DistinctRootEntityResultTransformer.java
===================================================================
--- src/org/hibernate/transform/DistinctRootEntityResultTransformer.java	(Revision 9893)
+++ src/org/hibernate/transform/DistinctRootEntityResultTransformer.java	(Arbeitskopie)
@@ -51,4 +51,11 @@
 		return result;
 	}
 
+        public boolean equals(Object obj) {
+                return obj != null && obj.getClass().equals(getClass());
+        }
+
+        public int hashCode() {
+                return getClass().hashCode();
+        }
 }
Index: src/org/hibernate/transform/AliasToEntityMapResultTransformer.java
===================================================================
--- src/org/hibernate/transform/AliasToEntityMapResultTransformer.java	(Revision 9893)
+++ src/org/hibernate/transform/AliasToEntityMapResultTransformer.java	(Arbeitskopie)
@@ -25,4 +25,11 @@
 		return collection;
 	}
 
+        public boolean equals(Object obj) {
+                return obj != null && obj.getClass().equals(getClass());
+        }
+
+        public int hashCode() {
+                return getClass().hashCode();
+        }
 }
Index: src/org/hibernate/transform/PassThroughResultTransformer.java
===================================================================
--- src/org/hibernate/transform/PassThroughResultTransformer.java	(Revision 9893)
+++ src/org/hibernate/transform/PassThroughResultTransformer.java	(Arbeitskopie)
@@ -16,4 +16,11 @@
 		return collection;
 	}
 
+        public boolean equals(Object obj) {
+                return obj != null && obj.getClass().equals(getClass());
+        }
+
+        public int hashCode() {
+                return getClass().hashCode();
+        }
 }
