Index: src/java/org/hibernate/cfg/OneToOneSecondPass.java
===================================================================
--- src/java/org/hibernate/cfg/OneToOneSecondPass.java	(revision 18582)
+++ src/java/org/hibernate/cfg/OneToOneSecondPass.java	(working copy)
@@ -146,8 +146,9 @@
 				Iterator it = otherSide.getJoinIterator();
 				Join otherSideJoin = null;
 				while ( it.hasNext() ) {
-					otherSideJoin = (Join) it.next();
-					if ( otherSideJoin.containsProperty( otherSideProperty ) ) {
+					Join otherSideJoinValue = (Join) it.next();
+					if ( otherSideJoinValue.containsProperty( otherSideProperty ) ) {
+						otherSideJoin = otherSideJoinValue;
 						break;
 					}
 				}
Index: pom.xml
===================================================================
--- pom.xml	(revision 18582)
+++ pom.xml	(working copy)
@@ -8,7 +8,7 @@
 	<groupId>org.hibernate</groupId>
 	<artifactId>hibernate-annotations</artifactId>
 	<packaging>jar</packaging>
-	<version>@version@</version>
+	<version>3.4.0.1.GA</version>
 
 	<name>Hibernate Annotations</name>
 	<description>Annotations metadata for Hibernate</description>
@@ -35,7 +35,7 @@
         <dependency>
 			<groupId>org.hibernate</groupId>
 			<artifactId>hibernate-core</artifactId>
-			<version>3.3.0.SP1</version>
+			<version>3.3.2.GA</version>
 		</dependency>
 		<dependency>
             <groupId>org.slf4j</groupId>
