Index: src/org/hibernate/loader/hql/QueryLoader.java
===================================================================
--- src/org/hibernate/loader/hql/QueryLoader.java	(revision 10700)
+++ src/org/hibernate/loader/hql/QueryLoader.java	(working copy)
@@ -501,7 +501,7 @@
 			ParameterSpecification spec = ( ParameterSpecification ) itr.next();
 			position += spec.bind( st, queryParameters, session, position );
 		}
-		return position;
+		return startPosition - position;
 	}
 
 	private int bindFilterParameterValues(
