--- ../tmp/hibernate-3.0/src/org/hibernate/cfg/HbmBinder.java	2005-05-25 10:32:09.000000000 +0530
+++ src/org/hibernate/cfg/HbmBinder.java	2006-05-17 13:41:39.000000000 +0530
@@ -539,7 +539,17 @@
 
 		// IMPORT
 		mappings.addImport( entity.getEntityName(), entity.getEntityName() );
-		if ( mappings.isAutoImport() && entity.getEntityName().indexOf( '.' ) > 0 ) {
+		
+		boolean unqualify = mappings.isAutoImport() && entity.getEntityName().indexOf( '.' ) > 0;
+		
+		Attribute autoImportNode = node.attribute("auto-import");
+		
+		if(autoImportNode != null)
+		{
+			unqualify = autoImportNode.getValue().equalsIgnoreCase("true");
+		}
+		
+		if(unqualify) {
 			mappings.addImport( entity.getEntityName(), StringHelper.unqualify( entity
 				.getEntityName() ) );
 		}
