Index: testsuite/src/test/java/org/hibernate/test/criteria/CriteriaQueryTest.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/criteria/CriteriaQueryTest.java	(revision 20006)
+++ testsuite/src/test/java/org/hibernate/test/criteria/CriteriaQueryTest.java	(working copy)
@@ -45,7 +45,7 @@
 	}
 
 	public String[] getMappings() {
-		return new String[] { "criteria/Enrolment.hbm.xml","criteria/Foo.hbm.xml", "hql/Animal.hbm.xml" };
+		return new String[] { "criteria/Enrolment.hbm.xml","criteria/Foo.hbm.xml", "hql/Animal.hbm.xml","criteria/KeyManyToOne.hbm.xml" };
 	}
 
 	public void configure(Configuration cfg) {
@@ -1584,5 +1584,22 @@
 		t.commit();
 		session.close();
 	}
+	
+	/**
+	 * Test for HHH-1570
+	 */
+	public void testJoinKeyManyToOne() {
+		Session session = openSession();
+		Transaction t = session.beginTransaction();
+
+		session.createCriteria(LeftTable.class)
+			.createCriteria("middleTables",	"mid")
+			.createCriteria("mid.right", "right")
+			.add(Restrictions.eq("right.value", "foo"))
+			.list();
+
+		t.commit();
+		session.close();
+	}
 }
 
Index: testsuite/src/test/java/org/hibernate/test/criteria/KeyManyToOne.hbm.xml
===================================================================
--- testsuite/src/test/java/org/hibernate/test/criteria/KeyManyToOne.hbm.xml	(revision 0)
+++ testsuite/src/test/java/org/hibernate/test/criteria/KeyManyToOne.hbm.xml	(revision 0)
@@ -0,0 +1,35 @@
+<?xml version="1.0"?>
+<!DOCTYPE hibernate-mapping PUBLIC "-//Hibernate/Hibernate Mapping DTD 3.0//EN" "http://hibernate.sourceforge.net/hibernate-mapping-3.0.dtd">
+
+<hibernate-mapping package="org.hibernate.test.criteria">
+	<class name="LeftTable" table="LEFT">
+		<id name="id" access="field">
+			<column name="ID" />
+			<generator class="native" />
+		</id>
+
+		<set name="middleTables" access="field">
+            <key>
+                <column name="ID"></column>
+            </key>
+            <one-to-many class="MiddleTable"/>
+        </set>
+	</class>
+    <class name="MiddleTable" table="MIDDLE">
+        <composite-id>
+            <key-many-to-one name="left" access="field">
+                <column name="LEFT_ID"></column>
+            </key-many-to-one>
+            <key-many-to-one name="right" access="field">
+                <column name="RIGHT_ID"></column>
+            </key-many-to-one>
+        </composite-id>
+    </class>
+    <class name="RightTable" table="RIGHT">
+        <id name="id" access="field">
+            <column name="ID"/>
+            <generator class="native" />
+        </id>
+        <property name="value" access="field"></property>
+    </class>
+</hibernate-mapping>
\ No newline at end of file
Index: testsuite/src/test/java/org/hibernate/test/criteria/LeftTable.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/criteria/LeftTable.java	(revision 0)
+++ testsuite/src/test/java/org/hibernate/test/criteria/LeftTable.java	(revision 0)
@@ -0,0 +1,11 @@
+package org.hibernate.test.criteria;
+
+import java.util.Set;
+
+public class LeftTable {
+
+	private String id;
+
+	private Set<MiddleTable> middleTables;
+
+}
Index: testsuite/src/test/java/org/hibernate/test/criteria/MiddleTable.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/criteria/MiddleTable.java	(revision 0)
+++ testsuite/src/test/java/org/hibernate/test/criteria/MiddleTable.java	(revision 0)
@@ -0,0 +1,11 @@
+package org.hibernate.test.criteria;
+
+import java.io.Serializable;
+
+public class MiddleTable implements Serializable {
+
+	private LeftTable left;
+
+	private RightTable right;
+
+}
Index: testsuite/src/test/java/org/hibernate/test/criteria/RightTable.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/criteria/RightTable.java	(revision 0)
+++ testsuite/src/test/java/org/hibernate/test/criteria/RightTable.java	(revision 0)
@@ -0,0 +1,9 @@
+package org.hibernate.test.criteria;
+
+public class RightTable {
+
+	private String id;
+
+	private String value;
+
+}
