Index: src/main/java/org/hibernate/dialect/DialectFactory.java
===================================================================
--- src/main/java/org/hibernate/dialect/DialectFactory.java	(revision 14832)
+++ src/main/java/org/hibernate/dialect/DialectFactory.java	(working copy)
@@ -130,6 +130,8 @@
 
 		MAPPERS.put( "Informix Dynamic Server", new VersionInsensitiveMapper( "org.hibernate.dialect.InformixDialect" ) );
 
+		MAPPERS.put( "OpenEdge RDBMS", new VersionInsensitiveMapper( "org.hibernate.dialect.ProgressDialect" ) );
+
 		// thanks goodness for "universal" databases...
 		MAPPERS.put( "DB2/NT", new VersionInsensitiveMapper( "org.hibernate.dialect.DB2Dialect" ) );
 		MAPPERS.put( "DB2/LINUX", new VersionInsensitiveMapper( "org.hibernate.dialect.DB2Dialect" ) );
