diff -ruN hibernate-2.1/test/org/hibernate/test/FooBarTest.java hibernate-2.1-mssql-locking/test/org/hibernate/test/FooBarTest.java
--- hibernate-2.1/test/org/hibernate/test/FooBarTest.java	Wed Feb  4 22:46:32 2004
+++ hibernate-2.1-mssql-locking/test/org/hibernate/test/FooBarTest.java	Sun Feb 29 18:53:50 2004
@@ -1598,7 +1598,8 @@
 			!( getDialect() instanceof DB2Dialect) && 
 			!( getDialect() instanceof HSQLDialect) &&
 			!( getDialect() instanceof SQLServerDialect) &&
-			!( getDialect() instanceof SybaseDialect)
+			!( getDialect() instanceof SybaseDialect) &&
+			!( getDialect() instanceof PostgreSQLDialect)
 		) { // oracle barfs on "x in ()" 
 			q.setParameterList("nameList", Collections.EMPTY_LIST);
 			list = q.list();
diff -ruN hibernate-2.1/test/org/hibernate/test/XY.hbm.xml hibernate-2.1-mssql-locking/test/org/hibernate/test/XY.hbm.xml
--- hibernate-2.1/test/org/hibernate/test/XY.hbm.xml	Wed Feb  4 22:46:32 2004
+++ hibernate-2.1-mssql-locking/test/org/hibernate/test/XY.hbm.xml	Sun Feb 29 18:54:37 2004
@@ -22,7 +22,7 @@
 		<property name="x"/>
 		<one-to-one name="theX" cascade="save-update"/>
 	</class>
-	<class name="org.hibernate.test.X$XX">
+	<class name="org.hibernate.test.X$XX" table="X_XX">
 		<id name="id">
 			<generator class="hilo"/>
 		</id>
