Index: src/net/sf/hibernate/cfg/Environment.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/cfg/Environment.java,v
retrieving revision 1.18.2.12
diff -u -r1.18.2.12 Environment.java
--- src/net/sf/hibernate/cfg/Environment.java   10 Sep 2003 12:56:56 -0000      1.18.2.12
+++ src/net/sf/hibernate/cfg/Environment.java   11 Sep 2003 09:19:38 -0000
@@ -268,7 +268,11 @@
         * Should we validate the connection on checkout. (optional)
         */
        public static final String C3P0_VALIDATE_CONNECTION = "hibernate.c3p0.validate";
-
+       /**
+        * On connection close, commit any pending transactions or rollback. (optional)
+        */
+       public static final String C3P0_AUTOCOMMIT = "hibernate.c3p0.autocommit";
+
        /**
         * Maximum number of checked out connections for DBCP connection pool
         */
Index: src/net/sf/hibernate/connection/C3P0ConnectionProvider.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/connection/C3P0ConnectionProvider.java,v
retrieving revision 1.7.2.2
diff -u -r1.7.2.2 C3P0ConnectionProvider.java
--- src/net/sf/hibernate/connection/C3P0ConnectionProvider.java 7 Sep 2003 00:59:51 -0000       1.7.2.2
+++ src/net/sf/hibernate/connection/C3P0ConnectionProvider.java 11 Sep 2003 09:20:24 -0000
@@ -82,6 +82,7 @@
                        int maxIdleTime = PropertiesHelper.getInt(Environment.C3P0_TIMEOUT, props, 0);
                        int maxStatements = PropertiesHelper.getInt(Environment.C3P0_MAX_STATEMENTS, props, 0);
                        boolean validateConnection = PropertiesHelper.getBoolean(Environment.C3P0_VALIDATE_CONNECTION, props);
+                       boolean autoCommit = PropertiesHelper.getBoolean(Environment.C3P0_AUTOCOMMIT, props);
 
                        PoolConfig pcfg = new PoolConfig();
                        pcfg.setMinPoolSize(minPoolSize);
@@ -90,6 +91,7 @@
                        pcfg.setMaxIdleTime(maxIdleTime);
                        pcfg.setMaxStatements(maxStatements);
                        pcfg.setTestConnectionOnCheckout(validateConnection);
+                       pcfg.setAutoCommitOnClose(autoCommit);
 
                        /*DataSource unpooled = DataSources.unpooledDataSource(
                                jdbcUrl, props.getProperty(Environment.USER), props.getProperty(Environment.PASS)
Index: src/hibernate.properties
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/hibernate.properties,v
retrieving revision 1.19.2.6
diff -u -r1.19.2.6 hibernate.properties
--- src/hibernate.properties    2 Sep 2003 10:54:12 -0000       1.19.2.6
+++ src/hibernate.properties    11 Sep 2003 09:37:31 -0000
@@ -174,6 +174,7 @@
 #hibernate.c3p0.timeout 5000
 #hibernate.c3p0.max_statements 100
 #hibernate.c3p0.validate false
+#hibernate.c3p0.autocommit false                                