--- org\hibernate\criterion\Order.java.orig	Thu Sep 29 19:11:39 2005
+++ org\hibernate\criterion\Order.java	Thu Sep 29 19:09:28 2005
@@ -47,7 +47,7 @@
 		StringBuffer fragment = new StringBuffer();
 		for ( int i=0; i<columns.length; i++ ) {
 			SessionFactoryImplementor factory = criteriaQuery.getFactory();
-			boolean lower = ignoreCase && type.sqlTypes( factory )[i]==Types.VARCHAR;
+			boolean lower = ignoreCase && type.sqlTypes( factory )[i]==Types.VARCHAR || type.sqlTypes( factory )[i]==Types.CHAR;
 			if (lower) {
 				fragment.append( factory.getDialect().getLowercaseFunction() )
 					.append('(');
