Index: AbstractEntityPersister.java
===================================================================
RCS file: /home/cvs/plumcreek/hibernate/src/net/sf/hibernate/persister/AbstractEntityPersister.java,v
retrieving revision 1.1
diff -u -r1.1 AbstractEntityPersister.java
--- AbstractEntityPersister.java	3 Oct 2003 19:28:54 -0000	1.1
+++ AbstractEntityPersister.java	11 Oct 2003 06:44:01 -0000
@@ -201,6 +201,13 @@
 			}  
 		}
 		catch (Throwable t) {
+			Class[] types = optimizer.getPropertyTypes();
+			String[] setters = optimizer.getSetters();
+			for (int i = 0; i < values.length; i++) {
+				if ((values[i] == null) && types[i].isPrimitive()) {
+					log.warn( "Setter " + setters[i] + " expects primitive " + types[i].getName() + " but value is null.");
+				}
+			}
 			throw new PropertyAccessException(t, "exception setting property value with CGLIB (set hibernate.cglib.use_reflection_optimizer=false for more info)", true, mappedClass, "?");
 		}		
