Index: src/main/java/org/hibernate/transaction/OpenEJBTransactionManagerLookup.java
===================================================================
--- src/main/java/org/hibernate/transaction/OpenEJBTransactionManagerLookup.java	(revision 0)
+++ src/main/java/org/hibernate/transaction/OpenEJBTransactionManagerLookup.java	(revision 0)
@@ -0,0 +1,31 @@
+package org.hibernate.transaction;
+
+import org.hibernate.HibernateException;
+
+import javax.transaction.TransactionManager;
+import java.util.Properties;
+
+/**
+ * A <tt>TransactionManager</tt> lookup strategy for OpenEJB
+ * @author David Blevins
+ */
+public class OpenEJBTransactionManagerLookup implements TransactionManagerLookup {
+
+    /**
+     * @see org.hibernate.transaction.TransactionManagerLookup#getTransactionManager(java.util.Properties)
+     */
+    public TransactionManager getTransactionManager(Properties props) throws HibernateException {
+        try {
+            Class clazz = Class.forName("org.apache.openejb.OpenEJB");
+            return (TransactionManager) clazz.getMethod("getTransactionManager", null).invoke(null, null);
+        }
+        catch (Exception e) {
+            throw new HibernateException( "Could not obtain OpenEJB transaction manager instance", e );
+        }
+    }
+
+    public String getUserTransactionName() {
+        return "java:comp/UserTransaction";
+    }
+
+}
