--- InstrumentTask.java	2006-07-24 12:14:52.000000000 -0400
+++ /disk2/src/Hibernate3/src/org/hibernate/tool/instrument/cglib/InstrumentTask.java	2006-07-31 16:27:57.000000000 -0400
@@ -1,4 +1,4 @@
-//$Id: InstrumentTask.java 9210 2006-02-03 22:15:19Z steveebersole $
+//$Id: InstrumentTask.java 10182 2006-07-31 20:27:50Z scottmarlownovell $
 package org.hibernate.tool.instrument.cglib;
 
 import java.util.Arrays;
@@ -43,21 +43,22 @@
 	 * Override the {@link AbstractTransformTask#getClassTransformer} method
 	 * in order to define field access interception transformation should occur.
 	 */
-	protected ClassTransformer getClassTransformer(String[] classInfo) {
+	protected ClassTransformer getClassTransformer(final String[] classInfo) {
 		if ( Arrays.asList( classInfo ).contains( InterceptFieldEnabled.class.getName() ) ) {
 			// The class is already instrumented, so skip this step
 			return null;
 		}
 		else {
+
 			// Class was not yet enhanced, so apply the transformation
 			return new InterceptFieldTransformer(
 					new InterceptFieldFilter() {
 						public boolean acceptRead(Type owner, String name) {
-							return true;
+							return (owner.getClassName().equals(classInfo[0]));
 						}
 
 						public boolean acceptWrite(Type owner, String name) {
-							return true;
+							return (owner.getClassName().equals(classInfo[0]));
 						}
 					}
 			);
