From bed9bf548792ad9a30670bc4ecba0e55163cbd40 Mon Sep 17 00:00:00 2001
From: Julien Bonjean <julien.bonjean@adenops.com>
Date: Tue, 25 Oct 2016 22:25:06 -0400
Subject: [PATCH] fix HHH-10938

---
 .../boot/archive/internal/JarFileBasedArchiveDescriptor.java          | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/hibernate-core/src/main/java/org/hibernate/boot/archive/internal/JarFileBasedArchiveDescriptor.java b/hibernate-core/src/main/java/org/hibernate/boot/archive/internal/JarFileBasedArchiveDescriptor.java
index eb1212f..18ce1b3 100644
--- a/hibernate-core/src/main/java/org/hibernate/boot/archive/internal/JarFileBasedArchiveDescriptor.java
+++ b/hibernate-core/src/main/java/org/hibernate/boot/archive/internal/JarFileBasedArchiveDescriptor.java
@@ -118,8 +118,8 @@ public class JarFileBasedArchiveDescriptor extends AbstractArchiveDescriptor {
 				final String name = extractName( zipEntry );
 				final String relativeName = extractRelativeName( zipEntry );
 				final InputStreamAccess inputStreamAccess;
-				try {
-					inputStreamAccess = buildByteBasedInputStreamAccess( name, jarFile.getInputStream( zipEntry ) );
+				try (InputStream is = jarFile.getInputStream( zipEntry )) {
+					inputStreamAccess = buildByteBasedInputStreamAccess( name, is );
 				}
 				catch (IOException e) {
 					throw new ArchiveException(
-- 
2.10.1

