Index: tools/src/java/org/hibernate/tool/hbm2x/pojo/BasicPOJOClass.java
===================================================================
--- tools/src/java/org/hibernate/tool/hbm2x/pojo/BasicPOJOClass.java	(revision 16240)
+++ tools/src/java/org/hibernate/tool/hbm2x/pojo/BasicPOJOClass.java	(working copy)
@@ -616,7 +616,7 @@
 
 	public String generateHashCode(Property property, String result, String thisName, boolean jdk5) {
 		StringBuffer buf = new StringBuffer();
-		if ( c2j.getMetaAsBool( property, "use-in-equals" ) ) {
+		if (usePropertyInEquals(property)) {
 			String javaTypeName = c2j.getJavaTypeName( property, jdk5, this );
 			boolean isPrimitive = c2j.isPrimitive( javaTypeName );
 			if ( isPrimitive ) {
Index: tools/src/templates/pojo/PojoEqualsHashcode.ftl
===================================================================
--- tools/src/templates/pojo/PojoEqualsHashcode.ftl	(revision 16240)
+++ tools/src/templates/pojo/PojoEqualsHashcode.ftl	(working copy)
@@ -1,16 +1,16 @@
-<#if pojo.needsEqualsHashCode() && !clazz.superclass?exists>   public boolean equals(Object other) {
+<#if pojo.needsEqualsHashCode() && !clazz.superclass?exists>	public boolean equals(Object other) {
          if ( (this == other ) ) return true;
 		 if ( (other == null ) ) return false;
 		 if ( !(other instanceof ${pojo.getDeclarationName()}) ) return false;
 		 ${pojo.getDeclarationName()} castOther = ( ${pojo.getDeclarationName()} ) other; 
          
 		 return ${pojo.generateEquals("this", "castOther", jdk5)};
-   }
+	}
    
-   public int hashCode() {
+	public int hashCode() {
          int result = 17;
          
-<#foreach property in pojo.getAllPropertiesIterator()>         ${pojo.generateHashCode(property, "result", "this", jdk5)}
+<#foreach property in pojo.getEqualsHashCodePropertiesIterator()>         ${pojo.generateHashCode(property, "result", "this", jdk5)}
 </#foreach>         return result;
-   }   
+	}   
 </#if>
\ No newline at end of file
