Index: documentation/src/test/java/org/hibernate/userguide/collections/ElementCollectionMapTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- documentation/src/test/java/org/hibernate/userguide/collections/ElementCollectionMapTest.java	(date 1521507992000)
+++ documentation/src/test/java/org/hibernate/userguide/collections/ElementCollectionMapTest.java	(revision )
@@ -76,7 +76,6 @@
 		@ElementCollection
 		@CollectionTable(name = "phone_register")
 		@Column(name = "since")
-		@MapKeyJoinColumn(name = "phone_id", referencedColumnName = "id")
 		private Map<Phone, Date> phoneRegister = new HashMap<>();
 
 		public Person() {}
