diff --git a/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/EntityModeConverter.java b/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/EntityModeConverter.java
index eeee468..70f9cef 100644
--- a/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/EntityModeConverter.java
+++ b/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/EntityModeConverter.java
@@ -17,6 +17,6 @@ public class EntityModeConverter {
 	}
 
 	public static String toXml(EntityMode entityMode) {
-		return entityMode.name();
+		return (null==entityMode) ? null : entityMode.getExternalName();
 	}
 }
