Index: hibernate-core/src/main/java/org/hibernate/engine/spi/QueryParameters.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- hibernate-core/src/main/java/org/hibernate/engine/spi/QueryParameters.java	(revision b1a322d1f389b991811b8bcfff8fee7f29760627)
+++ hibernate-core/src/main/java/org/hibernate/engine/spi/QueryParameters.java	(revision )
@@ -652,6 +652,7 @@
 		copy.processedSQL = this.processedSQL;
 		copy.processedPositionalParameterTypes = this.processedPositionalParameterTypes;
 		copy.processedPositionalParameterValues = this.processedPositionalParameterValues;
+		copy.passDistinctThrough = this.passDistinctThrough;
 		return copy;
 	}
 
