Index: src/org/hibernate/hibernate-mapping-3.0.dtd
===================================================================
--- src/org/hibernate/hibernate-mapping-3.0.dtd	(revision 11516)
+++ src/org/hibernate/hibernate-mapping-3.0.dtd	(working copy)
@@ -807,6 +807,7 @@
 <!ELEMENT list-index (column?)>
 	<!ATTLIST list-index column CDATA #IMPLIED>
 	<!ATTLIST list-index base CDATA "0">
+	<!ATTLIST list-index node CDATA #IMPLIED>
 
 <!ELEMENT map-key ((column|formula)*)>
 	<!ATTLIST map-key column CDATA #IMPLIED>
