Index: hibernate-core/src/main/java/org/hibernate/type/descriptor/java/ByteArrayTypeDescriptor.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- hibernate-core/src/main/java/org/hibernate/type/descriptor/java/ByteArrayTypeDescriptor.java	(revision 4780f117c52e2eca99561374d6a912b0caad5068)
+++ hibernate-core/src/main/java/org/hibernate/type/descriptor/java/ByteArrayTypeDescriptor.java	(revision )
@@ -27,6 +27,7 @@
 import java.io.InputStream;
 import java.sql.Blob;
 import java.sql.SQLException;
+import java.util.Arrays;
 
 import org.hibernate.HibernateException;
 import org.hibernate.engine.jdbc.BinaryStream;
@@ -45,7 +46,23 @@
 	public ByteArrayTypeDescriptor() {
 		super( Byte[].class, ArrayMutabilityPlan.INSTANCE );
 	}
+
+    @Override
+    public boolean areEqual(Byte[] one, Byte[] another) {
+        return one == another
+                || ( one != null && another != null && Arrays.equals(one, another) );
+    }
+
+    @Override
+    public int extractHashCode(Byte[] bytes) {
+        int hashCode = 1;
+        for ( byte aByte : bytes ) {
+            hashCode = 31 * hashCode + aByte;
+        }
+        return hashCode;
+    }
+
-	@Override
+    @Override
 	public String toString(Byte[] bytes) {
 		final StringBuilder buf = new StringBuilder();
 		for ( Byte aByte : bytes ) {
Index: hibernate-core/src/test/java/org/hibernate/test/annotations/lob/VersionedLobTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- hibernate-core/src/test/java/org/hibernate/test/annotations/lob/VersionedLobTest.java	(revision 4780f117c52e2eca99561374d6a912b0caad5068)
+++ hibernate-core/src/test/java/org/hibernate/test/annotations/lob/VersionedLobTest.java	(revision )
@@ -28,7 +28,6 @@
 import org.hibernate.Session;
 import org.hibernate.Transaction;
 import org.hibernate.testing.DialectChecks;
-import org.hibernate.testing.FailureExpected;
 import org.hibernate.testing.RequiresDialectFeature;
 
 import static org.junit.Assert.assertEquals;
@@ -146,7 +145,6 @@
 	}
 
 	@Test
-	@FailureExpected( jiraKey = "HHH-5811")
 	public void testVersionUnchangedByteArray() throws Exception {
 		Session s;
 		Transaction tx;
