--- hibernate-3.2.5.ga/test/org/hibernate/test/sql/hand/query/NativeSQLQueriesTest.java	2007-06-07 15:48:34.000000000 -0700
+++ NativeSQLQueriesTest.java	2007-09-07 14:43:46.593750000 -0700
@@ -160,11 +160,20 @@
 		assertEquals( 2, l.size() );
 
 		l = s.createSQLQuery( getOrgEmpPersonSQL() )
+		.addEntity("org", Organization.class)
+		.addEntity("emp", Employment.class)
+		.addEntity("pers", Person.class)
+		.list();
+		assertEquals( l.size(), 1 );
+		assertTrue(l.get(0) instanceof Object[]);
+
+		l = s.createSQLQuery( getOrgEmpPersonSQL() )
 				.addEntity("org", Organization.class)
 				.addJoin("emp", "org.employments")
 				.addJoin("pers", "emp.employee")
 				.list();
 		assertEquals( l.size(), 1 );
+		assertTrue(l.get(0) instanceof Organization);
 
 		t.commit();
 		s.close();
