Index: src/org/hibernate/persister/entity/AbstractEntityPersister.java
===================================================================
--- src/org/hibernate/persister/entity/AbstractEntityPersister.java	(revision 11393)
+++ src/org/hibernate/persister/entity/AbstractEntityPersister.java	(working copy)
@@ -300,9 +300,11 @@
 		if ( sqlUpdateByRowIdString == null ) {
 			throw new AssertionFailure( "no update by row id" );
 		}
-		String[] result = new String[getTableSpan() + 1];
+		String[] result = new String[getTableSpan()];
 		result[0] = sqlUpdateByRowIdString;
-		System.arraycopy( sqlUpdateStrings, 0, result, 1, getTableSpan() );
+		if ( getTableSpan() > 1 ) {
+			System.arraycopy( sqlUpdateStrings, 1, result, 1, getTableSpan() - 1 );
+		}
 		return result;
 	}
 
@@ -2001,6 +2003,7 @@
 		}
 
 		if ( rowId != null ) {
+			//log.trace( "binding '" + ((oracle.sql.ROWID)rowId).stringValue() + "' to parameter: " + index );
 			ps.setObject( index, rowId );
 			index += 1;
 		}
