--- /home/cgs/dist/hibernate-distribution-3.6.0.Final/project/core/src/main/java/org/hibernate/engine/jdbc/ClobProxy.java	2010-10-13 15:43:09.000000000 -0500
+++ org/hibernate/engine/jdbc/ClobProxy.java	2010-11-02 01:31:20.000000000 -0500
@@ -142,7 +142,7 @@
 			if ( start < 1 ) {
 				throw new SQLException( "Start position 1-based; must be 1 or more." );
 			}
-			if ( start > getLength() ) {
+			if ( start - 1 > getLength() ) {
 				throw new SQLException( "Start position [" + start + "] cannot exceed overall CLOB length [" + getLength() + "]" );
 			}
 			int length = (Integer) args[1];
