Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/AbstractDelegateSessionImplementor.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/AbstractDelegateSessionImplementor.java	(revision 15803)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/AbstractDelegateSessionImplementor.java	(working copy)
@@ -40,7 +40,7 @@
 import org.hibernate.Transaction;
 import org.hibernate.collection.PersistentCollection;
 import org.hibernate.engine.EntityKey;
-import org.hibernate.engine.LoadQueryInfluencers;
+//import org.hibernate.engine.LoadQueryInfluencers;
 import org.hibernate.engine.PersistenceContext;
 import org.hibernate.engine.QueryParameters;
 import org.hibernate.engine.SessionFactoryImplementor;
@@ -72,9 +72,9 @@
 
     // Delegate methods
 
-	public LoadQueryInfluencers getLoadQueryInfluencers() {
-		return delegate.getLoadQueryInfluencers();
-	}
+//	public LoadQueryInfluencers getLoadQueryInfluencers() {
+//		return delegate.getLoadQueryInfluencers();
+//	}
 
 	public Interceptor getInterceptor() {
         return delegate.getInterceptor();
