Index: src/org/hibernate/eclipse/console/views/properties/HibernatePropertySourceProvider.java
===================================================================
RCS file: /cvsroot/jboss/jbosside/hibernatetools/plugins/org.hibernate.eclipse.console/src/org/hibernate/eclipse/console/views/properties/HibernatePropertySourceProvider.java,v
retrieving revision 1.7
diff -u -r1.7 HibernatePropertySourceProvider.java
--- src/org/hibernate/eclipse/console/views/properties/HibernatePropertySourceProvider.java	27 Apr 2006 22:54:52 -0000	1.7
+++ src/org/hibernate/eclipse/console/views/properties/HibernatePropertySourceProvider.java	7 Jul 2006 06:03:26 -0000
@@ -22,7 +22,10 @@
 	}
 
 	public IPropertySource getPropertySource(Object object) {
-		if (object instanceof QueryPage)
+		if(  object == null  ) {
+			return null;
+		}
+		else if (object instanceof QueryPage)
 		{
 			return new QueryPagePropertySource( (QueryPage)object);
 		}
