Index: core/src/main/java/org/hibernate/dialect/FirebirdDialect.java
===================================================================
--- core/src/main/java/org/hibernate/dialect/FirebirdDialect.java	(revision 20099)
+++ core/src/main/java/org/hibernate/dialect/FirebirdDialect.java	(working copy)
@@ -35,9 +35,21 @@
 	}
 
 	public String getLimitString(String sql, boolean hasOffset) {
+		int afterSelectIndex = 6; /* default: SQL starts with "select", after which we insert "first" and optional "skip" */
+		if (sql.startsWith("/*")) { /* except if comments are enabled */
+			afterSelectIndex = sql.indexOf("*/") + 2; /* then we have to find the "select" */
+			if (afterSelectIndex < 2) {
+				throw new IllegalArgumentException("SQL starts with a non-terminated comment");
+			}
+			final int maxIndex = sql.length() - 6; /* minus the length of "select" */
+			while ((afterSelectIndex < maxIndex) && Character.isWhitespace(sql.charAt(afterSelectIndex))) {
+				afterSelectIndex++; /* skip whitespace in front of "select" */
+			}
+			afterSelectIndex += 6; /* now move over the "select" */
+		}
 		return new StringBuffer( sql.length()+20 )
 			.append(sql)
-			.insert(6, hasOffset ? " first ? skip ?" : " first ?")
+			.insert(afterSelectIndex, hasOffset ? " first ? skip ?" : " first ?")
 			.toString();
 	}
 
@@ -49,4 +61,4 @@
 		return true;
 	}
 
-}
\ No newline at end of file
+}
