--- src/net/sf/hibernate/tool/hbm2java/ClassMapping.java.orig	Wed Mar 10 17:40:56 2004
+++ src/net/sf/hibernate/tool/hbm2java/ClassMapping.java	Wed Mar 10 17:18:15 2004
@@ -268,6 +268,7 @@
       doCollections(classElement, "map", "java.util.Map", "java.util.HashMap", metaattribs);
       doCollections(classElement, "set", "java.util.Set", "java.util.HashSet", metaattribs);
       doCollections(classElement, "bag", System.getProperty("hbm2java.bag.interface","java.util.List"), "java.util.ArrayList", metaattribs);
+      doCollections(classElement, "idbag", System.getProperty("hbm2java.bag.interface","java.util.List"), "java.util.ArrayList", metaattribs);
       doArrays(classElement, "array", metaattribs);
       doArrays(classElement, "primitive-array", metaattribs);
     
