Index: src/main/java/org/hibernate/transform/AliasToEntityMapResultTransformer.java
===================================================================
--- src/main/java/org/hibernate/transform/AliasToEntityMapResultTransformer.java	(revision 14969)
+++ src/main/java/org/hibernate/transform/AliasToEntityMapResultTransformer.java	(working copy)
@@ -6,16 +6,19 @@
 import java.util.Map;
 
 /**
+ * Generates a list of Maps from the result with the key as the column alias -
+ * one map per result row.
+ * 
  * @author Gavin King
  */
 public class AliasToEntityMapResultTransformer implements ResultTransformer {
 
 	public Object transformTuple(Object[] tuple, String[] aliases) {
 		Map result = new HashMap(tuple.length);
-		for ( int i=0; i<tuple.length; i++ ) {
+		for (int i = 0; i < tuple.length; i++) {
 			String alias = aliases[i];
-			if ( alias!=null ) {
-				result.put( alias, tuple[i] );
+			if (alias != null) {
+				result.put(alias, tuple[i]);
 			}
 		}
 		return result;
