Index: core/src/main/java/org/hibernate/sql/ordering/antlr/OrderByFragmentTranslator.java
===================================================================
--- core/src/main/java/org/hibernate/sql/ordering/antlr/OrderByFragmentTranslator.java	(revision 19152)
+++ core/src/main/java/org/hibernate/sql/ordering/antlr/OrderByFragmentTranslator.java	(working copy)
@@ -71,7 +71,7 @@
 			log.trace( printer.showAsString( parser.getAST(), "--- {order-by fragment} ---" ) );
 		}
 
-		GeneratedOrderByFragmentRenderer renderer = new GeneratedOrderByFragmentRenderer();
+		OrderByFragmentRenderer renderer = new OrderByFragmentRenderer();
 		try {
 			renderer.orderByFragment( parser.getAST() );
 		}
