Index: tools/src/java/net/sf/hibernate/tool/hbm2java/AbstractRenderer.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/tools/src/java/net/sf/hibernate/tool/hbm2java/AbstractRenderer.java,v
retrieving revision 1.6
diff -u -u -r1.6 AbstractRenderer.java
--- tools/src/java/net/sf/hibernate/tool/hbm2java/AbstractRenderer.java	4 Dec 2003 21:42:17 -0000	1.6
+++ tools/src/java/net/sf/hibernate/tool/hbm2java/AbstractRenderer.java	26 Mar 2004 14:39:11 -0000
@@ -7,6 +7,7 @@
 package net.sf.hibernate.tool.hbm2java;
 
 import java.io.PrintWriter;
+import java.util.Map;
 import java.util.Properties;
 
 /**
@@ -41,4 +42,25 @@
             w.println("// default package");
         }
     }
+
+    public String getSaveToClassName(ClassMapping classMapping) {
+        if (properties.containsKey("useGeneratedClassName"))
+        {
+            String val = properties.getProperty("useGeneratedClassName");
+            if ("false".equals(val))
+                return classMapping.getName();
+        }
+        return classMapping.getGeneratedName();
+    }
+
+    public String getSaveToPackage(ClassMapping classMapping) {
+        if (properties.containsKey("useGeneratedClassName"))
+        {
+            String val = properties.getProperty("useGeneratedPackageName");
+            if ("false".equals(val))
+                return classMapping.getPackageName();
+        }
+        return classMapping.getGeneratedPackageName();
+    }
+
 }
Index: tools/src/java/net/sf/hibernate/tool/hbm2java/Generator.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/tools/src/java/net/sf/hibernate/tool/hbm2java/Generator.java,v
retrieving revision 1.2
diff -u -u -r1.2 Generator.java
--- tools/src/java/net/sf/hibernate/tool/hbm2java/Generator.java	4 Dec 2003 21:42:17 -0000	1.2
+++ tools/src/java/net/sf/hibernate/tool/hbm2java/Generator.java	26 Mar 2004 14:39:11 -0000
@@ -27,6 +27,7 @@
      private String baseDirName = "generated";
      private String packageName = null;
      private String suffix = StringHelper.EMPTY_STRING;
+     private String prefix = StringHelper.EMPTY_STRING;
      private String extension = "java";
      private boolean lowerFirstLetter = false;
 
@@ -56,6 +57,11 @@
          // set packageName field
          this.packageName = generateElement.getAttributeValue("package");
          
+         // set prefix
+         if ((value = generateElement.getAttributeValue("prefix")) != null) {
+             this.prefix = value;
+         }
+
          // set suffix
          if ((value = generateElement.getAttributeValue("suffix")) != null) {
              this.suffix = value;
@@ -121,8 +127,8 @@
      *
      */
     private void write(ClassMapping classMapping, Map class2classmap, Renderer renderer) throws Exception {
-        File dir = this.getDir(classMapping.getGeneratedPackageName());
-        File file = new File(dir, this.getFileName(classMapping.getGeneratedName()));
+        File dir = this.getDir(renderer.getSaveToPackage(classMapping));
+        File file = new File(dir, this.getFileName(renderer.getSaveToClassName(classMapping)));
         log.debug("Writing " + file);
 
         PrintWriter writer = new PrintWriter(new FileOutputStream(file));
@@ -150,7 +156,7 @@
             name = className;
         }
 
-        return name + this.suffix;
+        return this.prefix+ name + this.suffix;
     }
 
     /**
Index: tools/src/java/net/sf/hibernate/tool/hbm2java/Renderer.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/tools/src/java/net/sf/hibernate/tool/hbm2java/Renderer.java,v
retrieving revision 1.2
diff -u -u -r1.2 Renderer.java
--- tools/src/java/net/sf/hibernate/tool/hbm2java/Renderer.java	7 Dec 2003 19:41:34 -0000	1.2
+++ tools/src/java/net/sf/hibernate/tool/hbm2java/Renderer.java	26 Mar 2004 14:39:11 -0000
@@ -22,6 +22,22 @@
      * @throws Exception
      */    
 	public void render(String savedToPackage, String savedToClass, ClassMapping classMapping, Map class2classmap, PrintWriter writer) throws Exception;
+    
+    /**
+     * Called by the generator to determine the package name of the rendered class.
+     * 
+     * @param classMapping The class mapping of the generated class
+     * @return the package name the class should be saved to
+     */
+    public String getSaveToPackage(ClassMapping classMapping);
+    
+    /**
+     * Called by the generator to determine the class name of the rendered class.
+     * 
+     * @param classMapping The class mapping of the generated class
+     * @return the class name the class should be saved to
+     */
+    public String getSaveToClassName(ClassMapping classMapping);
 }
 
 
