Index: hibernate-core/src/main/java/org/hibernate/type/EntityType.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- hibernate-core/src/main/java/org/hibernate/type/EntityType.java	(revision 58405)
+++ hibernate-core/src/main/java/org/hibernate/type/EntityType.java	(revision )
@@ -504,7 +504,13 @@
 				id = ( Serializable ) value;
 			}
 			else {
+                if (value instanceof HibernateProxy) {
+              			id = ( (HibernateProxy) value ).getHibernateLazyInitializer()
+              					.getIdentifier();
+              		}
+              		else {
-				id = persister.getIdentifier( value );
+                        id = persister.getIdentifier( value );
+              		}
 			}
 			
 			result.append( '#' )
