Index: IdentifierGeneratorFactory.java
===================================================================
--- IdentifierGeneratorFactory.java	(revision 14012)
+++ IdentifierGeneratorFactory.java	(working copy)
@@ -95,7 +95,7 @@
 			return idgen;
 		}
 		catch (Exception e) {
-			throw new MappingException("could not instantiate id generator", e);
+			throw new MappingException("could not instantiate id generator for entity \'" + params.get("entity_name") + "\'", e);
 		}
 	}
 
