Index: hibernate-entitymanager/src/test/java/org/hibernate/jpa/test/cascade/DetachAndContainsTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- hibernate-entitymanager/src/test/java/org/hibernate/jpa/test/cascade/DetachAndContainsTest.java	(revision ac2f06800ea45aed66475a042a6f62cc99783565)
+++ hibernate-entitymanager/src/test/java/org/hibernate/jpa/test/cascade/DetachAndContainsTest.java	(revision )
@@ -8,13 +8,7 @@
 
 import java.util.ArrayList;
 import java.util.Collection;
-import javax.persistence.Entity;
-import javax.persistence.EntityManager;
-import javax.persistence.GeneratedValue;
-import javax.persistence.Id;
-import javax.persistence.ManyToOne;
-import javax.persistence.OneToMany;
-import javax.persistence.Table;
+import javax.persistence.*;
 
 import org.junit.Test;
 
@@ -64,6 +58,32 @@
 		em.close();
 	}
 
+	@Test
+	public void testDetachCollectionElementAndFlush(){
+		EntityManager em = getOrCreateEntityManager();
+		em.getTransaction().begin();
+
+		Tooth tooth = new Tooth();
+		Mouth mouth = new Mouth();
+		em.persist( mouth );
+		em.persist(tooth);
+		mouth.teeth = new ArrayList<Tooth>();
+		em.getTransaction().commit();
+		em.close();
+
+		em = getOrCreateEntityManager();
+		em.getTransaction().begin();
+
+		Mouth mouth1 = em.find(Mouth.class, mouth.id);
+		mouth1.teeth.add(tooth);
+		tooth.mouth = mouth1;
+		em.flush();
+		em.merge(mouth1);
+
+		em.getTransaction().commit();
+		em.close();
+	}
+
 	@Override
 	public Class[] getAnnotatedClasses() {
 		return new Class[] {
@@ -78,7 +98,7 @@
 		@Id
 		@GeneratedValue
 		public Integer id;
-		@OneToMany(mappedBy = "mouth", cascade = { DETACH, REMOVE } )
+		@OneToMany(mappedBy = "mouth", cascade = {CascadeType.ALL} )
 		public Collection<Tooth> teeth;
 	}
 
