Index: src/org/hibernate/engine/loading/LoadContexts.java
===================================================================
--- src/org/hibernate/engine/loading/LoadContexts.java	(revision 11404)
+++ src/org/hibernate/engine/loading/LoadContexts.java	(working copy)
@@ -134,10 +134,10 @@
 		LoadingCollectionEntry rtn = ( LoadingCollectionEntry ) xrefLoadingCollectionEntries.get( key );
 		if ( log.isTraceEnabled() ) {
 			if ( rtn == null ) {
-				log.trace( "collection [" + key + "] located in load context" );
+				log.trace( "collection [" + key + "] not located in load context" );
 			}
 			else {
-				log.trace( "collection [" + key + "] not located in load context" );
+				log.trace( "collection [" + key + "] located in load context" );
 			}
 		}
 		return rtn;
