diff -ru8pN core-orig/src/main/java/org/hibernate/cfg/SettingsFactory.java core/src/main/java/org/hibernate/cfg/SettingsFactory.java
--- core-orig/src/main/java/org/hibernate/cfg/SettingsFactory.java	2008-10-13 14:14:56.000000000 +0900
+++ core/src/main/java/org/hibernate/cfg/SettingsFactory.java	2008-10-13 16:07:52.000000000 +0900
@@ -142,16 +142,22 @@ public class SettingsFactory implements 
 				log.warn("Could not obtain connection metadata", sqle);
 			}
 			catch (UnsupportedOperationException uoe) {
 				// user supplied JDBC connections
 			}
 		}
 		settings.setDataDefinitionImplicitCommit( metaReportsDDLCausesTxnCommit );
 		settings.setDataDefinitionInTransactionSupported( metaReportsDDLInTxnSupported );
+		
+		// If the SQL dialect was not determined at this point, try with only props
+		if (dialect == null) {
+			dialect = determineDialectWithoutConnection(props);
+		}
+		
 		settings.setDialect( dialect );
 		
 		//use dialect default properties
 		final Properties properties = new Properties();
 		properties.putAll( dialect.getDefaultProperties() );
 		properties.putAll(props);
 		
 		// Transaction settings:
@@ -449,9 +455,13 @@ public class SettingsFactory implements 
 	protected TransactionManagerLookup createTransactionManagerLookup(Properties properties) {
 		return TransactionManagerLookupFactory.getTransactionManagerLookup(properties);		
 	}
 
 	private Dialect determineDialect(Properties props, Connection conn) {
 		return DialectFactory.buildDialect( props, conn );
 	}
 	
+	private Dialect determineDialectWithoutConnection(Properties props) {
+		return DialectFactory.buildDialect( props, null );
+	}
+	
 }
