Index: src/net/sf/hibernate/jmx/HibernateServiceMBean.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/jmx/HibernateServiceMBean.java,v
retrieving revision 1.9
diff -u -r1.9 HibernateServiceMBean.java
--- src/net/sf/hibernate/jmx/HibernateServiceMBean.java	4 Jun 2004 01:27:41 -0000	1.9
+++ src/net/sf/hibernate/jmx/HibernateServiceMBean.java	21 Jul 2004 10:16:42 -0000
@@ -224,6 +224,18 @@
 	 * @param showSql
 	 */
 	public void setShowSql(boolean showSql);
+
+        /**
+         * What is the automatic application of hbm?
+         * @return string
+         */
+        public String getHbmToDdlAuto();
+
+        /**
+         * Set automatic application of hbm
+         * @param hbmToDdlAuto
+         */
+        public void setHbmToDdlAuto(String hbmToDdlAuto);
 	
 	/**
 	 * Create the <tt>SessionFactory</tt> and bind to the jndi name on startup
