Index: src/net/sf/hibernate/jmx/HibernateService.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/jmx/HibernateService.java,v
retrieving revision 1.13
diff -u -r1.13 HibernateService.java
--- src/net/sf/hibernate/jmx/HibernateService.java	4 Jun 2004 01:27:41 -0000	1.13
+++ src/net/sf/hibernate/jmx/HibernateService.java	21 Jul 2004 10:16:35 -0000
@@ -118,6 +118,14 @@
 	public void setShowSql(boolean showSql) {
 		setProperty(Environment.SHOW_SQL, showSql);
 	}
+
+        public String getHbmToDdlAuto() {
+                return getProperty(Environment.HBM2DDL_AUTO);
+        }
+
+        public void setHbmToDdlAuto(String hbmToDdlAuto) {
+                setProperty(Environment.HBM2DDL_AUTO, hbmToDdlAuto);
+        }
 	
 	public String getUserName() {
 		return getProperty(Environment.USER);
