Index: D:/dev/workspace/hibernate3/src/org/hibernate/mapping/SimpleValue.java
===================================================================
--- D:/dev/workspace/hibernate3/src/org/hibernate/mapping/SimpleValue.java	(revision 11406)
+++ D:/dev/workspace/hibernate3/src/org/hibernate/mapping/SimpleValue.java	(working copy)
@@ -260,6 +260,9 @@
 		Type result = TypeFactory.heuristicType(typeName, typeParameters);
 		if (result==null) {
 			String msg = "Could not determine type for: " + typeName;
+			if(table != null){
+				msg += ", at table: " + table.getName();
+			} 
 			if(columns!=null && columns.size()>0) {
 				msg += ", for columns: " + columns;
 			}
