Index: src/main/java/org/hibernate/engine/profile/FetchProfile.java
===================================================================
--- src/main/java/org/hibernate/engine/profile/FetchProfile.java	(revision 19461)
+++ src/main/java/org/hibernate/engine/profile/FetchProfile.java	(working copy)
@@ -49,9 +49,7 @@
 	private Map fetches = new HashMap();
 
 	private boolean containsJoinFetchedCollection = false;
-	private boolean containsJoinFetchedBag = false;
-	private Fetch bagJoinFetch;
-
+	
 	/**
 	 * A 'fetch profile' is uniquely named within a
 	 * {@link SessionFactoryImplementor SessionFactory}, thus it is also
@@ -106,18 +104,6 @@
 					}
 				}
 
-				// also, in cases where we are asked to add a collection join
-				// fetch where we had already added a bag join fetch previously,
-				// we need to go back and ignore that previous bag join fetch.
-				if ( containsJoinFetchedBag ) {
-					if ( fetches.remove( bagJoinFetch.getAssociation().getRole() ) != bagJoinFetch ) {
-						// just for safety...
-						log.warn( "Unable to erase previously added bag join fetch" );
-					}
-					bagJoinFetch = null;
-					containsJoinFetchedBag = false;
-				}
-
 				containsJoinFetchedCollection = true;
 			}
 		}
@@ -157,13 +143,5 @@
 		return containsJoinFetchedCollection;
 	}
 
-	/**
-	 * Getter for property 'containsJoinFetchedBag', which flags whether this
-	 * fetch profile contained any bag join fetches
-	 *
-	 * @return Value for property 'containsJoinFetchedBag'.
-	 */
-	public boolean isContainsJoinFetchedBag() {
-		return containsJoinFetchedBag;
-	}
+	
 }
