Index: src/main/java/org/hibernate/engine/profile/FetchProfile.java
===================================================================
--- src/main/java/org/hibernate/engine/profile/FetchProfile.java	(revision 19461)
+++ src/main/java/org/hibernate/engine/profile/FetchProfile.java	(working copy)
@@ -109,7 +109,7 @@
 				// also, in cases where we are asked to add a collection join
 				// fetch where we had already added a bag join fetch previously,
 				// we need to go back and ignore that previous bag join fetch.
-				if ( containsJoinFetchedBag ) {
+				if ( containsJoinFetchedBag && bagJoinFetch != null) {
 					if ( fetches.remove( bagJoinFetch.getAssociation().getRole() ) != bagJoinFetch ) {
 						// just for safety...
 						log.warn( "Unable to erase previously added bag join fetch" );
