--- Hibernate2/src/net/sf/hibernate/tool/hbm2ddl/SchemaExport.java	2004-10-11 10:59:28.000000000 -0700
+++ Hibernate2.modified/src/net/sf/hibernate/tool/hbm2ddl/SchemaExport.java	2004-06-03 18:28:51.000000000 -0700
@@ -1,7 +1,6 @@
 //$Id: SchemaExport.java,v 1.15 2004/06/04 01:28:51 steveebersole Exp $
 package net.sf.hibernate.tool.hbm2ddl;
 
-import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileWriter;
 import java.io.IOException;
@@ -11,6 +10,9 @@
 import java.util.Properties;
 import java.util.StringTokenizer;
 
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
 import net.sf.hibernate.HibernateException;
 import net.sf.hibernate.cfg.Configuration;
 import net.sf.hibernate.cfg.NamingStrategy;
@@ -21,9 +23,6 @@
 import net.sf.hibernate.util.ReflectHelper;
 import net.sf.hibernate.util.StringHelper;
 
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
-
 /**
  * Commandline tool to export table schema to the database. This
  * class may also be called from inside an application.
@@ -273,8 +272,7 @@
 						delim = args[i].substring(12);
 					}
 					else if ( args[i].startsWith("--config=") ) {
-					   File file = new File(args[i].substring(9));
-						cfg.configure( file );
+						cfg.configure( args[i].substring(9) );
 					}
 					else if ( args[i].startsWith("--naming=") ) {
 						cfg.setNamingStrategy( 
