Index: tools/src/java/org/hibernate/tool/hbm2x/pojo/ImportContextImpl.java
===================================================================
--- tools/src/java/org/hibernate/tool/hbm2x/pojo/ImportContextImpl.java	(revision 14267)
+++ tools/src/java/org/hibernate/tool/hbm2x/pojo/ImportContextImpl.java	(working copy)
@@ -136,9 +136,9 @@
 					// dont add automatically "imported" stuff
 				} else {
 					if(staticImports.contains(next)) {
-						buf.append("import static " + next + ";\r\n");
+						buf.append("import static " + next + ";\n");
 					} else {
-						buf.append("import " + next + ";\r\n");
+						buf.append("import " + next + ";\n");
 					}
 				}
 		}
