Index: D:/dev/workspace/hibernate3/src/org/hibernate/criterion/Expression.java
===================================================================
--- D:/dev/workspace/hibernate3/src/org/hibernate/criterion/Expression.java	(revision 11406)
+++ D:/dev/workspace/hibernate3/src/org/hibernate/criterion/Expression.java	(working copy)
@@ -22,6 +22,7 @@
 	 * parameters. Any occurrences of <tt>{alias}</tt> will be
 	 * replaced by the table alias.
 	 *
+	 * @deprecated use {@link org.hibernate.criterion.Restrictions#sqlRestriction(String, Object[], Type[])}
 	 * @param sql
 	 * @param values
 	 * @param types
@@ -35,6 +36,7 @@
 	 * parameter. Any occurrences of <tt>{alias}</tt> will be replaced
 	 * by the table alias.
 	 *
+	 * @deprecated use {@link org.hibernate.criterion.Restrictions#sqlRestriction(String, Object, Type)}
 	 * @param sql
 	 * @param value
 	 * @param type
@@ -47,6 +49,7 @@
 	 * Apply a constraint expressed in SQL. Any occurrences of <tt>{alias}</tt>
 	 * will be replaced by the table alias.
 	 *
+	 * @deprecated use {@link org.hibernate.criterion.Restrictions#sqlRestriction(String)}
 	 * @param sql
 	 * @return Criterion
 	 */
