Index: MySQL5InnoDBDialect.java
===================================================================
--- MySQL5InnoDBDialect.java	(revision 0)
+++ MySQL5InnoDBDialect.java	(revision 0)
@@ -0,0 +1,21 @@
+//$Id:  $
+package org.hibernate.dialect;
+
+/**
+ * @author Scott Marlow
+ */
+public class MySQL5InnoDBDialect extends MySQL5Dialect {
+
+	public boolean supportsCascadeDelete() {
+		return true;
+	}
+	
+	public String getTableTypeString() {
+		return " ENGINE=InnoDB";
+	}
+
+	public boolean hasSelfReferentialForeignKeyBug() {
+		return true;
+	}
+	
+}

Property changes on: MySQL5InnoDBDialect.java
___________________________________________________________________
Name: svn:executable
   + *

