Index: /javadev/prj/Hibernate3/src/org/hibernate/type/CharacterArrayType.java
===================================================================
--- /javadev/prj/Hibernate3/src/org/hibernate/type/CharacterArrayType.java	(revision 11516)
+++ /javadev/prj/Hibernate3/src/org/hibernate/type/CharacterArrayType.java	(working copy)
@@ -1,6 +1,8 @@
 //$Id: $
 package org.hibernate.type;
 
+import java.util.Arrays;
+
 import org.hibernate.HibernateException;
 
 /**
@@ -34,4 +36,8 @@
 	}
 
 	public String getName() { return "wrapper-characters"; }
+	
+	public boolean isEqual(Object x, Object y) {
+		return Arrays.equals((Character[])x, (Character[])y);
+	}
 }
Index: /javadev/prj/Hibernate3/src/org/hibernate/type/CharArrayType.java
===================================================================
--- /javadev/prj/Hibernate3/src/org/hibernate/type/CharArrayType.java	(revision 11516)
+++ /javadev/prj/Hibernate3/src/org/hibernate/type/CharArrayType.java	(working copy)
@@ -1,6 +1,8 @@
 //$Id: $
 package org.hibernate.type;
 
+import java.util.Arrays;
+
 /**
  * put char[] into VARCHAR
  * @author Emmanuel Bernard
@@ -20,4 +22,8 @@
 	}
 
 	public String getName() { return "characters"; }
+	
+	public boolean isEqual(Object x, Object y) {
+		return Arrays.equals((char[])x, (char[])y);
+	}
 }

