Index: src/main/java/org/hibernate/mapping/Table.java
===================================================================
--- src/main/java/org/hibernate/mapping/Table.java	(revision 17527)
+++ src/main/java/org/hibernate/mapping/Table.java	(working copy)
@@ -322,14 +322,15 @@
 				if ( defaultValue != null ) {
 					alter.append( " default " ).append( defaultValue );
 
-					if ( column.isNullable() ) {
-						alter.append( dialect.getNullColumnString() );
-					}
-					else {
-						alter.append( " not null" );
-					}
+				
 
 				}
+				if ( column.isNullable() ) {
+					alter.append( dialect.getNullColumnString() );
+				}
+				else {
+					alter.append( " not null" );
+				}
 
 				boolean useUniqueConstraint = column.isUnique() &&
 						dialect.supportsUnique() &&
