Index: SubqueryExpression.java
===================================================================
--- SubqueryExpression.java	(revision 14324)
+++ SubqueryExpression.java	(working copy)
@@ -9,6 +9,7 @@
 import org.hibernate.engine.QueryParameters;
 import org.hibernate.engine.SessionFactoryImplementor;
 import org.hibernate.engine.TypedValue;
+import org.hibernate.engine.SessionImplementor;
 import org.hibernate.impl.CriteriaImpl;
 import org.hibernate.loader.criteria.CriteriaJoinWalker;
 import org.hibernate.loader.criteria.CriteriaQueryTranslator;
@@ -46,6 +47,8 @@
 		final OuterJoinLoadable persister = (OuterJoinLoadable) factory.getEntityPersister( criteriaImpl.getEntityOrClassName() );
 
 		createAndSetInnerQuery( criteriaQuery, factory );
+                
+		criteriaImpl.setSession(((CriteriaImpl) criteria).getSession());
 		
 		CriteriaJoinWalker walker = new CriteriaJoinWalker(
 				persister,
@@ -53,7 +56,7 @@
 				factory,
 				criteriaImpl,
 				criteriaImpl.getEntityOrClassName(),
-				new HashMap(),
+                ((CriteriaImpl) criteria).getSession().getEnabledFilters(),
 				innerQuery.getRootSQLALias());
 
 		String sql = walker.getSQLString();
