Index: src/org/hibernate/hql/ast/HqlSqlWalker.java
===================================================================
--- src/org/hibernate/hql/ast/HqlSqlWalker.java	(revision 15548)
+++ src/org/hibernate/hql/ast/HqlSqlWalker.java	(working copy)
@@ -343,7 +343,10 @@
 			SqlGenerator sql = new SqlGenerator( getSessionFactoryHelper().getFactory() );
 			sql.whereExpr( hqlSqlWithNode.getFirstChild() );
 			fromElement.setWithClauseFragment( visitor.getJoinAlias(), "(" + sql.getSQL() + ")" );
-
+			List params = sql.getCollectedParameters();
+			for (Iterator it = params.iterator(); it.hasNext(); ) {
+				fromElement.addEmbeddedParameter((ParameterSpecification) it.next());
+			}
 		}
 		catch( SemanticException e ) {
 			throw e;
