Index: PostgreSQLDialect.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/dialect/PostgreSQLDialect.java,v
retrieving revision 1.10
diff -u -r1.10 PostgreSQLDialect.java
--- PostgreSQLDialect.java	14 Jan 2005 19:19:20 -0000	1.10
+++ PostgreSQLDialect.java	18 Jan 2005 00:16:35 -0000
@@ -110,6 +110,11 @@
 	public String getCascadeConstraintsString() {
 		return "";//" cascade";
 	}
+
+    public String getQuerySequencesString() {
+        return "select relname as \"SEQUENCE_NAME\" from pg_class where relkind = 'S'";
+    }
+
 	public boolean dropConstraints() {
 		return true;
 	}
