Index: src/main/java/org/hibernate/mapping/UniqueKey.java
===================================================================
--- src/main/java/org/hibernate/mapping/UniqueKey.java	(révision 14918)
+++ src/main/java/org/hibernate/mapping/UniqueKey.java	(copie de travail)
@@ -15,6 +15,10 @@
 public class UniqueKey extends Constraint {
 
 	public String sqlConstraintString(Dialect dialect) {
+		if (getColumnSpan() == 1 && getColumn(0).isUnique() && dialect.supportsUnique()) {
+			// the column was already marked as 'unique' in the column description
+			return null;
+		}
 		StringBuffer buf = new StringBuffer( "unique (" );
 		Iterator iter = getColumnIterator();
 		boolean nullable = false;
