Index: tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/util/TestUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/util/TestUtil.java	(date 1414010775000)
+++ tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/util/TestUtil.java	(revision )
@@ -56,6 +56,8 @@
 	private static final String META_MODEL_CLASS_POSTFIX = "_";
 	private static final File OUT_BASE_DIR;
 
+	public static final String RESOURCE_SEPARATOR = "/";
+
 	static {
 		File targetDir = getTargetDir();
 		OUT_BASE_DIR = new File( targetDir, "processor-generated-test-classes" );
@@ -249,7 +251,7 @@
 	}
 
 	public static String fcnToPath(String fcn) {
-		return fcn.replace( PACKAGE_SEPARATOR, PATH_SEPARATOR );
+		return fcn.replace( PACKAGE_SEPARATOR, RESOURCE_SEPARATOR );
 	}
 
 	private static boolean hasFieldInMetamodelFor(Class<?> clazz, String fieldName) {
Index: tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/util/CompilationRunner.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/util/CompilationRunner.java	(date 1414010775000)
+++ tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/util/CompilationRunner.java	(revision )
@@ -98,7 +98,7 @@
 		if ( withMappingFiles != null ) {
 			String packageNameAsPath = TestUtil.fcnToPath( packageName );
 			for ( String mappingFile : withMappingFiles.value() ) {
-				mappingFiles.add( packageNameAsPath + File.separator + mappingFile );
+				mappingFiles.add( packageNameAsPath + TestUtil.RESOURCE_SEPARATOR + mappingFile );
 			}
 		}
 	}
