Index: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/util/xml/XmlParserHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/util/xml/XmlParserHelper.java	(date 1414010775000)
+++ tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/util/xml/XmlParserHelper.java	(revision )
@@ -84,6 +84,7 @@
 	 * @return an input stream for the specified resource or {@code null} in case resource cannot be loaded
 	 */
 	public InputStream getInputStreamForResource(String resource) {
+		resource = resource.replace('\\', '/'); // windows
 		// METAGEN-75
 		if ( !resource.startsWith( RESOURCE_PATH_SEPARATOR ) ) {
 			resource = RESOURCE_PATH_SEPARATOR + resource;
Index: .gitattributes
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .gitattributes	(revision )
+++ .gitattributes	(revision )
@@ -0,0 +1,5 @@
+# This file is used by unit tests and data I/O testing the CRLF/LF
+#  mode of the checked out working tree is important to ensure the unit
+#  test passes.
+hibernate-core/src/test/resources/org/hibernate/test/fileimport/multi-line-statements.sql eol=lf text
+
