Index: src/org/hibernate/bytecode/javassist/FastClass.java
===================================================================
--- src/org/hibernate/bytecode/javassist/FastClass.java	(revision 9939)
+++ src/org/hibernate/bytecode/javassist/FastClass.java	(working copy)
@@ -157,7 +157,7 @@
 	}
 
 	public boolean equals(Object o) {
-		if ( o == null || ( o instanceof FastClass ) ) {
+		if ( o == null || !( o instanceof FastClass ) ) {
 			return false;
 		}
 		return this.type.equals( ( ( FastClass ) o ).type );
