Index: src/main/java/org/hibernate/bytecode/javassist/FieldTransformer.java
===================================================================
--- src/main/java/org/hibernate/bytecode/javassist/FieldTransformer.java	(revision 18223)
+++ src/main/java/org/hibernate/bytecode/javassist/FieldTransformer.java	(working copy)
@@ -338,7 +338,8 @@
 			}
 			CodeAttribute codeAttr = minfo.getCodeAttribute();
 			if (codeAttr == null) {
-				return;
+				continue;
+				//return;
 			}
 			CodeIterator iter = codeAttr.iterator();
 			while (iter.hasNext()) {
