Subject: [PATCH] working in 6.1
---
Index: hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/EmbeddedTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/EmbeddedTest.java b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/EmbeddedTest.java
--- a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/EmbeddedTest.java	(revision 6110c0a98a3546b73b4abd0dc1f5333b874efb97)
+++ b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/EmbeddedTest.java	(date 1711094512758)
@@ -100,10 +100,13 @@
 			assertNotNull( p );
 			assertNotNull( p.address );
 			assertEquals( "Springfield", p.address.city );
+			assertEquals( "colorado street Springfield Matt Damon Land", p.address.fullAddress );
+
 			assertNotNull( p.address.country );
 			assertEquals( "DM", p.address.country.getIso2() );
 			assertNotNull( p.bornIn );
 			assertEquals( "US", p.bornIn.getIso2() );
+			assertEquals( "colorado street Springfield Matt Damon Land bis", p.addressBis.fullAddress );
 		} );
 	}
 
Index: hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Person.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Person.java b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Person.java
--- a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Person.java	(revision 6110c0a98a3546b73b4abd0dc1f5333b874efb97)
+++ b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Person.java	(date 1711094512763)
@@ -33,6 +33,9 @@
 	@Embedded
 	Address address;
 
+	@Embedded
+	AddressBis addressBis;
+
 	@Embedded
 	@AttributeOverrides( {
 			@AttributeOverride(name = "iso2", column = @Column(name = "bornIso2")),
Index: hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/AddressBis.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/AddressBis.java b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/AddressBis.java
new file mode 100644
--- /dev/null	(date 1711094697396)
+++ b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/AddressBis.java	(date 1711094697396)
@@ -0,0 +1,11 @@
+package org.hibernate.orm.test.bootstrap.binding.annotations.embedded;
+
+import jakarta.persistence.Embeddable;
+import org.hibernate.annotations.Formula;
+
+@Embeddable
+public class AddressBis {
+
+    @Formula("concat(concat(concat(concat(concat(address1, ' '), fld_city), ' '), countryName), ' bis')")
+    String fullAddress;
+}
Index: hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Address.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Address.java b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Address.java
--- a/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Address.java	(revision 6110c0a98a3546b73b4abd0dc1f5333b874efb97)
+++ b/hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/embedded/Address.java	(date 1711094512739)
@@ -11,6 +11,7 @@
 import jakarta.persistence.Column;
 import jakarta.persistence.Embeddable;
 import jakarta.persistence.ManyToOne;
+import org.hibernate.annotations.Formula;
 
 /**
  * @author Emmanuel Bernard
@@ -23,4 +24,7 @@
 	Country country;
 	@ManyToOne
 	AddressType type;
+
+	@Formula("concat(concat(concat(concat(address1, ' '), fld_city), ' '), countryName)")
+	String fullAddress;
 }
