diff --git a/hibernate-core/src/main/java/org/hibernate/context/internal/ThreadLocalSessionContext.java b/hibernate-core/src/main/java/org/hibernate/context/internal/ThreadLocalSessionContext.java
index 90cb644..1f8c00b 100644
--- a/hibernate-core/src/main/java/org/hibernate/context/internal/ThreadLocalSessionContext.java
+++ b/hibernate-core/src/main/java/org/hibernate/context/internal/ThreadLocalSessionContext.java
@@ -337,7 +337,8 @@ public class ThreadLocalSessionContext extends AbstractCurrentSessionContext {
 					     || "isTransactionInProgress".equals( methodName )
 					     || "setFlushMode".equals( methodName )
 						 || "getFactory".equals( methodName ) //from SessionImplementor
-					     || "getSessionFactory".equals( methodName ) ) {
+					     || "getSessionFactory".equals( methodName )
+						 || "getTenantIdentifier".equals( methodName ) ) {
 						LOG.tracev( "Allowing method [{0}] in non-transacted context", methodName );
 					}
 					else if ( "reconnect".equals( methodName )
