Index: ../../../../hibernate3.2.5/trunk/core/src/main/java/org/hibernate/dialect/MySQLDialect.java
===================================================================
--- ../../../../hibernate3.2.5/trunk/core/src/main/java/org/hibernate/dialect/MySQLDialect.java	(revision 17009)
+++ ../../../../hibernate3.2.5/trunk/core/src/main/java/org/hibernate/dialect/MySQLDialect.java	Tue Jul 07 12:37:27 EEST 2009
@@ -76,7 +76,7 @@
 		registerFunction("character_length", new StandardSQLFunction("character_length", Hibernate.LONG) );
 		registerFunction("lcase", new StandardSQLFunction("lcase") );
 		registerFunction("lower", new StandardSQLFunction("lower") );
-		registerFunction("length", new StandardSQLFunction("length", Hibernate.LONG) );
+		registerFunction("length", new StandardSQLFunction("char_length", Hibernate.LONG) );
 		registerFunction("ltrim", new StandardSQLFunction("ltrim") );
 		registerFunction("ord", new StandardSQLFunction("ord", Hibernate.INTEGER) );
 		registerFunction("quote", new StandardSQLFunction("quote") );
