--- hibernate-3.2.orig/src/org/hibernate/criterion/LikeExpression.java	2006-06-28 08:20:52.000000000 +0200
+++ hibernate-3.2/src/org/hibernate/criterion/LikeExpression.java	2007-12-07 10:42:20.426468800 +0100
@@ -68,8 +68,10 @@
 	public TypedValue[] getTypedValues(
 			Criteria criteria,
 			CriteriaQuery criteriaQuery) throws HibernateException {
+			
+		Object icvalue = ignoreCase ? value.toString().toLowerCase() : value;		
 		return new TypedValue[] {
-				criteriaQuery.getTypedValue( criteria, propertyName, value.toString().toLowerCase() )
+				criteriaQuery.getTypedValue( criteria, propertyName, icvalue.toString() )
 		};
 	}
 }
