Index: org/hibernate/hql/QueryTranslator.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/hql/QueryTranslator.java,v
retrieving revision 1.21
diff -u -r1.21 QueryTranslator.java
--- org/hibernate/hql/QueryTranslator.java	20 Jun 2005 19:36:56 -0000	1.21
+++ org/hibernate/hql/QueryTranslator.java	30 Aug 2005 07:22:27 -0000
@@ -138,6 +138,8 @@
 	 */
 	int[] getNamedParameterLocs(String name) throws QueryException;
 
+	String[] getNamedParameters();
+	
 	void validateScrollability() throws HibernateException;
 
 	boolean containsCollectionFetches();
Index: org/hibernate/hql/ast/HqlSqlWalker.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/hql/ast/HqlSqlWalker.java,v
retrieving revision 1.106
diff -u -r1.106 HqlSqlWalker.java
--- org/hibernate/hql/ast/HqlSqlWalker.java	13 Aug 2005 23:03:34 -0000	1.106
+++ org/hibernate/hql/ast/HqlSqlWalker.java	30 Aug 2005 07:22:27 -0000
@@ -882,4 +882,9 @@
 	public static void panic() {
 		throw new QueryException( "TreeWalker: panic" );
 	}
+
+
+	public String[] getNamedParameters() {
+		return (String[]) namedParameters.keySet().toArray(new String[0]);
+	}
 }
Index: org/hibernate/hql/ast/QueryTranslatorImpl.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/hql/ast/QueryTranslatorImpl.java,v
retrieving revision 1.63
diff -u -r1.63 QueryTranslatorImpl.java
--- org/hibernate/hql/ast/QueryTranslatorImpl.java	15 Aug 2005 22:42:27 -0000	1.63
+++ org/hibernate/hql/ast/QueryTranslatorImpl.java	30 Aug 2005 07:22:28 -0000
@@ -31,6 +31,7 @@
 import org.hibernate.loader.hql.QueryLoader;
 import org.hibernate.persister.entity.Queryable;
 import org.hibernate.type.Type;
+import org.hibernate.util.ArrayHelper;
 import org.hibernate.util.StringHelper;
 
 import java.util.HashMap;
@@ -346,6 +347,10 @@
 	public int[] getNamedParameterLocs(String name) {
 		return getWalker().getNamedParameterLocations( name );
 	}
+	
+	public String[] getNamedParameters() {
+		return getWalker().getNamedParameters();
+	}
 
 	public boolean containsCollectionFetches() {
 		errorIfDML();
Index: org/hibernate/hql/classic/QueryTranslatorImpl.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/hql/classic/QueryTranslatorImpl.java,v
retrieving revision 1.28
diff -u -r1.28 QueryTranslatorImpl.java
--- org/hibernate/hql/classic/QueryTranslatorImpl.java	15 Aug 2005 22:42:28 -0000	1.28
+++ org/hibernate/hql/classic/QueryTranslatorImpl.java	30 Aug 2005 07:22:29 -0000
@@ -489,6 +489,10 @@
 			return ArrayHelper.toIntArray( ( ArrayList ) o );
 		}
 	}
+	
+	public String[] getNamedParameters() {
+		return (String[]) namedParameters.keySet().toArray(new String[0]);
+	}
 
 	private void renderSQL() throws QueryException, MappingException {
 
