*** Configuration.java	Wed Dec 17 15:07:12 2003
--- C:\temp\clean\Configuration.java	Fri Jan 16 18:54:11 2004
***************
*** 242,252 ****
  		return this;
  	}
  
  	protected void add(org.dom4j.Document doc) throws Exception {
  		try {
! 			Binder.bindRoot( doc, createMappings() );
  		}
  		catch (MappingException me) {
  			log.error("Could not compile the mapping document", me);
  			throw me;
  		}
--- 242,252 ----
  		return this;
  	}
  
  	protected void add(org.dom4j.Document doc) throws Exception {
  		try {
! 			Binder.bindRoot( doc, createMappings(), this.properties.getProperty(Environment.DEFAULT_SCHEMA) );
  		}
  		catch (MappingException me) {
  			log.error("Could not compile the mapping document", me);
  			throw me;
  		}
