diff --git a/hibernate-core/src/test/java/org/hibernate/test/onetoone/optional/OptionalOneToOneTest.java b/hibernate-core/src/test/java/org/hibernate/test/onetoone/optional/OptionalOneToOneTest.java
index 3661494..cfa10d8 100644
--- a/hibernate-core/src/test/java/org/hibernate/test/onetoone/optional/OptionalOneToOneTest.java
+++ b/hibernate-core/src/test/java/org/hibernate/test/onetoone/optional/OptionalOneToOneTest.java
@@ -24,7 +24,6 @@
 package org.hibernate.test.onetoone.optional;
 
 import org.junit.Test;
-
 import org.hibernate.Session;
 import org.hibernate.cfg.Configuration;
 import org.hibernate.cfg.Environment;
@@ -65,4 +64,26 @@
 		s.getTransaction().commit();
 		s.close();
 	}
+
+	@Test
+	public void testRemoveAndInsertProxyEntity() {
+			Session s = openSession();
+			s.beginTransaction();
+			Person me = new Person();
+			me.name = "Steve";
+
+			s.persist( me );
+			s.getTransaction().commit();
+			s.close();
+
+			s = openSession();
+			s.beginTransaction();
+			me = ( Person ) s.load( Person.class, me.name );
+			s.delete( me );
+
+			s.persist(me);
+
+			s.getTransaction().commit();
+			s.close();
+	}
 }