Index: project/core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaExport.java
===================================================================
--- project/core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaExport.java	(revision 34)
+++ project/core/src/main/java/org/hibernate/tool/hbm2ddl/SchemaExport.java	Wed Jan 20 16:40:24 EET 2010
@@ -476,9 +476,7 @@
 					.setOutputFile( outFile )
 					.setImportFile( importFile )
 					.setDelimiter( delim );
-			if ( format ) {
-				se.setFormat( true );
-			}
+      se.setFormat(format);
 			se.execute( script, export, drop, create );
 
 		}
