Index: Oracle9Dialect.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/dialect/Oracle9Dialect.java,v
retrieving revision 1.17
diff -u -r1.17 Oracle9Dialect.java
--- Oracle9Dialect.java	19 Jun 2004 18:36:57 -0000	1.17
+++ Oracle9Dialect.java	25 Jun 2004 20:49:11 -0000
@@ -25,7 +25,7 @@
 		registerColumnType( Types.DOUBLE, "DOUBLE PRECISION" );
 		registerColumnType( Types.DATE, "DATE" );
 		registerColumnType( Types.TIME, "DATE" );
-		registerColumnType( Types.TIMESTAMP, "DATE" );
+		registerColumnType( Types.TIMESTAMP, "TIMESTAMP" );
 		//registerColumnType( Types.VARBINARY, "RAW" );
 		registerColumnType( Types.VARBINARY, "LONG RAW" );
 		registerColumnType( Types.VARBINARY, 2000, "RAW($l)" );
