diff --git a/hibernate-core/src/main/java/org/hibernate/mapping/SimpleValue.java b/hibernate-core/src/main/java/org/hibernate/mapping/SimpleValue.java
index a4f4d52..fdb15fc 100644
--- a/hibernate-core/src/main/java/org/hibernate/mapping/SimpleValue.java
+++ b/hibernate-core/src/main/java/org/hibernate/mapping/SimpleValue.java
@@ -435,11 +435,8 @@
 
 		// todo : cache the AttributeConverterTypeAdapter in case that AttributeConverter is applied multiple times.
 
-		final String name = String.format(
-				"BasicType adapter for AttributeConverter<%s,%s>",
-				entityAttributeJavaType.getSimpleName(),
-				databaseColumnJavaType.getSimpleName()
-		);
+		final String name = entityAttributeJavaType.getCanonicalName();
+
 		return new AttributeConverterTypeAdapter(
 				name,
 				attributeConverterDefinition.getAttributeConverter(),
