From 060e204361abdbc0370e01675e2d9829ee35d52c Mon Sep 17 00:00:00 2001
From: "Artem V. Navrotskiy" <bozaro@yandex.ru>
Date: Mon, 20 Feb 2012 00:19:05 +0400
Subject: [PATCH] ClassLoader fix

---
 .../internal/ClassLoaderServiceImpl.java           |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hibernate-core/src/main/java/org/hibernate/service/classloading/internal/ClassLoaderServiceImpl.java b/hibernate-core/src/main/java/org/hibernate/service/classloading/internal/ClassLoaderServiceImpl.java
index c175a87..0632755 100644
--- a/hibernate-core/src/main/java/org/hibernate/service/classloading/internal/ClassLoaderServiceImpl.java
+++ b/hibernate-core/src/main/java/org/hibernate/service/classloading/internal/ClassLoaderServiceImpl.java
@@ -86,7 +86,7 @@ public class ClassLoaderServiceImpl implements ClassLoaderService {
 		classLoadingClassLoaders.add( hibernateClassLoader );
 		classLoadingClassLoaders.add( environmentClassLoader );
 
-		this.classClassLoader = new ClassLoader() {
+		this.classClassLoader = new ClassLoader( applicationClassLoader ) {
 			@Override
 			protected Class<?> findClass(String name) throws ClassNotFoundException {
 				for ( ClassLoader loader : classLoadingClassLoaders ) {
-- 
1.7.5.4

