Index: src/org/hibernate/cfg/Configuration.java
===================================================================
--- src/org/hibernate/cfg/Configuration.java	(revision 9939)
+++ src/org/hibernate/cfg/Configuration.java	(working copy)
@@ -1595,7 +1595,7 @@
 
 	public void setListener(String type, Object listener) {
 		if ( listener == null ) {
-			setListener( type, null );
+			setListeners( type, (Object[]) null );
 		}
 		else {
 			Object[] listeners = (Object[]) Array.newInstance( eventListeners.getListenerClassFor(type), 1 );
