From 3165c8f7a0bb64f23dd78e593c062135bccc86ab Tue, 17 May 2011 01:22:56 +0200
From: Stevo Slavic <sslavic@gmail.com>
Date: Tue, 17 May 2011 01:17:39 +0200
Subject: [PATCH] Fixes build, compile tasks used platform encoding making build platform dependent.

Change-Id: I64d5496a71c96008b175f53b3febdad7ed6febfd


diff --git a/build.gradle b/build.gradle
index 33b4e61..1c6745d 100644
--- a/build.gradle
+++ b/build.gradle
@@ -163,6 +163,7 @@
                     compilerArgs: [
                             "-nowarn",
                             "-proc:only",
+							"-encoding", "UTF-8",
                             "-processor", "org.jboss.logging.LoggingToolsProcessor",
                             "-s", "$sourceSets.main.generatedLoggingSrcDir.absolutePath"
                     ]
@@ -177,7 +178,7 @@
         generateMainLoggingClasses.logging.captureStandardError(LogLevel.INFO)
 
 		compileJava.dependsOn generateMainLoggingClasses
-        compileJava.options.define(compilerArgs: ["-proc:none"])
+        compileJava.options.define(compilerArgs: ["-proc:none", "-encoding", "UTF-8"])
 
         manifest.mainAttributes(
                 provider: 'gradle',
