diff -bur hibernate-extensions-2.0/tools/src/net/sf/hibernate/tool/hbm2java/FinderRenderer.java hibernate-extensions-2.0-fix/tools/src/net/sf/hibernate/tool/hbm2java/FinderRenderer.java
--- hibernate-extensions-2.0/tools/src/net/sf/hibernate/tool/hbm2java/FinderRenderer.java	2003-06-13 06:15:31.000000000 -0600
+++ hibernate-extensions-2.0-fix/tools/src/net/sf/hibernate/tool/hbm2java/FinderRenderer.java	2003-08-14 07:34:53.000000000 -0600
@@ -333,6 +333,7 @@
 
         hibType.put("byte", "Hibernate.BYTE");
         hibType.put("short", "Hibernate.SHORT");
+        hibType.put("Integer", "Hibernate.INTEGER");
         hibType.put("int", "Hibernate.INTEGER");
         hibType.put("long", "Hibernate.LONG");
 
