--- hibernate-core\src\main\java\org\hibernate\type\DbTimestampType.java	2011-05-04 14:44:42.000000000 +0200
+++ d:\home\Downloads\DbTimestampType.java	2011-05-20 16:11:20.754870600 +0200
@@ -89,9 +89,10 @@
 
 	private Timestamp usePreparedStatement(String timestampSelectString, SessionImplementor session) {
 		PreparedStatement ps = null;
+		ResultSet rs = null;
 		try {
 			ps = session.getBatcher().prepareStatement( timestampSelectString );
-			ResultSet rs = session.getBatcher().getResultSet( ps );
+			rs = session.getBatcher().getResultSet( ps );
 			rs.next();
 			Timestamp ts = rs.getTimestamp( 1 );
 			if ( log.isTraceEnabled() ) {
@@ -114,7 +115,7 @@
 		finally {
 			if ( ps != null ) {
 				try {
-					session.getBatcher().closeStatement( ps );
+					session.getBatcher().closeQueryStatement( ps, rs );
 				}
 				catch( SQLException sqle ) {
 					log.warn( "unable to clean up prepared statement", sqle );

