Index: PropertyContainer.java
===================================================================
--- PropertyContainer.java	(revision 56219)
+++ PropertyContainer.java	(working copy)
@@ -276,9 +276,13 @@
 
 	private static boolean mustBeSkipped(XProperty property) {
 		//TODO make those hardcoded tests more portable (through the bytecode provider?)
-		return property.isAnnotationPresent( Transient.class )
-				|| "net.sf.cglib.transform.impl.InterceptFieldCallback".equals( property.getType().getName() )
-				|| "org.hibernate.bytecode.javassist.FieldHandler".equals( property.getType().getName() );
+        boolean result = property.isAnnotationPresent(Transient.class);
+        if (!result && property.getType() != null) {
+            result = result
+                    || "net.sf.cglib.transform.impl.InterceptFieldCallback".equals(property.getType().getName())
+                    || "org.hibernate.bytecode.javassist.FieldHandler".equals(property.getType().getName());
+        } 
+        return result;
 	}
 }
 
