diff --git a/hibernate-core/src/main/java/org/hibernate/engine/ActionQueue.java b/hibernate-core/src/main/java/org/hibernate/engine/ActionQueue.java
index cd12806..f76e01c 100644
--- a/hibernate-core/src/main/java/org/hibernate/engine/ActionQueue.java
+++ b/hibernate-core/src/main/java/org/hibernate/engine/ActionQueue.java
@@ -665,14 +665,10 @@ public class ActionQueue {
 			// now rebuild the insertions list. There is a batch for each entry in the name list.
 			for ( int i = 0; i < actionBatches.size(); i++ ) {
 				List batch = ( List ) actionBatches.get( new Integer( i ) );
-				ArrayList<EntityInsertAction> subInserts = new ArrayList<EntityInsertAction>();
-				for(int j=0; j < batch.size(); j++) {
-					Object aBatch = batch.get(j);
+				for ( Object aBatch : batch ) {
 					EntityInsertAction action = (EntityInsertAction) aBatch;
-					subInserts.add( action );
+					insertions.add( action );
 				}
-				java.util.Collections.sort(subInserts);
-				insertions.addAll(subInserts);
 			}
 		}
 
