Index: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/xml/JpaDescriptorParser.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/xml/JpaDescriptorParser.java	(revision e283d0f7ef4c3799f857f3016529edcef3d30544)
+++ tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/xml/JpaDescriptorParser.java	(revision )
@@ -138,7 +138,12 @@
 			context.logMessage(
 					Diagnostic.Kind.WARNING, "Unable to parse persistence.xml: " + e.getMessage()
 			);
+		} finally {
+			try {
+				stream.close();
+			} catch(IOException eat) {
-		}
+			}
+		}
 		return persistence;
 	}
 
@@ -157,6 +162,11 @@
 				context.logMessage(
 						Diagnostic.Kind.WARNING, "Unable to parse " + mappingFile + ": " + e.getMessage()
 				);
+			} finally {
+				try {
+					stream.close();
+				} catch(IOException eat) {
+				}
 			}
 			if ( mapping != null ) {
 				entityMappings.add( mapping );
