diff --git a/hibernate-entitymanager/src/test/java/org/hibernate/ejb/test/query/QueryTest.java b/hibernate-entitymanager/src/test/java/org/hibernate/ejb/test/query/QueryTest.java
index b64cc8c..e224226 100644
--- a/hibernate-entitymanager/src/test/java/org/hibernate/ejb/test/query/QueryTest.java
+++ b/hibernate-entitymanager/src/test/java/org/hibernate/ejb/test/query/QueryTest.java
@@ -53,6 +53,22 @@ public class QueryTest extends TestCase {
 		em.close();
 	}
 
+	public void testTypeExpression() throws Exception {
+		EntityManager em = getOrCreateEntityManager();
+		em.getTransaction().begin();
+		Item item = new Item( "Mouse", "Micro$oft mouse" );
+		em.persist( item );
+		item = new Item( "Computer", "Apple II" );
+		em.persist( item );
+		Query q = em.createQuery( "select i from Item i where TYPE(i) = :itemType" );
+		q.setParameter( "itemType", Item.class );
+		List result = q.getResultList();
+		assertNotNull( result );
+		assertEquals( 2, result.size() );
+		em.getTransaction().rollback();
+		em.close();
+	}
+
 	public void testParameterList() throws Exception {
 		final Item item = new Item( "Mouse", "Micro$oft mouse" );
 		final Item item2 = new Item( "Computer", "Dell computer" );
