--- hql-sql.g	2010-05-10 16:40:52.000000000 -0300
+++ /home/scott/old.scott/downloads/hql-sql.g	2010-04-19 22:58:06.000000000 -0300
@@ -502,6 +500,7 @@
 	| #(OR logicalExpr logicalExpr)
 	| #(NOT logicalExpr)
 	| comparisonExpr
+	| functionCall
 	;
 
 // TODO: Add any other comparison operators here.
--- sql-gen.g	2010-05-10 16:40:52.000000000 -0300
+++ /home/scott/old.scott/downloads/sql-gen.g	2010-04-19 23:24:42.000000000 -0300
@@ -278,6 +278,7 @@
 	: booleanOp [ parens ]
 	| comparisonExpr [ parens ]
 	| st:SQL_TOKEN { out(st); } // solely for the purpose of mapping-defined where-fragments
+	| methodCall
 	;
 	
 comparisonExpr[ boolean parens ]
