--- /cygdrive/c/oldtg.java	2004-01-10 17:51:25.123812800 +0100
+++ TableGenerator.java	2004-01-10 17:25:05.071809600 +0100
@@ -142,7 +142,13 @@
 	}
 	
 	public String sqlDropString(Dialect dialect) {
-		return "drop table " + tableName + dialect.getCascadeConstraintsString();
+        StringBuffer sqlDropString = new StringBuffer();
+        sqlDropString.append("drop table ");
+        if ( dialect.supportsIfExistsBeforeTableName() ) sqlDropString.append("if exists ");
+        sqlDropString.append(tableName);
+        sqlDropString.append(dialect.getCascadeConstraintsString());
+        if ( dialect.supportsIfExistsAfterTableName() ) sqlDropString.append(" if exists");   
+        return sqlDropString.toString();
 	}
 
 	public Object generatorKey() {
