Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/MapProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/MapProxy.java	(revision 18223)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/MapProxy.java	(working copy)
@@ -36,7 +36,7 @@
 public class MapProxy<K, V> implements Map<K, V>, Serializable {
     private static final long serialVersionUID = 8418037541773074646L;
 
-    private Initializor<Map<K, V>> initializor;
+    private transient Initializor<Map<K, V>> initializor;
     protected Map<K, V> delegate;
 
     public MapProxy() {
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedMapProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedMapProxy.java	(revision 18223)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedMapProxy.java	(working copy)
@@ -38,7 +38,7 @@
 public class SortedMapProxy<K, V> implements SortedMap<K, V>, Serializable {
     private static final long serialVersionUID = 2645817952901452375L;
 
-    private Initializor<SortedMap<K, V>> initializor;
+    private transient Initializor<SortedMap<K, V>> initializor;
     protected SortedMap<K, V> delegate;
 
     public SortedMapProxy() {
