diff --git a/hibernate-core/src/main/java/org/hibernate/resource/jdbc/internal/AbstractLogicalConnectionImplementor.java b/hibernate-core/src/main/java/org/hibernate/resource/jdbc/internal/AbstractLogicalConnectionImplementor.java
index 0202fa637d..dfc6d40a34 100644
--- a/hibernate-core/src/main/java/org/hibernate/resource/jdbc/internal/AbstractLogicalConnectionImplementor.java
+++ b/hibernate-core/src/main/java/org/hibernate/resource/jdbc/internal/AbstractLogicalConnectionImplementor.java
@@ -9,6 +9,7 @@ package org.hibernate.resource.jdbc.internal;
 import java.sql.Connection;
 import java.sql.SQLException;
 
+import org.hibernate.ConnectionReleaseMode;
 import org.hibernate.TransactionException;
 import org.hibernate.resource.jdbc.ResourceRegistry;
 import org.hibernate.resource.jdbc.spi.LogicalConnectionImplementor;
@@ -52,7 +53,11 @@ public abstract class AbstractLogicalConnectionImplementor implements LogicalCon
 	public void afterTransaction() {
 		log.trace( "LogicalConnection#afterTransaction" );
 
-		resourceRegistry.releaseResources();
+		if ( getConnectionHandlingMode().getReleaseMode() == ConnectionReleaseMode.AFTER_STATEMENT ||
+				getConnectionHandlingMode().getReleaseMode() == ConnectionReleaseMode.AFTER_TRANSACTION ) {
+
+			resourceRegistry.releaseResources();
+		}
 	}
 
 	// PhysicalJdbcTransaction impl ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
