Index: C:/Documents and Settings/Alex Burgel/workspace/Hibernate3/src/org/hibernate/impl/SessionFactoryImpl.java
===================================================================
--- C:/Documents and Settings/Alex Burgel/workspace/Hibernate3/src/org/hibernate/impl/SessionFactoryImpl.java	(revision 9854)
+++ C:/Documents and Settings/Alex Burgel/workspace/Hibernate3/src/org/hibernate/impl/SessionFactoryImpl.java	(working copy)
@@ -227,7 +227,7 @@
 			CacheConcurrencyStrategy cache = CacheFactory.createCache(
 				model.getCacheConcurrencyStrategy(),
 			    model.getCacheRegionName(),
-			    true,
+			    model.isMutable(),
 			    settings,
 			    properties
 			);
