Index: src/main/java/org/hibernate/id/enhanced/TableGenerator.java
===================================================================
--- src/main/java/org/hibernate/id/enhanced/TableGenerator.java	(revision 15564)
+++ src/main/java/org/hibernate/id/enhanced/TableGenerator.java	(working copy)
@@ -511,7 +511,7 @@
 						.append( segmentColumnName )
 						.append( ' ' )
 						.append( dialect.getTypeName( Types.VARCHAR, segmentValueLength, 0, 0 ) )
-						.append( ",  " )
+						.append( " not null,  " )
 						.append( valueColumnName )
 						.append( ' ' )
 						.append( dialect.getTypeName( Types.BIGINT ) )
