Index: D:/workspace/grammar/hql-sql.g
===================================================================
--- D:/workspace/grammar/hql-sql.g	(revision 11105)
+++ D:/workspace/grammar/hql-sql.g	(working copy)
@@ -382,6 +382,7 @@
 aggregateExpr
 	: expr //p:propertyRef { resolve(#p); }
 	| collectionFunction
+	| caseExpr	
 	;
 
 // Establishes the list of aliases being used by this query.
Index: D:/workspace/grammar/hql.g
===================================================================
--- D:/workspace/grammar/hql.g	(revision 11105)
+++ D:/workspace/grammar/hql.g	(working copy)
@@ -651,7 +651,7 @@
 aggregate
 	: ( SUM^ | AVG^ | MAX^ | MIN^ ) OPEN! additiveExpression CLOSE! { #aggregate.setType(AGGREGATE); }
 	// Special case for count - It's 'parameters' can be keywords.
-	|  COUNT^ OPEN! ( STAR { #STAR.setType(ROW_STAR); } | ( ( DISTINCT | ALL )? ( path | collectionExpr ) ) ) CLOSE!
+	|  COUNT^ OPEN! ( STAR { #STAR.setType(ROW_STAR); } | ( ( DISTINCT | ALL )? ( path | collectionExpr | caseExpression ) ) ) CLOSE!
 	|  collectionExpr
 	;
 
