--- hibernate-distribution-3.6.6.Final/project/hibernate-core/src/main/java/org/hibernate/engine/SubselectFetch.java	2011-07-20 18:21:22.000000000 -0700
+++ SubselectFetch.java	2011-11-16 09:05:50.063596237 -0800
@@ -58,7 +58,7 @@
 		
 		//TODO: ugly here:
 		final String queryString = queryParameters.getFilteredSQL();
-		int fromIndex = queryString.indexOf(" from ");
+		int fromIndex = getFromIndex(queryString);
 		int orderByIndex = queryString.lastIndexOf("order by");
 		this.queryString = orderByIndex>0 ? 
 				queryString.substring(fromIndex, orderByIndex) : 
@@ -66,6 +66,47 @@
 			
 	}
 
+	private int getFromIndex(String queryString)
+	{
+		String fromStr = " from ";
+
+		int index = queryString.indexOf(fromStr);
+
+		if (index < 0)
+			return index;
+
+		while (!parenthesesMatch(queryString.substring(0, index)))
+		{
+			String subString = queryString.substring(index + fromStr.length());
+
+			int subIndex = subString.indexOf(fromStr);
+
+			if (subIndex < 0)
+				return subIndex;
+
+			index += fromStr.length() + subIndex;
+		}
+
+		return index;
+	}
+
+	private boolean parenthesesMatch(String string)
+	{
+		int parenCount = 0;
+
+		for (int i = 0; i < string.length(); i++)
+		{
+			char character = string.charAt(i);
+
+			if (character == '(')
+				parenCount++;
+			else if (character == ')')
+				parenCount--;
+		}
+
+		return parenCount == 0;
+	}
+
 	public QueryParameters getQueryParameters() {
 		return queryParameters;
 	}
