Index: src/java/net/sf/hibernate/tool/hbm2java/BasicRenderer.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/tools/src/java/net/sf/hibernate/tool/hbm2java/BasicRenderer.java,v
retrieving revision 1.16
diff -u -w -b -B -u -r1.16 BasicRenderer.java
--- src/java/net/sf/hibernate/tool/hbm2java/BasicRenderer.java	9 Apr 2004 16:55:52 -0000	1.16
+++ src/java/net/sf/hibernate/tool/hbm2java/BasicRenderer.java	19 Apr 2004 14:55:30 -0000
@@ -376,7 +376,7 @@
 				classMapping.getImports().add("org.apache.commons.lang.builder.EqualsBuilder");
 				classMapping.getImports().add("org.apache.commons.lang.builder.HashCodeBuilder");
 		    	writer.println("    public boolean equals(Object other) {");
-		    	writer.println("        if ( (other == other ) ) return true;"); // == identity
+		    	writer.println("        if ( (this == other ) ) return true;"); // == identity
 		    	writer.println("        if ( !(other instanceof " + savedToClass + ") ) return false;"); // same class ?
 		    	writer.println("        " + savedToClass + " castOther = (" + savedToClass + ") other;"); 
 		    	writer.println("        return new EqualsBuilder()");
