Index: src/org/hibernate/bytecode/cglib/CglibClassTransformer.java
===================================================================
--- src/org/hibernate/bytecode/cglib/CglibClassTransformer.java	(.../base)	(revision 4329)
+++ src/org/hibernate/bytecode/cglib/CglibClassTransformer.java	(.../fixes/CAS-1100)	(revision 4329)
@@ -20,11 +20,11 @@
 import org.hibernate.bytecode.util.FieldFilter;
 import org.hibernate.bytecode.util.ClassFilter;
 import org.hibernate.HibernateException;
-import org.objectweb.asm.Attribute;
-import org.objectweb.asm.Type;
-import org.objectweb.asm.ClassReader;
-import org.objectweb.asm.ClassWriter;
-import org.objectweb.asm.attrs.Attributes;
+import net.sf.cglib.asm.Attribute;
+import net.sf.cglib.asm.Type;
+import net.sf.cglib.asm.ClassReader;
+import net.sf.cglib.asm.ClassWriter;
+import net.sf.cglib.asm.attrs.Attributes;
 
 /**
  * Enhance the classes allowing them to implements InterceptFieldEnabled
Index: src/org/hibernate/bytecode/cglib/BytecodeProviderImpl.java
===================================================================
--- src/org/hibernate/bytecode/cglib/BytecodeProviderImpl.java	(.../base)	(revision 4329)
+++ src/org/hibernate/bytecode/cglib/BytecodeProviderImpl.java	(.../fixes/CAS-1100)	(revision 4329)
@@ -18,7 +18,7 @@
 import org.hibernate.bytecode.ReflectionOptimizer;
 import org.hibernate.bytecode.util.FieldFilter;
 import org.hibernate.util.StringHelper;
-import org.objectweb.asm.Type;
+import net.sf.cglib.asm.Type;
 
 /**
  * Bytecode provider implementation for CGLIB.
Index: src/org/hibernate/tool/instrument/cglib/InstrumentTask.java
===================================================================
--- src/org/hibernate/tool/instrument/cglib/InstrumentTask.java	(.../base)	(revision 4329)
+++ src/org/hibernate/tool/instrument/cglib/InstrumentTask.java	(.../fixes/CAS-1100)	(revision 4329)
@@ -6,7 +6,7 @@
 import org.hibernate.bytecode.cglib.BytecodeProviderImpl;
 import org.hibernate.bytecode.ClassTransformer;
 import org.hibernate.tool.instrument.BasicInstrumentationTask;
-import org.objectweb.asm.ClassReader;
+import net.sf.cglib.asm.ClassReader;
 
 import java.io.ByteArrayInputStream;
