--- hibernate-extensions-2.0.2/tools/src/net/sf/hibernate/tool/hbm2java/ClassMapping.java	Tue Oct 21 15:14:16 2003
+++ /users/linus/projects/scratch/net/sf/hibernate/tools/hbm2java/ClassMapping.java	Thu Dec  4 09:31:11 2003
@@ -488,6 +488,9 @@
    private void doCollections(Element classElement, String xmlName, String interfaceClass, String implementingClass, MultiMap inheritedMeta) {
       
       
+	   String originalInterface = interfaceClass;
+	   String originalImplementingClass = implementingClass;
+
       for (Iterator collections = classElement.getChildren(xmlName).iterator(); collections.hasNext();) {
       
          Element collection = (Element) collections.next();
@@ -495,15 +498,21 @@
          String name = collection.getAttributeValue("name");
          
 			//		Small hack to switch over to sortedSet/sortedMap if sort is specified. (that is sort != unsorted)
-			if (!"unsorted".equals(collection.getAttributeValue("sort"))) {
-				  if("map".equals(xmlName)) {
-					  interfaceClass = SortedMap.class.getName();
-					  implementingClass = TreeMap.class.getName();
-				  } else if("set".equals(xmlName)) {
-					  interfaceClass = SortedSet.class.getName();
-					  implementingClass = TreeSet.class.getName();
-				  }
-			   }
+		 String sort = collection.getAttributeValue("sort");
+		 if ("unsorted".equals(sort) ||
+			 "".equals(sort)) {
+				  interfaceClass = originalInterface;
+				  implementingClass = originalImplementingClass;
+		   }
+		 else {
+			  if("map".equals(xmlName)) {
+				  interfaceClass = SortedMap.class.getName();
+				  implementingClass = TreeMap.class.getName();
+			  } else if("set".equals(xmlName)) {
+				  interfaceClass = SortedSet.class.getName();
+				  implementingClass = TreeSet.class.getName();
+			  }
+		   }
         
 		ClassName interfaceClassName = new ClassName();
 		ClassName implementingClassName = new ClassName();
