diff -cr hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3MergeEventListener.java hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3MergeEventListener.java
*** hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3MergeEventListener.java	Sat Sep 16 15:49:58 2006
--- hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3MergeEventListener.java	Tue Sep 26 13:27:02 2006
***************
*** 1,6 ****
--- 1,8 ----
  //$Id: EJB3MergeEventListener.java 9796 2006-04-26 06:46:52Z epbernard $
  package org.hibernate.ejb.event;
  
+ import java.io.Serializable;
+ import org.hibernate.HibernateException;
  import org.hibernate.event.EventSource;
  import org.hibernate.event.def.DefaultMergeEventListener;
  import org.hibernate.persister.entity.EntityPersister;
***************
*** 29,37 ****
  	}
  
  	@Override
! 	protected boolean invokeSaveLifecycle(Object entity, EntityPersister persister, EventSource source) {
! 		callbackHandler.preCreate( entity ); //always call the precreate event even if on safe vetoe it
! 		return super.invokeSaveLifecycle( entity, persister, source );
! 
  	}
  }
--- 31,55 ----
  	}
  
  	@Override
! 	protected Serializable saveWithGeneratedId(
! 			Object entity,
! 			String entityName,
! 			Object anything,
! 			EventSource source,
! 			boolean requiresImmediateIdAccess)
! 	throws HibernateException {
! 		callbackHandler.preCreate( entity );
! 		return super.saveWithGeneratedId(entity, entityName, anything, source, requiresImmediateIdAccess);
!  	}
! 	
! 	@Override
! 	protected Serializable saveWithRequestedId(
! 			Object entity,
! 			Serializable requestedId,
! 			String entityName,
! 			Object anything,
! 			EventSource source) 
! 	throws HibernateException {
! 		return super.saveWithRequestedId(entity, requestedId, entityName, anything, source);
  	}
  }
Only in hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event: EJB3MergeEventListener.java.orig
diff -cr hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3PersistEventListener.java hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3PersistEventListener.java
*** hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3PersistEventListener.java	Sat Sep 16 15:49:58 2006
--- hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3PersistEventListener.java	Tue Sep 26 13:27:24 2006
***************
*** 1,6 ****
--- 1,8 ----
  //$Id: EJB3PersistEventListener.java 9869 2006-05-04 03:17:03Z epbernard $
  package org.hibernate.ejb.event;
  
+ import java.io.Serializable;
+ import org.hibernate.HibernateException;
  import org.hibernate.event.EventSource;
  import org.hibernate.event.def.DefaultPersistEventListener;
  import org.hibernate.persister.entity.EntityPersister;
***************
*** 27,34 ****
  	}
  
  	@Override
! 	protected boolean invokeSaveLifecycle(Object entity, EntityPersister persister, EventSource source) {
! 		callbackHandler.preCreate( entity ); //always call the precreate event even if on safe vetoe it
! 		return super.invokeSaveLifecycle( entity, persister, source );
  	}
  }
--- 29,54 ----
  	}
  
  	@Override
! 	protected Serializable saveWithGeneratedId(
! 			Object entity,
! 			String entityName,
! 			Object anything,
! 			EventSource source,
! 			boolean requiresImmediateIdAccess)
! 	throws HibernateException {
! 		callbackHandler.preCreate( entity );
! 		return super.saveWithGeneratedId(entity, entityName, anything, source, requiresImmediateIdAccess);
!  	}
! 	
! 	@Override
! 	protected Serializable saveWithRequestedId(
! 			Object entity,
! 			Serializable requestedId,
! 			String entityName,
! 			Object anything,
! 			EventSource source) 
! 	throws HibernateException {
! 		return super.saveWithRequestedId(entity, requestedId, entityName, anything, source);
  	}
+ 
  }
Only in hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event: EJB3PersistEventListener.java.orig
diff -cr hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3SaveEventListener.java hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3SaveEventListener.java
*** hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3SaveEventListener.java	Sat Sep 16 15:49:58 2006
--- hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3SaveEventListener.java	Tue Sep 26 13:27:10 2006
***************
*** 1,6 ****
--- 1,8 ----
  //$Id: EJB3SaveEventListener.java 9796 2006-04-26 06:46:52Z epbernard $
  package org.hibernate.ejb.event;
  
+ import java.io.Serializable;
+ import org.hibernate.HibernateException;
  import org.hibernate.event.EventSource;
  import org.hibernate.event.def.DefaultSaveEventListener;
  import org.hibernate.persister.entity.EntityPersister;
***************
*** 29,36 ****
  	}
  
  	@Override
! 	protected boolean invokeSaveLifecycle(Object entity, EntityPersister persister, EventSource source) {
! 		callbackHandler.preCreate( entity ); //always call the precreate event even if on safe vetoe it
! 		return super.invokeSaveLifecycle( entity, persister, source );
  	}
  }
--- 31,55 ----
  	}
  
  	@Override
! 	protected Serializable saveWithGeneratedId(
! 			Object entity,
! 			String entityName,
! 			Object anything,
! 			EventSource source,
! 			boolean requiresImmediateIdAccess)
! 	throws HibernateException {
! 		callbackHandler.preCreate( entity );
! 		return super.saveWithGeneratedId(entity, entityName, anything, source, requiresImmediateIdAccess);
! 	}
! 	
! 	@Override
! 	protected Serializable saveWithRequestedId(
! 			Object entity,
! 			Serializable requestedId,
! 			String entityName,
! 			Object anything,
! 			EventSource source) 
! 	throws HibernateException {
! 		return super.saveWithRequestedId(entity, requestedId, entityName, anything, source);
  	}
  }
Only in hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event: EJB3SaveEventListener.java.orig
diff -cr hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java
*** hibernate-entitymanager-3.2.0.CR2.orig/src/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java	Sat Sep 16 15:49:58 2006
--- hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java	Tue Sep 26 13:26:47 2006
***************
*** 1,6 ****
--- 1,8 ----
  //$Id: EJB3SaveOrUpdateEventListener.java 9796 2006-04-26 06:46:52Z epbernard $
  package org.hibernate.ejb.event;
  
+ import java.io.Serializable;
+ import org.hibernate.HibernateException;
  import org.hibernate.event.EventSource;
  import org.hibernate.event.def.DefaultSaveOrUpdateEventListener;
  import org.hibernate.persister.entity.EntityPersister;
***************
*** 21,36 ****
  		super();
  	}
  
- 	;
- 
  	public EJB3SaveOrUpdateEventListener(EntityCallbackHandler callbackHandler) {
  		super();
  		this.callbackHandler = callbackHandler;
  	}
  
  	@Override
! 	protected boolean invokeSaveLifecycle(Object entity, EntityPersister persister, EventSource source) {
! 		callbackHandler.preCreate( entity ); //always call the precreate event even if on safe vetoe it
! 		return super.invokeSaveLifecycle( entity, persister, source );
  	}
  }
--- 23,54 ----
  		super();
  	}
  
  	public EJB3SaveOrUpdateEventListener(EntityCallbackHandler callbackHandler) {
  		super();
  		this.callbackHandler = callbackHandler;
  	}
  
  	@Override
! 	protected Serializable saveWithGeneratedId(
! 			Object entity,
! 			String entityName,
! 			Object anything,
! 			EventSource source,
! 			boolean requiresImmediateIdAccess)
! 	throws HibernateException {
! 		callbackHandler.preCreate( entity );
! 		return super.saveWithGeneratedId(entity, entityName, anything, source, requiresImmediateIdAccess);
! 	}
! 	
! 	@Override
! 	protected Serializable saveWithRequestedId(
! 			Object entity,
! 			Serializable requestedId,
! 			String entityName,
! 			Object anything,
! 			EventSource source) 
! 	throws HibernateException {
! 		return super.saveWithRequestedId(entity, requestedId, entityName, anything, source);
  	}
+ 
  }
Only in hibernate-entitymanager-3.2.0.CR2/src/org/hibernate/ejb/event: EJB3SaveOrUpdateEventListener.java.orig
