Index: core/trunk/core/src/main/java/org/hibernate/connection/ConnectionProviderFactory.java
===================================================================
--- core/trunk/core/src/main/java/org/hibernate/connection/ConnectionProviderFactory.java	(revision 17975)
+++ core/trunk/core/src/main/java/org/hibernate/connection/ConnectionProviderFactory.java	(working copy)
@@ -153,7 +153,8 @@
 		Properties result = new Properties();
 		while ( iter.hasNext() ) {
 			String prop = (String) iter.next();
-			if ( prop.indexOf(Environment.CONNECTION_PREFIX) > -1 && !SPECIAL_PROPERTIES.contains(prop) ) {
+			// PATCH: startsWith() instead of indexOf()
+			if ( prop.startsWith(Environment.CONNECTION_PREFIX) && !SPECIAL_PROPERTIES.contains(prop) ) {
 				result.setProperty(
 					prop.substring( Environment.CONNECTION_PREFIX.length()+1 ),
 					properties.getProperty(prop)
