Index: SQLLoader.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/loader/Attic/SQLLoader.java,v
retrieving revision 1.1.2.9
diff -u -r1.1.2.9 SQLLoader.java
--- SQLLoader.java	10 Sep 2003 16:51:07 -0000	1.1.2.9
+++ SQLLoader.java	16 Oct 2003 17:20:22 -0000
@@ -144,6 +144,16 @@
 					else {*/
 						columnAliases = currentPersister.getSubclassPropertyColumnAliases(propertyName, suffixes[currentPersisterIndex]);
 					//}
+					if (columnAliases == null) {
+						String[] idAliases = currentPersister.getIdentifierColumnNames();
+						if (idAliases != null) {
+							for (int i = 0; i < idAliases.length && columnAliases == null; i++) {
+								if (propertyName.equals(idAliases[i])) {
+									columnAliases = new String[] {currentPersister.getIdentifierAliases(suffixes[currentPersisterIndex])[i]};
+								}
+							}
+						}
+					}
 
 					if (columnAliases == null || columnAliases.length == 0) {
 						throw new QueryException("No column name found for property [" + propertyName + "]", sqlQuery);
