Index: D:/projects/hibernate3/src/org/hibernate/event/def/DefaultLoadEventListener.java
===================================================================
--- D:/projects/hibernate3/src/org/hibernate/event/def/DefaultLoadEventListener.java	(revision 9932)
+++ D:/projects/hibernate3/src/org/hibernate/event/def/DefaultLoadEventListener.java	(working copy)
@@ -227,7 +227,8 @@
 			log.trace("entity found in session cache");
 			if ( options.isCheckDeleted() ) {
 				EntityEntry entry = persistenceContext.getEntry(existing);
-				throwObjectDeletedIfNecessary(event, entry);
+				Status status = entry.getStatus();
+				if ( status == Status.DELETED || status == Status.GONE ) return null;
 			}
 			return existing;
 		}
@@ -407,24 +408,14 @@
 			// this object was already loaded
 			EntityEntry oldEntry = session.getPersistenceContext().getEntry(old);
 			if ( options.isCheckDeleted() ) {
-				throwObjectDeletedIfNecessary( event, oldEntry );
+				Status status = oldEntry.getStatus();
+				if ( status == Status.DELETED || status == Status.GONE ) return null;
 			}
 			upgradeLock( old, oldEntry, event.getLockMode(), session );
 		}
 		return old;
 	}
 
-	private void throwObjectDeletedIfNecessary(LoadEvent event, EntityEntry oldEntry) {
-		Status status = oldEntry.getStatus();
-		if ( status == Status.DELETED || status == Status.GONE ) {
-			throw new ObjectDeletedException(
-					"The object with that id was deleted",
-					event.getEntityId(),
-					event.getEntityClassName()
-				);
-		}
-	}
-
 	/**
 	 * Attempts to load the entity from the second-level cache.
 	 *
