Index: src/main/java/org/hibernate/envers/configuration/metadata/reader/AuditedPropertiesReader.java
===================================================================
--- src/main/java/org/hibernate/envers/configuration/metadata/reader/AuditedPropertiesReader.java	(revision 17248)
+++ src/main/java/org/hibernate/envers/configuration/metadata/reader/AuditedPropertiesReader.java	Sun Sep 13 20:16:02 EEST 2009
@@ -242,6 +242,7 @@
 						if (override.auditJoinTable() != null) {
 							propertyData.setJoinTable(override.auditJoinTable());
 						}
+						propertyData.setRelationTargetAuditMode(override.targetAuditMode());
 					}
 				}
 			}
Index: src/main/java/org/hibernate/envers/AuditOverride.java
===================================================================
--- src/main/java/org/hibernate/envers/AuditOverride.java	(revision 15899)
+++ src/main/java/org/hibernate/envers/AuditOverride.java	Sun Sep 13 20:16:02 EEST 2009
@@ -38,4 +38,11 @@
 	 * is ignored if {@link #isAudited()} equals to {@code false}.
 	 */
 	AuditJoinTable auditJoinTable() default @AuditJoinTable;
+
+	/**
+	 * To override {@link RelationTargetAuditMode}; defaults to {@link RelationTargetAuditMode#AUDITED}.
+	 * Its value is ignored if {@link #isAudited()} equals to {@code false}.  
+	 * @return {@link RelationTargetAuditMode}
+	 */
+	RelationTargetAuditMode targetAuditMode() default RelationTargetAuditMode.AUDITED;
 }
