Index: C:/Documents and Settings/Alex Burgel/workspace/Hibernate3.2/src/org/hibernate/cfg/Configuration.java
===================================================================
--- C:/Documents and Settings/Alex Burgel/workspace/Hibernate3.2/src/org/hibernate/cfg/Configuration.java	(revision 10840)
+++ C:/Documents and Settings/Alex Burgel/workspace/Hibernate3.2/src/org/hibernate/cfg/Configuration.java	(working copy)
@@ -938,13 +938,14 @@
 		String defaultCatalog = properties.getProperty( Environment.DEFAULT_CATALOG );
 		String defaultSchema = properties.getProperty( Environment.DEFAULT_SCHEMA );
 
+		Settings settings = buildSettings();
+
 		ArrayList script = new ArrayList( 50 );
 
 		Iterator iter = getTableMappings();
 		while ( iter.hasNext() ) {
 			Table table = (Table) iter.next();
 			if ( table.isPhysicalTable() ) {
-				Settings settings = buildSettings();
 				TableMetadata tableInfo = databaseMetadata.getTableMetadata(
 						table.getName(),
 						( table.getSchema() == null ) ? settings.getDefaultSchemaName() : table.getSchema(),
@@ -1062,11 +1063,12 @@
 			throws HibernateException {
 		secondPassCompile();
 
+		Settings settings = buildSettings();
+
 		Iterator iter = getTableMappings();
 		while ( iter.hasNext() ) {
 			Table table = (Table) iter.next();
 			if ( table.isPhysicalTable() ) {
-				Settings settings = buildSettings();
 
 				TableMetadata tableInfo = databaseMetadata.getTableMetadata(
 						table.getName(),
