Index: NewConfigurationWizardPage.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/org.hibernate.eclipse.console/src/org/hibernate/eclipse/console/wizards/NewConfigurationWizardPage.java,v
retrieving revision 1.2
diff -u -r1.2 NewConfigurationWizardPage.java
--- NewConfigurationWizardPage.java     20 Jan 2005 21:20:19 -0000      1.2
+++ NewConfigurationWizardPage.java     23 Jan 2005 15:57:23 -0000
@@ -332,7 +332,7 @@
      * @return
      */
     public String getDialect() {
-        return nullIfEmpty(dialectCombo.getText());
+        return nullIfEmpty(helper.getDialectClass(dialectCombo.getText()));
     }
 
     /**
