Index: tools/src/java/org/hibernate/cfg/reveng/ReverseEngineeringStrategyUtil.java
===================================================================
--- tools/src/java/org/hibernate/cfg/reveng/ReverseEngineeringStrategyUtil.java	(revision 17230)
+++ tools/src/java/org/hibernate/cfg/reveng/ReverseEngineeringStrategyUtil.java	(working copy)
@@ -134,6 +134,15 @@
 				singular += "es";
 				break;
 			case 'y':
+				if(singular.length()==1) {
+					// do not pluralize y to ies
+					break;
+				}
+				char penultimate = singular.charAt( singular.length()-2 );
+				if("aeiou".indexOf(penultimate)>=0) {
+					// do not pluralize key to keies or hay to haies
+					break;
+				}
 				singular = singular.substring( 0, singular.length()-1 ) + "ies";
 				break;
 			default:
