Index: EntityPersister.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/persister/EntityPersister.java,v
retrieving revision 1.38.2.40
diff -u -r1.38.2.40 EntityPersister.java
--- EntityPersister.java    10 Jan 2004 12:42:24 -0000  1.38.2.40
+++ EntityPersister.java    24 Feb 2004 23:10:33 -0000
@@ -505,7 +505,10 @@
                PreparedStatement insertSelect = session.getBatcher().prepareStatement(insertSelectSQL);
                try {
                    dehydrate(null, fields, notNull, insertSelect, session);
-                   return getGeneratedIdentity( object, session, insertSelect.executeQuery() );
+                   if (!insertSelect.execute()) {
+                       insertSelect.getMoreResults();
+                   }
+                   return getGeneratedIdentity( object, session, insertSelect.getResultSet() );
                }
                catch (SQLException sqle) {
                    session.getBatcher().closeStatement(insertSelect);
                    
Index: NormalizedEntityPersister.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/persister/NormalizedEntityPersister.java,v
retrieving revision 1.32.2.44
diff -u -r1.32.2.44 NormalizedEntityPersister.java
--- NormalizedEntityPersister.java  25 Jan 2004 03:54:15 -0000  1.32.2.44
+++ NormalizedEntityPersister.java  24 Feb 2004 23:11:38 -0000
@@ -501,8 +501,11 @@
                //use one statement to insert the row and get the generated id
                PreparedStatement insertSelect = session.getBatcher().prepareStatement(insertSelectSQL);
                try {
+                   if (!insertSelect.execute()) {
+                       insertSelect.getMoreResults();
+                   }
                    dehydrate(null, fields, notNull, 0, insertSelect, session);
-                   id = getGeneratedIdentity( object, session, insertSelect.executeQuery() );
+                   id = getGeneratedIdentity( object, session, insertSelect.getResultSet());
                }
                catch (SQLException sqle) {
                    session.getBatcher().closeStatement(insertSelect);