Index: PostgreSQLDialect.java
===================================================================
--- PostgreSQLDialect.java	(Revision 17199)
+++ PostgreSQLDialect.java	(Arbeitskopie)
@@ -359,4 +359,13 @@
 		// seems to have spotty LOB suppport
 		return false;
 	}
+
+	public boolean supportsPooledSequences() {
+		return true;
+	}
+
+	protected String getCreateSequenceString(String sequenceName, int initialValue,
+		int incrementSize) throws MappingException {
+		return getCreateSequenceString(sequenceName) + " minvalue " + initialValue + " increment " + incrementSize;
+	}
 }
