Index: src/org/hibernate/context/ManagedSessionContext.java
===================================================================
RCS file: /cvs/hibernate-3.1/src/org/hibernate/context/ManagedSessionContext.java,v
retrieving revision 1.1.1.1
diff -b -u -r1.1.1.1 ManagedSessionContext.java
--- src/org/hibernate/context/ManagedSessionContext.java	13 Jul 2006 21:10:43 -0000	1.1.1.1
+++ src/org/hibernate/context/ManagedSessionContext.java	5 Dec 2006 19:27:12 -0000
@@ -70,7 +70,7 @@
 	 * @param session The session to be bound.
 	 * @return Any previously bound session (should be null in most cases).
 	 */
-	public static Session bind(Session session) {
+	public static Session bind(org.hibernate.Session session) {
 		return ( Session ) sessionMap( true ).put( session.getSessionFactory(), session );
 	}
 
