Index: C:/workspace/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java
===================================================================
--- C:/workspace/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java	(revision 10368)
+++ C:/workspace/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java	(working copy)
@@ -74,6 +74,7 @@
 import org.hibernate.sql.SimpleSelect;
 import org.hibernate.sql.Template;
 import org.hibernate.sql.Update;
+import org.hibernate.stat.PersistedEntities;
 import org.hibernate.tuple.entity.EntityMetamodel;
 import org.hibernate.tuple.entity.EntityTuplizer;
 import org.hibernate.tuple.Tuplizer;
@@ -3536,6 +3537,7 @@
 
 	public Object instantiate(Serializable id, EntityMode entityMode)
 			throws HibernateException {
+		PersistedEntities.register(this);
 		return getTuplizer( entityMode ).instantiate( id );
 	}
 
