Index: src/org/hibernate/tool/hbm2ddl/SchemaUpdate.java
===================================================================
--- src/org/hibernate/tool/hbm2ddl/SchemaUpdate.java	(revision 14121)
+++ src/org/hibernate/tool/hbm2ddl/SchemaUpdate.java	(working copy)
@@ -1,7 +1,11 @@
 //$Id$
 package org.hibernate.tool.hbm2ddl;
 
+import java.io.File;
 import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Writer;
 import java.sql.Connection;
 import java.sql.SQLException;
 import java.sql.Statement;
@@ -31,6 +35,7 @@
 	private Configuration configuration;
 	private Dialect dialect;
 	private List exceptions;
+	private String outputFile = null;
 
 	public SchemaUpdate(Configuration cfg) throws HibernateException {
 		this( cfg, cfg.getProperties() );
@@ -116,6 +121,7 @@
 
 		Connection connection = null;
 		Statement stmt = null;
+		Writer outputFileWriter = null;
 
 		exceptions.clear();
 
@@ -134,6 +140,11 @@
 				log.error( "could not get database metadata", sqle );
 				throw sqle;
 			}
+			
+			if ( outputFile != null ) {
+				log.info( "writing generated schema to file: " + outputFile );
+				outputFileWriter = new FileWriter( outputFile );
+			}
 
 			log.info( "updating schema" );
 
@@ -145,6 +156,14 @@
 					if ( script ) {
 						System.out.println( sql );
 					}
+					
+					if (outputFileWriter!=null)
+					{
+						outputFileWriter.write(sql);
+						outputFileWriter.write("\n");
+					}
+					
+					
 					if ( doUpdate ) {
 						log.debug( sql );
 						stmt.executeUpdate( sql );
@@ -176,6 +195,17 @@
 				exceptions.add( e );
 				log.error( "Error closing connection", e );
 			}
+			
+			try {
+				if ( outputFileWriter != null ) {
+					outputFileWriter.close();
+				}
+				
+			}
+			catch ( IOException ioe ) {
+				exceptions.add( ioe );
+				log.error( "Error closing output file: " + outputFile, ioe );
+			}
 
 		}
 	}
@@ -188,4 +218,8 @@
 	public List getExceptions() {
 		return exceptions;
 	}
+	
+	public void setOutputFile(String outputFile) {
+		this.outputFile = outputFile;
+	}
 }
