Index: tools/src/java/org/hibernate/tool/ant/Hbm2DDLExporterTask.java
===================================================================
--- tools/src/java/org/hibernate/tool/ant/Hbm2DDLExporterTask.java	(revision 14121)
+++ tools/src/java/org/hibernate/tool/ant/Hbm2DDLExporterTask.java	(working copy)
@@ -37,10 +37,15 @@
 	}
 	
 	public void execute() {
-		if(schemaUpdate) {
+		if(schemaUpdate) { 
 			SchemaUpdate update = new SchemaUpdate(parent.getConfiguration() );
+			
+			if(outputfileName!=null) {
+				update.setOutputFile(new File(getDestdir(),outputfileName).toString() );
+			}
+			
 			update.execute(scriptToConsole, exportToDatabase);			
-		} 
+		}  
 		else {
 			SchemaExport export = new SchemaExport(parent.getConfiguration() );
 			if(outputfileName!=null) {
