Index: core/src/main/java/org/hibernate/type/ComponentType.java
===================================================================
--- core/src/main/java/org/hibernate/type/ComponentType.java	(revision 17621)
+++ core/src/main/java/org/hibernate/type/ComponentType.java	(working copy)
@@ -581,14 +581,15 @@
 			int length = propertyTypes[i].getColumnSpan( session.getFactory() );
 			String[] range = ArrayHelper.slice( names, begin, length ); //cache this
 			Object val = propertyTypes[i].hydrate( rs, range, session, owner );
-			if ( val == null ) {
-				if ( isKey ) {
-					return null; //different nullability rules for pk/fk
-				}
-			}
-			else {
-				notNull = true;
-			}
+//			if ( val == null ) {
+//				if ( isKey ) {
+//					return null; //different nullability rules for pk/fk
+//				}
+//			}
+//			else {
+//				notNull = true;
+//			}
+			notNull = true;
 			values[i] = val;
 			begin += length;
 		}
