Index: core/src/main/java/org/hibernate/id/TableGenerator.java
===================================================================
--- core/src/main/java/org/hibernate/id/TableGenerator.java	(revision 15563)
+++ core/src/main/java/org/hibernate/id/TableGenerator.java	Fri Nov 14 16:00:54 EST 2008
@@ -122,7 +122,7 @@
 
 	public String[] sqlCreateStrings(Dialect dialect) throws HibernateException {
 		return new String[] {
-			dialect.getCreateTableString() + " " + tableName + " ( " + columnName + " " + dialect.getTypeName(Types.INTEGER) + " )",
+			dialect.getCreateTableString() + " " + tableName + " ( " + columnName + " " + dialect.getTypeName(Types.INTEGER) + " )" + dialect.getTableTypeString(),
 			"insert into " + tableName + " values ( 0 )"
 		};
 	}
