Index: annotations/src/test/java/org/hibernate/test/annotations/manytoone/referencedcolumnname/ManyToOneReferencedColumnNameTest.java
===================================================================
--- annotations/src/test/java/org/hibernate/test/annotations/manytoone/referencedcolumnname/ManyToOneReferencedColumnNameTest.java	(revision 19799)
+++ annotations/src/test/java/org/hibernate/test/annotations/manytoone/referencedcolumnname/ManyToOneReferencedColumnNameTest.java	(working copy)
@@ -36,7 +36,7 @@
  */
 public class ManyToOneReferencedColumnNameTest extends TestCase {
 	@RequiresDialectFeature(DialectChecks.SupportsIdentityColumns.class)
-	public void testReoverableExceptionInFkOrdering() throws Exception {
+	public void testRecoverableExceptionInFkOrdering() throws Exception {
 		//SF should not blow up
 		Vendor v = new Vendor();
 		Item i = new Item();
Index: annotations/src/test/java/org/hibernate/test/annotations/entity/BasicHibernateAnnotationsTest.java
===================================================================
--- annotations/src/test/java/org/hibernate/test/annotations/entity/BasicHibernateAnnotationsTest.java	(revision 19799)
+++ annotations/src/test/java/org/hibernate/test/annotations/entity/BasicHibernateAnnotationsTest.java	(working copy)
@@ -114,7 +114,7 @@
 		forest.setLength( 22 );
 		try {
 			parallelTx.commit();
-			fail( "All optimistic locking should have make it fail" );
+			fail( "All optimistic locking should have made it fail" );
 		}
 		catch (HibernateException e) {
 			if ( parallelTx != null ) parallelTx.rollback();
Index: annotations/src/main/java/org/hibernate/annotations/PolymorphismType.java
===================================================================
--- annotations/src/main/java/org/hibernate/annotations/PolymorphismType.java	(revision 19799)
+++ annotations/src/main/java/org/hibernate/annotations/PolymorphismType.java	(working copy)
@@ -24,7 +24,7 @@
 package org.hibernate.annotations;
 
 /**
- * Type of avaliable polymorphism for a particular entity
+ * Type of available polymorphism for a particular entity
  *
  * @author Emmanuel Bernard
  */
@@ -34,7 +34,7 @@
 	 */
 	IMPLICIT,
 	/**
-	 * this entity is retrived only if explicitly asked
+	 * this entity is retrieved only if explicitly asked
 	 */
 	EXPLICIT
 }
\ No newline at end of file
