Index: EntityKey.java
===================================================================
--- EntityKey.java	(revision 15138)
+++ EntityKey.java	(working copy)
@@ -94,7 +94,7 @@
 
 	public boolean equals(Object other) {
 		EntityKey otherKey = (EntityKey) other;
-		return otherKey.rootEntityName.equals(this.rootEntityName) && 
+		return otherKey.entityName.equals(this.entityName) && 
 			identifierType.isEqual(otherKey.identifier, this.identifier, entityMode, factory);
 	}
 	
