Index: FooBarTest.java
===================================================================
retrieving revision 1.1.2.22
diff -u -r1.1.2.22 FooBarTest.java
--- FooBarTest.java	17 Jan 2004 23:09:26 -0000	1.1.2.22
+++ FooBarTest.java	2 Feb 2004 11:29:45 -0000
@@ -50,6 +50,7 @@
 import net.sf.hibernate.dialect.PointbaseDialect;
 import net.sf.hibernate.dialect.PostgreSQLDialect;
 import net.sf.hibernate.dialect.SAPDBDialect;
+import net.sf.hibernate.dialect.SQLServerDialect;
 import net.sf.hibernate.dialect.SybaseDialect;
 import net.sf.hibernate.engine.SessionFactoryImplementor;
 import net.sf.hibernate.expression.Example;
@@ -1478,11 +1479,10 @@
 			//should happen
 		}
 		
-		
-		if (! (getDialect() instanceof Oracle9Dialect) ) { // oracle barfs on "x in ()" 
-		q.setParameterList("nameList", Collections.EMPTY_LIST);
-		list = q.list();
-		assertTrue( list.size()==0 );
+		if (! (getDialect() instanceof Oracle9Dialect) && ! (getDialect() instanceof SQLServerDialect) ) { // oracle and SQLserver barf on "x in ()"
+			q.setParameterList("nameList", Collections.EMPTY_LIST);
+			list = q.list();
+			assertTrue( list.size()==0 );
 		}
 		
 		q = s.createQuery("select bar, b from Bar bar inner join bar.baz baz inner join baz.cascadingBars b where bar.name like 'Bar%'");
