Index: core/src/main/java/org/hibernate/loader/Loader.java
===================================================================
--- core/src/main/java/org/hibernate/loader/Loader.java	(revision 20078)
+++ core/src/main/java/org/hibernate/loader/Loader.java	(revision )
@@ -1775,7 +1775,9 @@
 		final int firstRow = dialect.convertToFirstRowValue( getFirstRow( selection ) );
 		final int lastRow = selection.getMaxRows().intValue();
 		if ( dialect.useMaxForLimit() ) {
-			return lastRow + firstRow;
+            //this should be overflow safe
+            long result = (long) lastRow + (long) firstRow;
+            return result > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) result;
 		}
 		else {
 			return lastRow;
@@ -1820,7 +1822,9 @@
 			final PreparedStatement st,
 			final RowSelection selection) throws SQLException {
 		if ( hasMaxRows( selection ) ) {
-			st.setMaxRows( selection.getMaxRows().intValue() + interpretFirstRow( getFirstRow( selection ) ) );
+            //to prevent integer overflow
+            long maxRows = (long) selection.getMaxRows().intValue() + (long) interpretFirstRow(getFirstRow(selection));
+            st.setMaxRows(maxRows > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) maxRows);
 		}
 	}
 
