Index: src/main/java/org/hibernate/event/def/DefaultRefreshEventListener.java
===================================================================
--- src/main/java/org/hibernate/event/def/DefaultRefreshEventListener.java	(revision 19408)
+++ src/main/java/org/hibernate/event/def/DefaultRefreshEventListener.java	(working copy)
@@ -152,7 +152,13 @@
 		// Keep the same read-only/modifiable setting for the entity that it had before refreshing;
 		// If it was transient, then set it to the default for the source.
 		if ( result != null ) {
-			source.setReadOnly( result, ( e == null ? source.isDefaultReadOnly() : e.isReadOnly() ) );
+			if ( ! persister.isMutable() ) {
+				// this is probably redundant; it should already be read-only
+				source.setReadOnly( result, true );
+			}
+			else {
+				source.setReadOnly( result, ( e == null ? source.isDefaultReadOnly() : e.isReadOnly() ) );
+			}
 		}
 		source.setFetchProfile(previousFetchProfile);
 		
