diff -Naur Hibernate2.orig/src/net/sf/hibernate/type/CharacterType.java Hibernate2/src/net/sf/hibernate/type/CharacterType.java
--- Hibernate2.orig/src/net/sf/hibernate/type/CharacterType.java	2003-03-01 13:15:28.000000000 +0100
+++ Hibernate2/src/net/sf/hibernate/type/CharacterType.java	2003-09-05 13:41:24.000000000 +0200
@@ -16,7 +16,7 @@
 			return null;
 		}
 		else {
-			return new Character( str.charAt(0) );
+			return new Character( str.length()>0 ? str.charAt(0) : ' ' );
 		}
 	}
 	
