Index: src/main/java/org/hibernate/cfg/Ejb3JoinColumn.java
===================================================================
--- src/main/java/org/hibernate/cfg/Ejb3JoinColumn.java	(revision 18147)
+++ src/main/java/org/hibernate/cfg/Ejb3JoinColumn.java	(working copy)
@@ -566,14 +566,15 @@
 		// columnDefinition can also be specified using @JoinColumn, hence we have to check
 		// whether it is set or not
 		if ( StringHelper.isEmpty( sqlType ) ) {
-			sqlType = column.getSqlType();
-			if ( getMappingColumn() != null ) getMappingColumn().setSqlType( sqlType );
+			sqlType = column.getSqlType();			
 		}
-		
-		// these properties can only be applied on the referenced column - we can just take them over
-		getMappingColumn().setLength(column.getLength());
-		getMappingColumn().setPrecision(column.getPrecision());
-		getMappingColumn().setScale(column.getScale());		
+		if ( getMappingColumn() != null ) {
+			getMappingColumn().setSqlType( sqlType );		
+			// these properties can only be applied on the referenced column - we can just take them over
+			getMappingColumn().setLength(column.getLength());
+			getMappingColumn().setPrecision(column.getPrecision());
+			getMappingColumn().setScale(column.getScale());
+		}
 	}
 
 	@Override
