Index: entitymanager/src/test/java/org/hibernate/ejb/test/lock/LockTest.java
===================================================================
--- entitymanager/src/test/java/org/hibernate/ejb/test/lock/LockTest.java	(revision 19476)
+++ entitymanager/src/test/java/org/hibernate/ejb/test/lock/LockTest.java	(working copy)
@@ -232,6 +232,32 @@
 		 em2.close();
 	}
 
+	public void testLockOptimisticForceIncrementDifferentEmDefaultLock() throws Exception {
+		Lock lock = new Lock();
+		lock.setName( "force" );
+		EntityManager em1 = createIsolatedEntityManager();
+		em1.getTransaction().begin();
+		em1.persist( lock );
+		em1.getTransaction().commit();
+		em1.close();
+	    
+		EntityManager em2 = createIsolatedEntityManager();
+		em2.getTransaction().begin();
+		lock = em2.find( Lock.class, lock.getId() );
+		if ( LockModeType.OPTIMISTIC == em2.getLockMode(lock) ) {
+			// In case OPTIMISTIC is default, it should still work correctly
+			try {
+				em2.lock( lock, LockModeType.OPTIMISTIC_FORCE_INCREMENT );
+				assertEquals( "lock mode should be OPTIMISTIC_FORCE_INCREMENT ", LockModeType.OPTIMISTIC_FORCE_INCREMENT, em2.getLockMode(lock) );
+			}
+			finally {
+				em2.remove( lock );
+				em2.getTransaction().commit();
+			}
+		}
+		em2.close();
+	}
+
 	public void testContendedPessimisticLock() throws Exception {
 
 		EntityManager em = getOrCreateEntityManager();
