diff -u -r src.orig/org/hibernate/annotations/Index.java src/org/hibernate/annotations/Index.java
--- src.orig/org/hibernate/annotations/Index.java	2006-01-20 14:10:54.000000000 +0100
+++ src/org/hibernate/annotations/Index.java	2006-02-25 18:05:53.837073227 +0100
@@ -12,6 +12,6 @@
  */
 @Target({FIELD, METHOD}) @Retention(RUNTIME)
 public @interface Index {
-	String name();
+	String name() default "";
 	String[] columnNames() default {};
 }
diff -u -r src.orig/org/hibernate/cfg/annotations/TableBinder.java src/org/hibernate/cfg/annotations/TableBinder.java
--- src.orig/org/hibernate/cfg/annotations/TableBinder.java	2006-01-20 14:10:54.000000000 +0100
+++ src/org/hibernate/cfg/annotations/TableBinder.java	2006-02-25 18:12:21.861930455 +0100
@@ -2,6 +2,7 @@
 package org.hibernate.cfg.annotations;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
 import javax.persistence.UniqueConstraint;
@@ -320,8 +321,12 @@
 
 	public static void addIndexes(Table hibTable, Index[] indexes, ExtendedMappings mappings) {
 		for ( Index index : indexes ) {
+			String indexName = index.name();
+			if ( indexName.length() == 0 ) {
+				indexName = "IX" + hibTable.uniqueColumnString( Arrays.asList( index.columnNames() ).iterator() );
+			}
 			mappings.addSecondPass(
-					new IndexSecondPass( hibTable, index.name(), index.columnNames(), mappings )
+					new IndexSecondPass( hibTable, indexName, index.columnNames(), mappings )
 			);
 		}
 	}
