diff -ru -x '*gradle*' -x '*target*' -x '*build*' -x '*.log' hibernate-orm-5.2.4-source/hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java hibernate-orm-5.2.4/hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java
--- hibernate-orm-5.2.4-source/hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java	2016-10-26 15:41:28.000000000 +0200
+++ hibernate-orm-5.2.4/hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java	2016-11-13 23:01:49.920852588 +0100
@@ -611,12 +611,15 @@
 			}
 		}
 		private void handleDotStructure(AST dotStructureRoot) {
-			final String expression = ASTUtil.getPathText( dotStructureRoot );
-			final Object constant = ReflectHelper.getConstantValue( expression, factory.getServiceRegistry().getService( ClassLoaderService.class ) );
-			if ( constant != null ) {
-				dotStructureRoot.setFirstChild( null );
-				dotStructureRoot.setType( HqlTokenTypes.JAVA_CONSTANT );
-				dotStructureRoot.setText( expression );
+			final String expression = ASTUtil.getPathText( dotStructureRoot ); 
+			final Object entity = factory.getMetamodel().entity( expression );
+			if( entity == null && expression.indexOf('.') != -1 ) {
+				final Object constant = ReflectHelper.getConstantValue( expression, factory.getServiceRegistry().getService( ClassLoaderService.class ) );
+				if ( constant != null ) {
+					dotStructureRoot.setFirstChild( null );
+					dotStructureRoot.setType( HqlTokenTypes.JAVA_CONSTANT );
+					dotStructureRoot.setText( expression );
+				}
 			}
 		}
 	}
