Index: test/org/hibernate/test/generated/AbstractGeneratedPropertyTest.java
===================================================================
--- test/org/hibernate/test/generated/AbstractGeneratedPropertyTest.java	(revision 11639)
+++ test/org/hibernate/test/generated/AbstractGeneratedPropertyTest.java	(working copy)
@@ -17,6 +17,7 @@
 	}
 
 	public final void testGeneratedProperty() {
+		for (int i=0; i<500; i++) {
 		GeneratedPropertyEntity entity = new GeneratedPropertyEntity();
 		entity.setName( "entity-1" );
 		Session s = openSession();
@@ -43,5 +44,6 @@
 		s.delete( entity );
 		t.commit();
 		s.close();
+		}
 	}
 }
Index: test/org/hibernate/test/instrument/cases/TestLazyExecutable.java
===================================================================
--- test/org/hibernate/test/instrument/cases/TestLazyExecutable.java	(revision 11639)
+++ test/org/hibernate/test/instrument/cases/TestLazyExecutable.java	(working copy)
@@ -17,6 +17,7 @@
  */
 public class TestLazyExecutable extends AbstractExecutable {
 	public void execute() {
+		for (int i=0; i<100; i++) {
 		SessionFactory factory = getFactory();
 		Session s = factory.openSession();
 		Transaction t = s.beginTransaction();
@@ -192,6 +193,7 @@
 		s.flush();
 		t.commit();
 		s.close();
+		}
 	}
 
 }
