Index: C:/workspace/bin-hibernate-3.2.3/src/org/hibernate/jdbc/AbstractBatcher.java
===================================================================
--- C:/workspace/bin-hibernate-3.2.3/src/org/hibernate/jdbc/AbstractBatcher.java	(revision 4324)
+++ C:/workspace/bin-hibernate-3.2.3/src/org/hibernate/jdbc/AbstractBatcher.java	(working copy)
@@ -119,15 +119,10 @@
 
 	public PreparedStatement prepareSelectStatement(String sql)
 			throws SQLException, HibernateException {
-		logOpenPreparedStatement();
-		return getPreparedStatement(
-				connectionManager.getConnection(),
-		        sql,
-		        false,
+		return prepareQueryStatement(
+				sql,
 		        false,
-		        null,
-		        null,
-		        false
+		        null
 		);
 	}
 
