Index: src/org/hibernate/collection/PersistentMap.java
===================================================================
--- src/org/hibernate/collection/PersistentMap.java	(revision 1773)
+++ src/org/hibernate/collection/PersistentMap.java	(revision 1774)
@@ -332,7 +332,7 @@
 	throws HibernateException {
 		final java.util.Map sn = (java.util.Map) getSnapshot();
 		java.util.Map.Entry e = (java.util.Map.Entry) entry;
-		return e.getValue()!=null && sn.get( e.getKey() )==null;
+		return e.getValue()!=null && !sn.containsKey( e.getKey() );
 	}
 
 	public boolean needsUpdating(Object entry, int i, Type elemType) 
@@ -340,8 +340,8 @@
 		final java.util.Map sn = (java.util.Map) getSnapshot();
 		java.util.Map.Entry e = (java.util.Map.Entry) entry;
 		Object snValue = sn.get( e.getKey() );
-		return e.getValue()!=null &&
-			snValue!=null &&
+		return e.getValue()!=null && 
+			sn.containsKey(e.getKey()) &&
 			elemType.isDirty( snValue, e.getValue(), getSession() );
 	}
 
