Index: src/test/java/org/hibernate/test/hql/HQLTest.java
===================================================================
--- src/test/java/org/hibernate/test/hql/HQLTest.java	(revision 18998)
+++ src/test/java/org/hibernate/test/hql/HQLTest.java	(working copy)
@@ -14,6 +14,7 @@
 
 import org.hibernate.Hibernate;
 import org.hibernate.QueryException;
+import org.hibernate.Session;
 import org.hibernate.dialect.DB2Dialect;
 import org.hibernate.dialect.HSQLDialect;
 import org.hibernate.dialect.MySQLDialect;
@@ -84,7 +85,22 @@
 		SqlGenerator.REGRESSION_STYLE_CROSS_JOINS = false;
 		super.cleanupTest();
 	}
+    public void testCast() {
+        Session session = openSession();
+        session.beginTransaction();
 
+        session
+                .createQuery(
+                        "select cast(h.id as string) from HeresAnotherCrazyIdFieldName h" )
+                .list();
+        session
+                .createQuery(
+                        "select h.id from HeresAnotherCrazyIdFieldName h where h.name = cast(? as string)" )
+                .setString( 0, "stliu" ).list();
+        session.createQuery( "select h.id from HeresAnotherCrazyIdFieldName h where h.name = cast('stliu' as string)" ).list();
+        session.getTransaction().commit();
+        session.close();
+    }
 	public void testModulo() {
 		assertTranslation( "from Animal a where a.bodyWeight % 2 = 0" );
 	}
Index: src/test/java/org/hibernate/test/hql/ClassicTranslatorTest.java
===================================================================
--- src/test/java/org/hibernate/test/hql/ClassicTranslatorTest.java	(revision 18998)
+++ src/test/java/org/hibernate/test/hql/ClassicTranslatorTest.java	(working copy)
@@ -61,4 +61,21 @@
 		session.getTransaction().commit();
 		session.close();
 	}
+	
+    public void testCast() {
+        Session session = openSession();
+        session.beginTransaction();
+
+        session
+                .createQuery(
+                        "select cast(h.id as string) from HeresAnotherCrazyIdFieldName h" )
+                .list();
+        session
+                .createQuery(
+                        "select h.id from HeresAnotherCrazyIdFieldName h where h.name = cast(? as string)" )
+                .setString( 0, "stliu" ).list();
+        session.createQuery( "select h.id from HeresAnotherCrazyIdFieldName h where h.name = cast('stliu' as string)" ).list();
+        session.getTransaction().commit();
+        session.close();
+    }
 }
