Index: PostgresUUIDType.java
===================================================================
--- PostgresUUIDType.java	(revision 19697)
+++ PostgresUUIDType.java	(working copy)
@@ -52,10 +52,17 @@
 		super( PostgresUUIDSqlTypeDescriptor.INSTANCE, UUIDTypeDescriptor.INSTANCE );
 	}
 
+	
 	public String getName() {
 		return "pg-uuid";
 	}
-
+	
+	@Override
+	protected boolean registerUnderJavaType() {
+		return true;
+	}
+	
+	
 	public static class PostgresUUIDSqlTypeDescriptor implements SqlTypeDescriptor {
 		public static final PostgresUUIDSqlTypeDescriptor INSTANCE = new PostgresUUIDSqlTypeDescriptor();
 
@@ -68,7 +75,7 @@
 			return new BasicBinder<X>( javaTypeDescriptor, this ) {
 				@Override
 				protected void doBind(PreparedStatement st, X value, int index, WrapperOptions options) throws SQLException {
-					st.setObject( index, javaTypeDescriptor.unwrap( value, UUID.class, options ) );
+					st.setObject( index, javaTypeDescriptor.unwrap( value, UUID.class, options ),getSqlType() );				
 				}
 			};
 		}
@@ -81,5 +88,5 @@
 				}
 			};
 		}
-	}
+	}	
 }
