--- .\org\hibernate\cfg\HbmBinder.java	Tue Dec 05 15:39:12 2006
+++ c:\workspace-jboss\HibernateMappingTests\mymod\HbmBinder.java	Sat May 05 13:39:05 2007
@@ -2240,7 +2240,7 @@
 			String entityName = ( (OneToMany) list.getElement() ).getReferencedEntityName();
 			PersistentClass referenced = mappings.getClass( entityName );
 			IndexBackref ib = new IndexBackref();
-			ib.setName( '_' + node.attributeValue( "name" ) + "IndexBackref" );
+			ib.setName( '_' + list.getOwnerEntityName() + "." + node.attributeValue( "name" ) + "IndexBackref" );
 			ib.setUpdateable( false );
 			ib.setSelectable( false );
 			ib.setCollectionRole( list.getRole() );
@@ -2343,7 +2343,7 @@
 			String entityName = ( (OneToMany) map.getElement() ).getReferencedEntityName();
 			PersistentClass referenced = mappings.getClass( entityName );
 			IndexBackref ib = new IndexBackref();
-			ib.setName( '_' + node.attributeValue( "name" ) + "IndexBackref" );
+			ib.setName( '_' + map.getOwnerEntityName() + "." + node.attributeValue( "name" ) + "IndexBackref" );
 			ib.setUpdateable( false );
 			ib.setSelectable( false );
 			ib.setCollectionRole( map.getRole() );
@@ -2476,7 +2476,7 @@
 			String entityName = ( (OneToMany) collection.getElement() ).getReferencedEntityName();
 			PersistentClass referenced = mappings.getClass( entityName );
 			Backref prop = new Backref();
-			prop.setName( '_' + node.attributeValue( "name" ) + "Backref" );
+			prop.setName( '_' + collection.getOwnerEntityName() + "." + node.attributeValue( "name" ) + "Backref" );
 			prop.setUpdateable( false );
 			prop.setSelectable( false );
 			prop.setCollectionRole( collection.getRole() );
