Index: impl/bridge/QueryResultsRegionAdapter.java
===================================================================
--- impl/bridge/QueryResultsRegionAdapter.java	(revision 15019)
+++ impl/bridge/QueryResultsRegionAdapter.java	(working copy)
@@ -37,4 +37,8 @@
 	protected QueryResultsRegionAdapter(Cache underlyingCache, Settings settings) {
 		super( underlyingCache, settings );
 	}
-}
+
+	public long nextTimestamp() {
+		return underlyingCache.nextTimestamp();
+	}
+}
\ No newline at end of file
Index: QueryResultsRegion.java
===================================================================
--- QueryResultsRegion.java	(revision 15019)
+++ QueryResultsRegion.java	(working copy)
@@ -31,4 +31,6 @@
  * @author Steve Ebersole
  */
 public interface QueryResultsRegion extends GeneralDataRegion {
+
+	long nextTimestamp();
 }
Index: StandardQueryCache.java
===================================================================
--- StandardQueryCache.java	(revision 15019)
+++ StandardQueryCache.java	(working copy)
@@ -88,7 +88,7 @@
 			return false;
 		}
 		else {
-			Long ts = new Long( session.getTimestamp() );
+			Long ts = new Long( cacheRegion.nextTimestamp() );
 
 			if ( log.isDebugEnabled() ) {
 				log.debug( "caching query results in region: " + cacheRegion.getName() + "; timestamp=" + ts );
