Index: C:/Documents and Settings/Alex Burgel/workspace/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java
===================================================================
--- C:/Documents and Settings/Alex Burgel/workspace/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java	(revision 9626)
+++ C:/Documents and Settings/Alex Burgel/workspace/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java	(working copy)
@@ -3453,6 +3453,9 @@
 
 					for ( int i = 0; i < naturalIdPropertyCount; i++ ) {
 						snapshot[i] = extractionTypes[i].hydrate( rs, getPropertyAliases( "", naturalIdPropertyIndexes[i] ), session, null );
+						if (extractionTypes[i].isEntityType()) {
+							snapshot[i] = extractionTypes[i].resolve(snapshot[i], session, null);
+						}
 					}
 					return snapshot;
 				}
