--- ./Escritorio/Ejb3TypeDeclaration.ftl	2009-02-20 12:42:01.338510800 -0600
+++ /cygdrive/c/workspace/persistencia/pojo/Ejb3TypeDeclaration.ftl	2009-02-20 10:52:52.125385000 -0600
@@ -1,13 +1,52 @@
 <#if ejb3?if_exists>
 <#if pojo.isComponent()>
 @${pojo.importType("javax.persistence.Embeddable")}
 <#else>
 @${pojo.importType("javax.persistence.Entity")}
+<#switch clazz.class.simpleName>
+  <#case "UnionSubclass">
+   <#if clazz.isInherited()>
+@${pojo.importType("javax.persistence.Inheritance")}(strategy = ${pojo.importType("javax.persistence.InheritanceType")}.TABLE_PER_CLASS)
+  </#if>
+  <#break>
+  <#case "SingleTableSubclass">
+   <#if !clazz.isInherited()>
+ @${pojo.importType("javax.persistence.Inheritance")}(strategy=${pojo.importType("javax.persistence.InheritanceType")}.SINGLE_TABLE)
+      <#if clazz.discriminator.isSimpleValue()>
+         <#list clazz.discriminator.columnIterator as discriminatorColumn>
+@${pojo.importType("javax.persistence.DiscriminatorColumn")}(
+    name="${discriminatorColumn.name}"
+)  
+         </#list>  
+      </#if>
+   </#if>
+@${pojo.importType("javax.persistence.DiscriminatorValue")}("${clazz.discriminatorValue}")   
+  <#break>
+  <#case "JoinedSubclass">
+@${pojo.importType("javax.persistence.Inheritance")}(strategy=${pojo.importType("javax.persistence.InheritanceType")}.JOINED)  
+   <#if clazz.discriminator.isSimpleValue()>
+      <#list clazz.discriminator.columnIterator as joinColumn>
+@${pojo.importType("javax.persistence.PrimaryKeyJoinColumn")}(
+    name="${joinColumn.name}"
+)
+     </#list>     
+<#else>
+@${pojo.importType("javax.persistence.PrimaryKeyJoinColumns")}({
+      <#list clazz.discriminator.columnIterator as joinColumn>
+@${pojo.importType("javax.persistence.PrimaryKeyJoinColumn")}(
+    name="${joinColumn.name}"
+)<#if joinColumn_has_next>,</#if>
+     </#list>     
+
+})
+</#if> 
+  <#break>
+</#switch>  
 @${pojo.importType("javax.persistence.Table")}(name="${clazz.table.name}"
 <#if clazz.table.schema?exists>
     ,schema="${clazz.table.schema}"
 </#if><#if clazz.table.catalog?exists>
     ,catalog="${clazz.table.catalog}"
 </#if>
 <#assign uniqueConstraint=pojo.generateAnnTableUniqueConstraint()>
 <#if uniqueConstraint?has_content>
