Index: src/org/hibernate/criterion/SimpleExpression.java
===================================================================
--- src/org/hibernate/criterion/SimpleExpression.java	(revision 2)
+++ src/org/hibernate/criterion/SimpleExpression.java	Mon Nov 26 16:48:31 PST 2007
@@ -67,7 +67,7 @@
 
 	public TypedValue[] getTypedValues(Criteria criteria, CriteriaQuery criteriaQuery)
 	throws HibernateException {
-		Object icvalue = ignoreCase ? value.toString().toLowerCase() : value;
+		Object icvalue = ignoreCase && value != null ? value.toString().toLowerCase() : value;
 		return new TypedValue[] { criteriaQuery.getTypedValue(criteria, propertyName, icvalue) };
 	}
 
