Index: hibernate-core/src/main/java/org/hibernate/id/IdentityGenerator.java
===================================================================
--- hibernate-core/src/main/java/org/hibernate/id/IdentityGenerator.java	(date 1328747554000)
+++ hibernate-core/src/main/java/org/hibernate/id/IdentityGenerator.java	(revision )
@@ -96,7 +96,7 @@
 				rs = insert.getGeneratedKeys();
 				return IdentifierGeneratorHelper.getGeneratedIdentity(
 						rs,
-						persister.getIdentifierColumnNames()[0],
+						persister.getRootTableKeyColumnNames()[0],
 						persister.getIdentifierType()
 				);
 			}
@@ -144,7 +144,7 @@
 			try {
 				return IdentifierGeneratorHelper.getGeneratedIdentity(
 						rs,
-						persister.getIdentifierColumnNames()[0],
+						persister.getRootTableKeyColumnNames()[0],
 						persister.getIdentifierType()
 				);
 			}
@@ -190,7 +190,7 @@
 		        Object object) throws SQLException {
 			return IdentifierGeneratorHelper.getGeneratedIdentity(
 					rs,
-					persister.getIdentifierColumnNames()[0],
+					persister.getRootTableKeyColumnNames()[0],
 					persister.getIdentifierType()
 			);
 		}
Index: hibernate-core/src/main/java/org/hibernate/id/SequenceIdentityGenerator.java
===================================================================
--- hibernate-core/src/main/java/org/hibernate/id/SequenceIdentityGenerator.java	(date 1328747554000)
+++ hibernate-core/src/main/java/org/hibernate/id/SequenceIdentityGenerator.java	(revision )
@@ -103,7 +103,7 @@
 			insert.executeUpdate();
 			return IdentifierGeneratorHelper.getGeneratedIdentity(
 					insert.getGeneratedKeys(),
-					getPersister().getIdentifierColumnNames()[0],
+					getPersister().getRootTableKeyColumnNames()[0],
 					getPersister().getIdentifierType()
 			);
 		}
