Index: src/net/sf/hibernate/connection/ConnectionProviderFactory.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/connection/ConnectionProviderFactory.java,v
retrieving revision 1.9.2.2
diff -u -r1.9.2.2 ConnectionProviderFactory.java
--- src/net/sf/hibernate/connection/ConnectionProviderFactory.java      7 Sep 2003 00:59:51 -0000       1.9.2.2
+++ src/net/sf/hibernate/connection/ConnectionProviderFactory.java      11 Sep 2003 09:42:30 -0000
@@ -67,15 +67,15 @@
                else if ( properties.getProperty(Environment.DBCP_MAXACTIVE)!=null ) {
                        connections = new DBCPConnectionProvider();
                }
-               else if ( properties.getProperty(Environment.URL)!=null ) {
-                       connections = new DriverManagerConnectionProvider();
-               }
                else if ( 
                        properties.getProperty(Environment.PROXOOL_XML)!=null ||
                        properties.getProperty(Environment.PROXOOL_PROPERTIES)!=null || 
                        properties.getProperty(Environment.PROXOOL_EXISTING_POOL)!=null 
                ) {
                        connections = new ProxoolConnectionProvider();
+               }
+               else if ( properties.getProperty(Environment.URL)!=null ) {
+                       connections = new DriverManagerConnectionProvider();
                }
                else {
                        connections = new UserSuppliedConnectionProvider();