Index: src/java/org/hibernate/tool/hbm2x/Cfg2HbmTool.java
===================================================================
--- src/java/org/hibernate/tool/hbm2x/Cfg2HbmTool.java	(revision 20972)
+++ src/java/org/hibernate/tool/hbm2x/Cfg2HbmTool.java	(working copy)
@@ -318,6 +318,16 @@
 			return true;
 		}
 	}
+	
+	public boolean hasCascade(Property property) {
+		String cascade = property.getCascade();
+		if(cascade==null || "none".equals(cascade)) {
+			return false;
+		} else {
+			return true;
+		}
+	}
+	
 	public String getFetchMode(Property property) {
 		FetchMode fetchMode = property.getValue().getFetchMode();
 		return (fetchMode== null) ? null : fetchMode.toString().toLowerCase();
Index: src/templates/hbm/list.hbm.ftl
===================================================================
--- src/templates/hbm/list.hbm.ftl	(revision 20972)
+++ src/templates/hbm/list.hbm.ftl	(working copy)
@@ -1,28 +0,0 @@
-<#assign value = property.value>
-<#assign keyValue = value.getKey()>
-<#assign elementValue = value.getElement()>
-<#assign indexValue = value.getIndex()>
-<#assign elementTag = c2h.getCollectionElementTag(property)>
-
-	<list name="${property.name}" inverse="${value.inverse?string}" 
-	<#include "collection-tableattr.hbm.ftl"> 
-	lazy="${c2h.getCollectionLazy(value)}"
-	<#if property.cascade != "none">
-        cascade="${property.cascade}"
-	</#if>
-	<#if !property.basicPropertyAccessor>
-        access="${property.propertyAccessorName}"
-	</#if>	
-	<#if c2h.hasFetchMode(property)> fetch="${c2h.getFetchMode(property)}"</#if>>
-		<#assign metaattributable=property>
-		<#include "meta.hbm.ftl">
-		<#include "key.hbm.ftl">		
-    		<list-index>
-    			<#foreach column in indexValue.columnIterator>
-    			<#include "column.hbm.ftl">
-			</#foreach>  
-    		</list-index>
-    		<#include "${elementTag}-element.hbm.ftl">
-	</list>
-
-
Index: src/templates/hbm/map.hbm.ftl
===================================================================
--- src/templates/hbm/map.hbm.ftl	(revision 20972)
+++ src/templates/hbm/map.hbm.ftl	(working copy)
@@ -1,34 +0,0 @@
-<#assign value = property.value>
-<#assign keyValue = value.getKey()>
-<#assign elementValue = value.getElement()>
-<#assign indexValue = value.getIndex()>
-<#assign elementTag = c2h.getCollectionElementTag(property)>
-
-	<map name="${property.name}"
-	<#include "collection-tableattr.hbm.ftl">
-	lazy="${c2h.getCollectionLazy(value)}"
-	<#if property.cascade != "none">
-        cascade="${property.cascade}"
-	</#if>
-	<#if !property.basicPropertyAccessor>
-        access="${property.propertyAccessorName}"
-	</#if>
-	<#if c2h.hasFetchMode(property)> fetch="${c2h.getFetchMode(property)}"</#if>>
-		<#assign metaattributable=property><#include "meta.hbm.ftl">
-		<#-- TODO table attributes-->
-    		<#include "key.hbm.ftl">
-    		<#if c2h.isManyToOne(indexValue)>
-    		<map-key-many-to-many class="${indexValue.getReferencedEntityName()}">
-    			<#foreach column in indexValue.columnIterator>
-    			<#include "column.hbm.ftl">
-			</#foreach>  
-    		</map-key-many-to-many>
-    		<#else>
-    		<map-key type="${indexValue.typeName}">
-    			<#foreach column in indexValue.columnIterator>
-    			<#include "column.hbm.ftl">
-			</#foreach> 
-		</map-key>
-    		</#if>
-    		<#include "${elementTag}-element.hbm.ftl">
-	</map>
\ No newline at end of file
