Index: src/main/java/org/hibernate/cfg/annotations/TableBinder.java
===================================================================
--- src/main/java/org/hibernate/cfg/annotations/TableBinder.java	(revision 18156)
+++ src/main/java/org/hibernate/cfg/annotations/TableBinder.java	(working copy)
@@ -431,7 +431,7 @@
 						for (Ejb3JoinColumn joinCol : columns) {
 							String referencedColumn = joinCol.getReferencedColumn();
 							referencedColumn = mappings.getPhysicalColumnName( referencedColumn, table );
-							if ( referencedColumn.equals( col.getName() ) ) {
+							if ( referencedColumn.equals( col.getQuotedName() ) ) {
 								//proper join column
 								if ( joinCol.isNameDeferred() ) {
 									joinCol.linkValueUsingDefaultColumnNaming(
