Index: src/main/java/org/hibernate/cfg/Configuration.java
===================================================================
--- src/main/java/org/hibernate/cfg/Configuration.java	(revision 18156)
+++ src/main/java/org/hibernate/cfg/Configuration.java	(working copy)
@@ -2711,7 +2711,7 @@
 					finalName = ( String ) binding.logicalToPhysical.get( logicalName );
 				}
 				String key = buildTableNameKey(
-						currentTable.getSchema(), currentTable.getCatalog(), currentTable.getName()
+						currentTable.getQuotedSchema(), currentTable.getCatalog(), currentTable.getQuotedName()
 				);
 				TableDescription description = ( TableDescription ) tableNameBinding.get( key );
 				if ( description != null ) {
