Index: src/main/java/org/hibernate/cache/RegionFactory.java
===================================================================
--- src/main/java/org/hibernate/cache/RegionFactory.java	(revision 15754)
+++ src/main/java/org/hibernate/cache/RegionFactory.java	(working copy)
@@ -31,12 +31,11 @@
 /**
  * Contract for building second level cache regions.
  * <p/>
- * Implementors should define a constructor in one of two forms:<ul>
+ * Implementors should define a constructor of the form:<ul>
  * <li>MyRegionFactoryImpl({@link java.util.Properties})</li>
- * <li>MyRegionFactoryImpl()</li>
  * </ul>
- * Use the first when we need to read config properties prior to
- * {@link #start} being called.  For an example, have a look at
+ * This enables us to read config properties prior to {@link #start} 
+ * being called.  For an example, have a look at
  * {@link org.hibernate.cache.impl.bridge.RegionFactoryCacheProviderBridge}
  * where we need the properties in order to determine which legacy 
  * {@link CacheProvider} to use so that we can answer the
