Index: src/org/hibernate/hql/ast/exec/AbstractStatementExecutor.java
===================================================================
--- src/org/hibernate/hql/ast/exec/AbstractStatementExecutor.java	(revision 10846)
+++ src/org/hibernate/hql/ast/exec/AbstractStatementExecutor.java	(working copy)
@@ -130,7 +130,7 @@
 				}
 			}
 		};
-		if ( getFactory().getSettings().isDataDefinitionImplicitCommit() || getFactory().getDialect().performTemporaryTableDDLInIsolation() ) {
+		if ( getFactory().getDialect().performTemporaryTableDDLInIsolation() ) {
 			if ( getFactory().getSettings().isDataDefinitionInTransactionSupported() ) {
 				Isolater.doIsolatedWork( work, session );
 			}
@@ -168,7 +168,7 @@
 					}
 				}
 			};
-			if ( getFactory().getSettings().isDataDefinitionImplicitCommit() || getFactory().getDialect().performTemporaryTableDDLInIsolation() ) {
+			if ( getFactory().getDialect().performTemporaryTableDDLInIsolation() ) {
 				if ( getFactory().getSettings().isDataDefinitionInTransactionSupported() ) {
 					Isolater.doIsolatedWork( work, session );
 				}
