Index: core/src/main/java/org/hibernate/event/def/AbstractFlushingEventListener.java
===================================================================
--- core/src/main/java/org/hibernate/event/def/AbstractFlushingEventListener.java	(revision 20694)
+++ core/src/main/java/org/hibernate/event/def/AbstractFlushingEventListener.java	(working copy)
@@ -321,8 +321,7 @@
 			session.getActionQueue().executeActions();
 		}
 		catch (HibernateException he) {
-			log.error("Could not synchronize database state with session", he);
-			throw he;
+			throw new HibernateException("Could not synchronize database state with session", he);
 		}
 		finally {
 			session.getJDBCContext().getConnectionManager().flushEnding();
