Index: Oracle9iDialect.java
===================================================================
--- Oracle9iDialect.java	(revision 20346)
+++ Oracle9iDialect.java	(working copy)
@@ -96,7 +96,7 @@
 	}
 
 	public String getCurrentTimestampSelectString() {
-		return "select systimestamp from dual";
+		return "select " + getCurrentTimestampSQLFunctionName() + " from dual";
 	}
 
 	public String getCurrentTimestampSQLFunctionName() {
@@ -127,7 +127,7 @@
 		return getWriteLockString( timeout );
 	}
 	/**
-	 * HHH-4907, I don't know if oracle 8 supports this syntax, so I'd think it is better add this 
+	 * HHH-4907, I don't know if oracle 8 supports this syntax, so I'd think it is better add this
 	 * method here. Reopen this issue if you found/know 8 supports it.
 	 */
 	public boolean supportsRowValueConstructorSyntaxInInList() {
@@ -136,5 +136,5 @@
 
 	public boolean supportsTupleDistinctCounts() {
 		return false;
-	}	
+	}
 }
