diff --git a/hibernate-core/src/main/java/org/hibernate/jdbc/util/BasicFormatterImpl.java b/hibernate-core/src/main/java/org/hibernate/jdbc/util/BasicFormatterImpl.java
index 68f6a8d..a3cb815 100755
--- a/hibernate-core/src/main/java/org/hibernate/jdbc/util/BasicFormatterImpl.java
+++ b/hibernate-core/src/main/java/org/hibernate/jdbc/util/BasicFormatterImpl.java
@@ -371,6 +371,7 @@ public class BasicFormatterImpl implements Formatter {
 		}
 
 		private static boolean isFunctionName(String tok) {
+			if(tok == null || "".equals(tok)) return false;
 			final char begin = tok.charAt( 0 );
 			final boolean isIdentifier = Character.isJavaIdentifierStart( begin ) || '"' == begin;
 			return isIdentifier &&
@@ -394,4 +395,4 @@ public class BasicFormatterImpl implements Formatter {
 		}
 	}
 
-}
\ No newline at end of file
+}
diff --git a/hibernate-core/src/test/java/org/hibernate/jdbc/util/BasicFormatterTest.java b/hibernate-core/src/test/java/org/hibernate/jdbc/util/BasicFormatterTest.java
index 15f86ec..3d1dff4 100644
--- a/hibernate-core/src/test/java/org/hibernate/jdbc/util/BasicFormatterTest.java
+++ b/hibernate-core/src/test/java/org/hibernate/jdbc/util/BasicFormatterTest.java
@@ -49,6 +49,9 @@ public class BasicFormatterTest extends TestCase {
 		assertNoLoss(
 				"/* Here we' go! */ select case when p.age > 50 then 'old' when p.age > 18 then 'adult' else 'child' end from Person p where ( case when p.age > 50 then 'old' when p.age > 18 then 'adult' else 'child' end ) like ?"
 		);
+		assertNoLoss(
+				"(select p.name from Person p WHERE p.age > 20)  UNION (select p.name from Person p WHERE p.age > 40) ORDER BY p.age LIMIT 10 "
+		);
 	}
 
 	private void assertNoLoss(String query) {
