--- org\hibernate\dialect\IngresDialect.java.orig	Wed Oct 05 14:34:05 2005
+++ org\hibernate\dialect\IngresDialect.java	Tue Oct 04 15:54:19 2005
@@ -201,8 +201,11 @@
 	 * Add a <tt>LIMIT</tt> clause to the given SQL <tt>SELECT</tt>
 	 * @return the modified SQL
 	 */
-	public String getLimitString(String querySelect, boolean hasOffset, int limit) {
-		if (hasOffset) return getLimitString(querySelect, true);
+	@Override
+	public String getLimitString(String querySelect, int offset, int limit) {
+		if(offset > 0) {
+			throw new UnsupportedOperationException( "paged queries not supported" );
+		}
 		return new StringBuffer( querySelect.length()+16 )
 			.append(querySelect)
 			.insert(6, " first " + limit )
