Index: C:/workspaces/oss/hibernate/src/main/java/org/hibernate/mapping/Table.java
===================================================================
--- C:/workspaces/oss/hibernate/src/main/java/org/hibernate/mapping/Table.java	(revision 14204)
+++ C:/workspaces/oss/hibernate/src/main/java/org/hibernate/mapping/Table.java	(working copy)
@@ -376,11 +376,19 @@
 		if ( hasPrimaryKey() && identityColumn ) {
 			pkname = ( (Column) getPrimaryKey().getColumnIterator().next() ).getQuotedName( dialect );
 		}
-
+		
+		List columnNames = new ArrayList();
 		Iterator iter = getColumnIterator();
 		while ( iter.hasNext() ) {
 			Column col = (Column) iter.next();
-
+			
+			String quotedName = col.getQuotedName( dialect );
+			if(columnNames.contains(quotedName)) {
+				continue;
+			} else {
+				columnNames.add(quotedName);
+			}
+			
 			buf.append( col.getQuotedName( dialect ) )
 					.append( ' ' );
 
