Index: src/org/hibernate/hql/ast/HqlSqlWalker.java
===================================================================
--- src/org/hibernate/hql/ast/HqlSqlWalker.java	(revision 9761)
+++ src/org/hibernate/hql/ast/HqlSqlWalker.java	(working copy)
@@ -346,7 +346,7 @@
 			traverser.traverseDepthFirst( hqlSqlWithNode );
 			FromElement referencedFromElement = visitor.getReferencedFromElement();
 			if ( referencedFromElement != fromElement ) {
-				throw new SemanticException( "with-clause expressions did not reference from-clause element to which the with-clause was associated" );
+				log.warn( "with-clause expressions did not reference from-clause element to which the with-clause was associated" );
 			}
 			SqlGenerator sql = new SqlGenerator( getSessionFactoryHelper().getFactory() );
 			sql.whereExpr( hqlSqlWithNode.getFirstChild() );
@@ -382,7 +382,7 @@
 				FromElement fromElement = dotNode.getFromElement();
 				if ( referencedFromElement != null ) {
 					if ( fromElement != referencedFromElement ) {
-						throw new HibernateException( "with-clause referenced two different from-clause elements" );
+						log.warn( "with-clause referenced two different from-clause elements" );
 					}
 				}
 				else {
@@ -393,7 +393,7 @@
 					//      creates and renders the join fragments for inheritence
 					//      hierarchies...
 					if ( !joinAlias.equals( referencedFromElement.getTableAlias() ) ) {
-						throw new HibernateException( "with clause can only reference columns in the driving table" );
+						log.warn( "with clause can only reference columns in the driving table" );
 					}
 				}
 			}
