Index: StringUtils.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/tools/src/java/org/hibernate/tool/hbm2x/StringUtils.java,v
retrieving revision 1.1
diff -u -r1.1 StringUtils.java
--- StringUtils.java	18 Mar 2005 08:09:14 -0000	1.1
+++ StringUtils.java	15 May 2005 17:19:29 -0000
@@ -1072,6 +1072,26 @@
     }
 
     /**
+	 * Strip any of the supplied string from the start of a string but
+	 * do not strip more than maxStrip characters.
+	 * <p>
+	 * If maxStrip value is zero then no restriction is imposed.
+	 * <p>
+     * @param str  the string to remove characters from
+     * @param strip  the string to remove
+     * @param max characters to strip
+     * @return the stripped string
+	 */
+	public static String stripStart(String str, String strip, int maxStrip) {
+        if (str == null) return null;
+
+		if (maxStrip > 0 && str.length() > maxStrip)
+			return stripStart(str.substring(0,maxStrip-1), strip) + str.substring(maxStrip);
+		else
+			return stripStart(str, strip);
+	}
+	
+    /**
      * Strip any of a supplied string from the start of a String.
      * If the strip string is null, whitespace is stripped.
      * 
