Index: C:/java/workspace/HibernateExt/tools/src/java/org/hibernate/tool/ant/JPAConfigurationTask.java
===================================================================
--- C:/java/workspace/HibernateExt/tools/src/java/org/hibernate/tool/ant/JPAConfigurationTask.java	(revision 10462)
+++ C:/java/workspace/HibernateExt/tools/src/java/org/hibernate/tool/ant/JPAConfigurationTask.java	(working copy)
@@ -11,6 +11,8 @@
 import org.hibernate.util.ReflectHelper;
 
 public class JPAConfigurationTask extends ConfigurationTask {
+	
+	private String persistenceUnit;
 
 	public JPAConfigurationTask() {
 		setDescription("JPA Configuration");
@@ -23,7 +25,9 @@
 			Class clazz = ReflectHelper.classForName("org.hibernate.ejb.Ejb3Configuration", JPAConfigurationTask.class);
 			Object ejb3cfg = clazz.newInstance();
 			Method method = clazz.getMethod("configure", new Class[] { String.class, Map.class });
-			method.invoke(ejb3cfg, new Object[] { null, overrides } );
+			if ( method.invoke(ejb3cfg, new Object[] { persistenceUnit, overrides } ) == null ) {
+				throw new BuildException("Persistence unit not found: " + persistenceUnit);
+			}
 			
 			method = clazz.getMethod("getHibernateConfiguration", new Class[0]);
 			return (Configuration) method.invoke(ejb3cfg, null);
@@ -31,6 +35,9 @@
 		catch(HibernateException he) {
 			throw new BuildException(he);
 		}
+		catch(BuildException be) {
+			throw be;
+		}
 		catch(Throwable t) {
 			throw new BuildException("Problems in creating a configuration for JPA. Have you remembered to add hibernate EntityManager jars to the classpath ?",t);			
 		}
@@ -44,6 +51,14 @@
 		
 	}
 	
+	public String getPersistenceUnit() {
+		return persistenceUnit;
+	}
+	
+	public void setPersistenceUnit(String persistenceUnit) {
+		this.persistenceUnit = persistenceUnit;
+	}
+	
 	public void setConfigurationFile(File configurationFile) {
 		complain("configurationfile");
 	}
