--- Order.java.orig	2007-05-25 22:19:29.000000000 +0200
+++ Order.java	2008-01-30 13:40:12.000000000 +0100
@@ -60,6 +60,18 @@
 		return fragment.toString();
 	}
 
+	public String getPropertyName() {
+	    return propertyName;
+	}
+
+	public boolean isAscending() {
+        return ascending;
+    }
+
+	public boolean isIgnoreCase() {
+        return ignoreCase;
+    }
+
 	/**
 	 * Ascending order
 	 *
