--- BasicRenderer.java	2004-02-16 12:09:24.000000000 -0800
+++ BasicRendererNew.java	2004-02-16 12:09:58.000000000 -0800
@@ -10,7 +10,7 @@
 
 import net.sf.hibernate.util.StringHelper;
 
-public class BasicRenderer extends AbstractRenderer {
+public class BasicRendererNew extends AbstractRenderer {
 	
 	static final protected int ORDINARY = 0;
 	static final protected int BOUND = 1;
@@ -276,24 +276,43 @@
 		    	writer.println("        if ( !(other instanceof " + savedToClass + ") ) return false;");
 		    	writer.println("        " + savedToClass + " castOther = (" + savedToClass + ") other;");
 		    	writer.println("        return new EqualsBuilder()");
+		    	StringBuffer equalsIdBuffer = new StringBuffer();
+		    	boolean useInEquals = false;
 		    	for (Iterator fields = classMapping.getFields().iterator(); fields.hasNext();) {
 		    		FieldProperty field = (FieldProperty) fields.next();
 		    		if ( field.isIdentifier() ) {
-		    			writer.println("            .append(this." + field.getGetterSignature() + ", castOther." + field.getGetterSignature() + StringHelper.CLOSE_PAREN);
+						equalsIdBuffer.append("            .append(this." + field.getGetterSignature() + ", castOther." + field.getGetterSignature() + StringHelper.CLOSE_PAREN);
 		    		}
+					else if(field.getMetaAsBool("use-in-equals")) {
+						writer.println("            .append(\"" + field.getFieldName() + "\", " + field.getGetterSignature() + StringHelper.CLOSE_PAREN);
+						useInEquals = true;
+					}
 		    	}
+		    	if(!useInEquals){
+					writer.println(equalsIdBuffer.toString());
+		    	}
+
 		    	writer.println("            .isEquals();");
 		    	writer.println("    }");
 				writer.println();
 				
 		    	writer.println("    public int hashCode() {");
 		    	writer.println("        return new HashCodeBuilder()");
+				StringBuffer hashCodeIdBuffer = new StringBuffer();
+				boolean useInHashCode = false;
 		    	for (Iterator fields = classMapping.getFields().iterator(); fields.hasNext();) {
 		    		FieldProperty field = (FieldProperty) fields.next();
 		    		if ( field.isIdentifier() ) {
-		    			writer.println("            .append(" + field.getGetterSignature() + StringHelper.CLOSE_PAREN);
+						hashCodeIdBuffer.append("            .append(" + field.getGetterSignature() + StringHelper.CLOSE_PAREN);
 		    		}
+					else if(field.getMetaAsBool("use-in-equals")) {
+						writer.println("            .append(" + field.getGetterSignature() + StringHelper.CLOSE_PAREN);
+						useInHashCode = true;
+					}
 		    	}
+				if(!useInHashCode){
+					writer.println(hashCodeIdBuffer.toString());
+				}
 		    	writer.println("            .toHashCode();");
 		    	writer.println("    }");
 				writer.println();
