Index: core/src/main/java/org/hibernate/cfg/Configuration.java
===================================================================
--- core/src/main/java/org/hibernate/cfg/Configuration.java	(revision 19116)
+++ core/src/main/java/org/hibernate/cfg/Configuration.java	(working copy)
@@ -1072,9 +1072,9 @@
 
 				TableMetadata tableInfo = databaseMetadata.getTableMetadata(
 						table.getName(),
-						table.getSchema(),
-						table.getCatalog(),
-						table.isQuoted()
+						( table.getSchema() == null ) ? defaultSchema : table.getSchema(),
+						( table.getCatalog() == null ) ? defaultCatalog : table.getCatalog(),
+								table.isQuoted()
 					);
 
 				if ( dialect.hasAlterTable() ) {
