diff -u -r src.orig/org/hibernate/cfg/Configuration.java src/org/hibernate/cfg/Configuration.java
--- src.orig/org/hibernate/cfg/Configuration.java	2006-01-28 13:12:21.000000000 +0100
+++ src/org/hibernate/cfg/Configuration.java	2006-02-26 16:20:17.000000000 +0100
@@ -79,6 +79,7 @@
 import org.hibernate.mapping.UniqueKey;
 import org.hibernate.secure.JACCConfiguration;
 import org.hibernate.tool.hbm2ddl.DatabaseMetadata;
+import org.hibernate.tool.hbm2ddl.IndexMetadata;
 import org.hibernate.tool.hbm2ddl.TableMetadata;
 import org.hibernate.type.SerializationException;
 import org.hibernate.type.Type;
@@ -906,6 +907,26 @@
 					}
 				}
 
+				Iterator subIter = table.getIndexIterator();
+				while ( subIter.hasNext() ) {
+					Index index = (Index) subIter.next();
+					// Skip if index already exists
+					if (tableInfo != null) {
+						IndexMetadata meta = tableInfo.getIndexMetadata(index.getName());
+						if (meta != null) {
+							continue;
+						}
+					}
+					script.add(
+							index.sqlCreateString(
+									dialect,
+									mapping,
+									defaultCatalog,
+									defaultSchema
+								)
+						);
+				}
+
 			}
 
 			/*//broken, 'cos we don't generate these with names in SchemaExport
