Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/ToOneDelegateSessionImplementor.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/ToOneDelegateSessionImplementor.java	(revision 17249)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/ToOneDelegateSessionImplementor.java	Sun Sep 13 16:46:19 EEST 2009
@@ -31,7 +31,6 @@
 import org.hibernate.envers.reader.AuditReaderImplementor;
 
 import org.hibernate.HibernateException;
-import org.hibernate.Session;
 
 /**
  * @author Adam Warski (adam at warski dot org)
@@ -62,8 +61,7 @@
 		if (notVersionedEntityConfiguration == null) {
 			return versionsReader.find(entityClass, entityId, revision);
 		} else {
-			Session session = versionsReader.getSession();
-			return session.get(entityClass, (Serializable) entityId);
+			return delegate.immediateLoad(entityName, (Serializable) entityId);
 		}
     }
 }
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/AbstractDelegateSessionImplementor.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/AbstractDelegateSessionImplementor.java	(revision 15833)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/AbstractDelegateSessionImplementor.java	Sun Sep 13 16:46:19 EEST 2009
@@ -57,7 +57,7 @@
  * @author Adam Warski (adam at warski dot org)
  */
 public abstract class AbstractDelegateSessionImplementor implements SessionImplementor {
-    private SessionImplementor delegate;
+    protected SessionImplementor delegate;
 
     public AbstractDelegateSessionImplementor(SessionImplementor delegate) {
         this.delegate = delegate;
