Index: src/net/sf/hibernate/util/ReflectHelper.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/util/ReflectHelper.java,v
retrieving revision 1.12.2.10
diff -u -r1.12.2.10 ReflectHelper.java
--- src/net/sf/hibernate/util/ReflectHelper.java	19 Nov 2003 10:01:52 -0000	1.12.2.10
+++ src/net/sf/hibernate/util/ReflectHelper.java	25 Nov 2003 20:59:12 -0000
@@ -133,14 +133,20 @@
 			return optimizer;
 		}
 		catch (Throwable t) {
-			log.info( 
+			String message =
 				"reflection optimizer disabled for: " + 
 				clazz.getName() + 
 				", " + 
 				StringHelper.unqualify( t.getClass().getName() ) + 
 				": " + 
-				t.getMessage() 
-			);
+				t.getMessage();
+			if (t instanceof BulkBeanException) {
+				int index = ((BulkBeanException)t).getIndex();
+				if (index >= 0) {
+					message += " (was setting property " + setterNames[index] + ")";
+				}
+			}
+			log.info(message);
 			return null;
 		}
 	}
