diff --git a/hibernate-core/src/main/java/org/hibernate/cache/internal/StandardQueryCache.java b/hibernate-core/src/main/java/org/hibernate/cache/internal/StandardQueryCache.java
index 3513931..c216169 100644
--- a/hibernate-core/src/main/java/org/hibernate/cache/internal/StandardQueryCache.java
+++ b/hibernate-core/src/main/java/org/hibernate/cache/internal/StandardQueryCache.java
@@ -178,7 +178,14 @@ public class StandardQueryCache implements QueryCache {
 					cacheRegion.evict( key );
 					return null;
 				}
-				throw ex;
+				else if (ObjectNotFoundException.class.isInstance(ex)) {
+					// object must have been removed from the entity cache after
+					// reading the query cache
+					cacheRegion.evict( key );
+				}
+				else {
+					throw ex;
+				}
 			}
 		}
 		return result;
