Index: src/org/hibernate/hibernate-mapping-3.0.dtd
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/hibernate-mapping-3.0.dtd,v
retrieving revision 1.64
diff -u -r1.64 hibernate-mapping-3.0.dtd
--- src/org/hibernate/hibernate-mapping-3.0.dtd	3 Feb 2005 19:15:20 -0000	1.64
+++ src/org/hibernate/hibernate-mapping-3.0.dtd	4 Feb 2005 21:54:04 -0000
@@ -708,12 +708,14 @@
 on entities. -->
 
 <!ELEMENT map-key-many-to-many (column*)>
-	<!ATTLIST map-key-many-to-many class CDATA #REQUIRED>
+	<!ATTLIST map-key-many-to-many class CDATA #IMPLIED>
+	<!ATTLIST map-key-many-to-many entity-name CDATA #IMPLIED>
 	<!ATTLIST map-key-many-to-many column CDATA #IMPLIED>
 	<!ATTLIST map-key-many-to-many foreign-key CDATA #IMPLIED>
 
 <!ELEMENT index-many-to-many (column*)>
-	<!ATTLIST index-many-to-many class CDATA #REQUIRED>
+	<!ATTLIST index-many-to-many class CDATA #IMPLIED>
+	<!ATTLIST index-many-to-many entity-name CDATA #IMPLIED>
 	<!ATTLIST index-many-to-many column CDATA #IMPLIED>
 	<!ATTLIST index-many-to-many foreign-key CDATA #IMPLIED>
 
