Index: C:/projects/workspace/Hibernate3/src/org/hibernate/impl/FetchingScrollableResultsImpl.java
===================================================================
--- C:/projects/workspace/Hibernate3/src/org/hibernate/impl/FetchingScrollableResultsImpl.java	(revision 9974)
+++ C:/projects/workspace/Hibernate3/src/org/hibernate/impl/FetchingScrollableResultsImpl.java	(working copy)
@@ -51,6 +51,18 @@
 			currentPosition = maxPosition.intValue() + 1;
 			return false;
 		}
+		
+		if (currentPosition == 0) {
+			try {
+				if (!getResultSet().next()) return false;
+			} catch (SQLException e) {
+				throw JDBCExceptionHelper.convert(
+				        getSession().getFactory().getSQLExceptionConverter(),
+				        e,
+				        "exception calling next()"
+					);
+			}
+		}
 
 		Object row = getLoader().loadSequentialRowsForward(
 				getResultSet(),
