Index: org/hibernate/cfg/Configuration.java
===================================================================
--- org/hibernate/cfg/Configuration.java	(revision 20769)
+++ org/hibernate/cfg/Configuration.java	(working copy)
@@ -1214,8 +1214,8 @@
 
 				TableMetadata tableInfo = databaseMetadata.getTableMetadata(
 						table.getName(),
-						table.getSchema(),
-						table.getCatalog(),
+						( table.getSchema() == null ) ? defaultSchema : table.getSchema(),
+						( table.getCatalog() == null ) ? defaultCatalog : table.getCatalog(),
 						table.isQuoted()
 					);
 
