diff --git a/hibernate-search-orm/src/test/java/org/hibernate/search/test/query/facet/Car.java b/hibernate-search-orm/src/test/java/org/hibernate/search/test/query/facet/Car.java
index 9c36d9a..b79569e 100644
--- a/hibernate-search-orm/src/test/java/org/hibernate/search/test/query/facet/Car.java
+++ b/hibernate-search-orm/src/test/java/org/hibernate/search/test/query/facet/Car.java
@@ -31,6 +31,7 @@ import javax.persistence.Id;
 import org.hibernate.search.annotations.Analyze;
 import org.hibernate.search.annotations.Field;
 import org.hibernate.search.annotations.Indexed;
+import org.hibernate.search.annotations.NumericField;
 import org.hibernate.search.annotations.Store;
 
 /**
@@ -50,6 +51,7 @@ public class Car {
 	private String make;
 
 	@Field(analyze = Analyze.NO)
+	@NumericField
 	private int cubicCapacity;
 
 	private Car() {
