--- hibernate-3.2.0.CR2/src/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java	2006-05-05 23:55:32.000000000 -0500
+++ JoinedSubclassEntityPersister.java	2006-07-24 12:01:26.000000000 -0500
@@ -43,7 +43,9 @@
 	private final String[][] tableKeyColumns;
 	private final String[][] naturalOrderTableKeyColumns;
 	private final boolean[] naturalOrderCascadeDeleteEnabled;
-	
+
+	private final String rootTableName;
+
 	private final String[] spaces;
 
 	private final String[] subclassClosure;
@@ -323,6 +325,8 @@
 			k++;
 		}
 
+		rootTableName = persistentClass.getRootTable().getName();
+
 		initLockers();
 
 		initSubclassPropertyAliasesMap(persistentClass);
@@ -362,7 +366,11 @@
 	protected String getTableName(int j) {
 		return naturalOrderTableNames[j];
 	}
-	
+
+	public String getRootTableName() {
+		return rootTableName;
+	}
+
 	protected String[] getKeyColumns(int j) {
 		return naturalOrderTableKeyColumns[j];
 	}
