Index: C:/Workspaces/leica-trunk/core/src/main/java/org/hibernate/cfg/HbmBinder.java
===================================================================
--- C:/Workspaces/leica-trunk/core/src/main/java/org/hibernate/cfg/HbmBinder.java	(revision 14832)
+++ C:/Workspaces/leica-trunk/core/src/main/java/org/hibernate/cfg/HbmBinder.java	(working copy)
@@ -2909,7 +2909,14 @@
 			final String paramName = param.attributeValue( "name" );
 			final String paramType = param.attributeValue( "type" );
 			log.debug( "adding filter parameter : " + paramName + " -> " + paramType );
-			final Type heuristicType = TypeFactory.heuristicType( paramType );
+			Type heuristicType = TypeFactory.heuristicType( paramType );
+			if ( heuristicType == null ) {
+				TypeDef typeDef = mappings.getTypeDef(paramType);
+				
+				if ( typeDef != null ) {
+					heuristicType = TypeFactory.heuristicType( typeDef.getTypeClass(), typeDef.getParameters() );
+				}
+			}
 			log.debug( "parameter heuristic type : " + heuristicType );
 			paramMappings.put( paramName, heuristicType );
 		}
