Index: src/org/hibernate/event/def/AbstractFlushingEventListener.java
===================================================================
--- src/org/hibernate/event/def/AbstractFlushingEventListener.java	(revision 14204)
+++ src/org/hibernate/event/def/AbstractFlushingEventListener.java	(working copy)
@@ -297,10 +297,6 @@
 			session.getActionQueue().prepareActions();
 			session.getActionQueue().executeActions();
 		}
-		catch (HibernateException he) {
-			log.error("Could not synchronize database state with session", he);
-			throw he;
-		}
 		finally {
 			session.getJDBCContext().getConnectionManager().flushEnding();
 		}
