Index: src/main/java/org/hibernate/cfg/HbmBinder.java
===================================================================
--- src/main/java/org/hibernate/cfg/HbmBinder.java      (revision 15600)
+++ src/main/java/org/hibernate/cfg/HbmBinder.java      (working copy)
@@ -1044,7 +1044,7 @@

                        if ( value != null ) {
                                Property prop = createProperty( value, propertyName, persistentClass
-                                       .getEntityName(), subnode, mappings, inheritedMetas );
+                                       .getClassName(), subnode, mappings, inheritedMetas );
                                prop.setOptional( join.isOptional() );
                                join.addProperty( prop );
                        }
