Index: org/hibernate/cfg/HbmBinder.java
===================================================================
--- org/hibernate/cfg/HbmBinder.java	(revision 12882)
+++ org/hibernate/cfg/HbmBinder.java	(working copy)
@@ -992,7 +992,7 @@
 
 			if ( value != null ) {
 				Property prop = createProperty( value, propertyName, persistentClass
-					.getEntityName(), subnode, mappings, inheritedMetas );
+					.getClassName(), subnode, mappings, inheritedMetas );
 				prop.setOptional( join.isOptional() );
 				join.addProperty( prop );
 			}
