--- src/org/hibernate/proxy/BasicLazyInitializer.java	2005-07-22 04:32:25.000000000 -0700
+++ src/org/hibernate/proxy/BasicLazyInitializer.java.orig	2005-07-22 04:30:56.000000000 -0700
@@ -46,7 +46,7 @@
 	protected final Object invoke(Method method, Object[] args, Object proxy) throws Throwable {
 
 		String methodName = method.getName();
-		int params = method.getParameterTypes().length;
+		int params = args.length;
 
 		if ( params==0 ) {
 
