Index: java/org/hibernate/tool/hbm2x/pojo/EntityPOJOClass.java
===================================================================
--- java/org/hibernate/tool/hbm2x/pojo/EntityPOJOClass.java	(revision 15608)
+++ java/org/hibernate/tool/hbm2x/pojo/EntityPOJOClass.java	(working copy)
@@ -454,10 +454,13 @@
 		OneToOne oneToOne = (OneToOne)property.getValue();
 		if (oneToOne.isConstrained())
 			ab.addQuotedAttribute("mappedBy", getOneToOneMappedBy(cfg, oneToOne));
-		
 		StringBuffer buffer = new StringBuffer(ab.getResult());
+		buffer.append(getHibernateCascadeTypeAnnotation(property));
+		if (!oneToOne.isConstrained()){
+			AnnotationBuilder ab1 = AnnotationBuilder.createAnnotation( importType("javax.persistence.PrimaryKeyJoinColumn") );
+			buffer.append(ab1.getResult());
+		}
 		
-		buffer.append(getHibernateCascadeTypeAnnotation(property));
 		return buffer.toString();
 	}
 	
