From 6ff4ff97a3ca74f762625c12f9485eb3c7b2daa9 Mon Sep 17 00:00:00 2001
From: JohnOhara <johara@redhat.com>
Date: Fri, 7 Jun 2013 16:19:45 +0100
Subject: [PATCH] Patched EnhanceEntities ant task

---
 .../hibernate/tool/enhance/EnhancementTask.java    |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hibernate-core/src/main/java/org/hibernate/tool/enhance/EnhancementTask.java b/hibernate-core/src/main/java/org/hibernate/tool/enhance/EnhancementTask.java
index 1ef02df..ae36dd0 100644
--- a/hibernate-core/src/main/java/org/hibernate/tool/enhance/EnhancementTask.java
+++ b/hibernate-core/src/main/java/org/hibernate/tool/enhance/EnhancementTask.java
@@ -140,7 +140,7 @@ public class EnhancementTask extends Task implements EnhancementContext {
 
 	private boolean shouldInclude(CtClass ctClass) {
 		// we currently only handle entity enhancement
-		return ! ctClass.hasAnnotation( Entity.class );
+		return ctClass.hasAnnotation( Entity.class );
 	}
 
 
-- 
1.7.7.6

