Index: ./src/org/hibernate/LazyInitializationException.java
===================================================================
--- ./src/org/hibernate/LazyInitializationException.java	(revision 14265)
+++ ./src/org/hibernate/LazyInitializationException.java	(working copy)
@@ -1,6 +1,7 @@
 //$Id$
 package org.hibernate;
 
+import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
 /**
@@ -14,9 +15,12 @@
  */
 public class LazyInitializationException extends HibernateException {
 
+	private static final Log LOG = LogFactory.getLog(LazyInitializationException.class);
+
 	public LazyInitializationException(String msg) {
 		super(msg);
-		LogFactory.getLog(LazyInitializationException.class).error(msg, this);
+		if (LOG.isErrorEnabled())
+			LOG.error(msg, this);
 	}
 
 }
