Index: metadata/src/java/org/hibernate/cfg/AnnotationBinder.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/metadata/src/java/org/hibernate/cfg/AnnotationBinder.java,v
retrieving revision 1.74
diff -u -r1.74 AnnotationBinder.java
--- metadata/src/java/org/hibernate/cfg/AnnotationBinder.java	13 Apr 2005 17:46:45 -0000	1.74
+++ metadata/src/java/org/hibernate/cfg/AnnotationBinder.java	16 Apr 2005 18:48:11 -0000
@@ -552,7 +552,9 @@
 			int index = 0;
 			while( index < methods.length) {
 				currentElt = (AnnotatedElement) methods[index];
-				addAnnotatedElement(currentElt, annElts, inferredDatas);
+				if (!methods[index].isBridge()) {
+					addAnnotatedElement(currentElt, annElts, inferredDatas);
+                }
 				index++;
 			}
 		}
