Index: hql-sql.g
===================================================================
--- hql-sql.g	(revision 9734)
+++ hql-sql.g	(working copy)
@@ -575,7 +575,7 @@
 	;
 
 functionCall
-	: #(METHOD_CALL  {inFunctionCall=true;} pathAsIdent ( #(EXPR_LIST (expr)* ) )? )
+	: #(METHOD_CALL  {inFunctionCall=true;} pathAsIdent ( #(EXPR_LIST (exprOrSubquery)* ) )? )
 		{ processFunction(#functionCall,inSelect); } {inFunctionCall=false;}
 	| #(AGGREGATE aggregateExpr )
 	;
