Index: QueryTest.java
===================================================================
--- QueryTest.java	(revision 19344)
+++ QueryTest.java	(working copy)
@@ -1,4 +1,4 @@
-//$Id$
+﻿//$Id$
 package org.hibernate.ejb.test.query;
 
 import java.util.ArrayList;
@@ -65,7 +65,7 @@
 		em.getTransaction().commit();
 
 		em.getTransaction().begin();
-		Query q = em.createQuery( "select item from Item item where item.name in (:names)" );
+		Query q = em.createQuery( "select item from Item item where item.name in :names" );
 		//test hint in value and string
 		q.setHint( "org.hibernate.fetchSize", 10 );
 		q.setHint( "org.hibernate.fetchSize", "10" );
@@ -77,7 +77,7 @@
 		assertNotNull( result );
 		assertEquals( 2, result.size() );
 
-		q = em.createQuery( "select item from Item item where item.name in ( ?1 )" );
+		q = em.createQuery( "select item from Item item where item.name in ?1" );
 		params = new ArrayList();
 		params.add( item.getName() );
 		params.add( item2.getName() );
@@ -201,7 +201,7 @@
 		em.persist( w );
 		em.getTransaction().commit();
 		em.getTransaction().begin();
-		Query query = em.createQuery( "select w from " + Wallet.class.getName() + " w where w.brand in (?1)" );
+		Query query = em.createQuery( "select w from " + Wallet.class.getName() + " w where w.brand in ?1" );
 		List brands = new ArrayList();
 		brands.add( "Lacoste" );
 		query.setParameter( 1, brands );
