Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/CollectionProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/CollectionProxy.java	(revision 17676)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/CollectionProxy.java	(working copy)
@@ -23,6 +23,7 @@
  */
 package org.hibernate.envers.entities.mapper.relation.lazy.proxy;
 
+import java.io.Serializable;
 import java.util.Collection;
 import java.util.Iterator;
 
@@ -31,8 +32,10 @@
 /**
  * @author Adam Warski (adam at warski dot org)
  */
-public abstract class CollectionProxy<U, T extends Collection<U>> implements Collection<U> {
-    private org.hibernate.envers.entities.mapper.relation.lazy.initializor.Initializor<T> initializor;
+public abstract class CollectionProxy<U, T extends Collection<U>> implements Collection<U>, Serializable {
+
+	private static final long serialVersionUID = 8698249863871832402L;
+	private transient org.hibernate.envers.entities.mapper.relation.lazy.initializor.Initializor<T> initializor;
     protected T delegate;
 
     protected CollectionProxy() {
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/ListProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/ListProxy.java	(revision 17676)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/ListProxy.java	(working copy)
@@ -33,7 +33,7 @@
 /**
  * @author Adam Warski (adam at warski dot org)
  */
-public class ListProxy<U> extends CollectionProxy<U, List<U>> implements List<U>, Serializable {
+public class ListProxy<U> extends CollectionProxy<U, List<U>> implements List<U> {
     private static final long serialVersionUID = -5479232938279790987L;
 
     public ListProxy() {
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/MapProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/MapProxy.java	(revision 17676)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/MapProxy.java	(working copy)
@@ -33,7 +33,7 @@
 /**
  * @author Adam Warski (adam at warski dot org)
  */
-public class MapProxy<K, V> implements Map<K, V>, Serializable {
+public class MapProxy<K, V> implements Map<K, V> {
     private static final long serialVersionUID = 8418037541773074646L;
 
     private Initializor<Map<K, V>> initializor;
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SetProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SetProxy.java	(revision 17676)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SetProxy.java	(working copy)
@@ -29,7 +29,7 @@
 /**
  * @author Adam Warski (adam at warski dot org)
  */
-public class SetProxy<U> extends CollectionProxy<U, Set<U>> implements Set<U>, Serializable {
+public class SetProxy<U> extends CollectionProxy<U, Set<U>> implements Set<U> {
     private static final long serialVersionUID = 131464133074137701L;
 
     public SetProxy() {
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedMapProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedMapProxy.java	(revision 17676)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedMapProxy.java	(working copy)
@@ -35,7 +35,7 @@
 /**
  * @author Adam Warski (adam at warski dot org)
  */
-public class SortedMapProxy<K, V> implements SortedMap<K, V>, Serializable {
+public class SortedMapProxy<K, V> implements SortedMap<K, V> {
     private static final long serialVersionUID = 2645817952901452375L;
 
     private Initializor<SortedMap<K, V>> initializor;
Index: src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedSetProxy.java
===================================================================
--- src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedSetProxy.java	(revision 17676)
+++ src/main/java/org/hibernate/envers/entities/mapper/relation/lazy/proxy/SortedSetProxy.java	(working copy)
@@ -30,7 +30,7 @@
 /**
  * @author Adam Warski (adam at warski dot org)
  */
-public class SortedSetProxy<U> extends CollectionProxy<U, SortedSet<U>> implements SortedSet<U>, Serializable {
+public class SortedSetProxy<U> extends CollectionProxy<U, SortedSet<U>> implements SortedSet<U> {
     private static final long serialVersionUID = 2092884107178125905L;
 
     public SortedSetProxy() {
