Index: src/main/java/org/hibernate/mapping/SimpleValue.java
===================================================================
--- src/main/java/org/hibernate/mapping/SimpleValue.java	(revision 15705)
+++ src/main/java/org/hibernate/mapping/SimpleValue.java	(working copy)

@@ -206,11 +206,11 @@
 	public String getIdentifierGeneratorStrategy() {
 		return identifierGeneratorStrategy;
 	}
 	
 	public boolean isIdentityColumn(IdentifierGeneratorFactory identifierGeneratorFactory, Dialect dialect) {
 		identifierGeneratorFactory.setDialect( dialect );
-		return identifierGeneratorFactory.getIdentifierGeneratorClass( identifierGeneratorStrategy )
-				.equals( IdentityGenerator.class );
+		return IdentityGenerator.class.isAssignableFrom(
+				identifierGeneratorFactory.getIdentifierGeneratorClass( identifierGeneratorStrategy ));
 	}
 
 	/**
