Index: ForeignKeys.java
===================================================================
--- ForeignKeys.java	(revision 9528)
+++ ForeignKeys.java	(working copy)
@@ -215,7 +215,7 @@
 				if ( isTransient(entityName, object, Boolean.FALSE, session) ) {
 					throw new TransientObjectException(
 							"object references an unsaved transient instance - save the transient instance before flushing: " +
-							entityName == null ? session.guessEntityName( object ) : entityName
+							(entityName == null ? session.guessEntityName( object ) : entityName)
 					);
 				}
 				id = session.getEntityPersister(entityName, object)
