Index: tools/src/java/org/hibernate/cfg/reveng/dialect/OracleMetaDataDialect.java
===================================================================
--- tools/src/java/org/hibernate/cfg/reveng/dialect/OracleMetaDataDialect.java	(revision 11404)
+++ tools/src/java/org/hibernate/cfg/reveng/dialect/OracleMetaDataDialect.java	Thu Jun 02 09:35:42 BST 2011
@@ -189,7 +189,7 @@
 			+ "left join all_constraints f on (p.owner = f.r_owner and p.constraint_name = f.r_constraint_name) "
 			+ "left join all_cons_columns fc on "
 			+ "(fc.owner = f.owner and fc.constraint_name = f.constraint_name and fc.table_name = f.table_name and"
-			+ " fc.position = pc.position) where f.constraint_type = 'R' AND  p.constraint_type = 'P' ";
+			+ " fc.position = pc.position) where f.constraint_type = 'R' AND  p.constraint_type in ('P','U') ";
 
 	private static final String SQL_FK_ORDER = " order by f.table_name, f.constraint_name, fc.position ";
 
