diff -ur hibernate-annotations-3.3.1.GA.orig/src/org/hibernate/cfg/annotations/MapBinder.java hibernate-annotations-3.3.1.GA/src/org/hibernate/cfg/annotations/MapBinder.java
--- hibernate-annotations-3.3.1.GA.orig/src/org/hibernate/cfg/annotations/MapBinder.java	2008-03-14 16:44:45.000000000 -0400
+++ hibernate-annotations-3.3.1.GA/src/org/hibernate/cfg/annotations/MapBinder.java	2008-07-11 15:03:40.000000000 -0400
@@ -233,11 +233,13 @@
 						column.setTable( mapValue.getCollectionTable() );
 					}
 					elementBinder.setColumns( elementColumns );
-					//do not call setType as it extract the type from @Type
-					//the algorithm generally does not apply for map key anyway
+					//use the @Type given in @MapKey rather than the one on the property itself if possible
+					//because the algorithm in setType may not always apply for map key generally speaking
 					MapKey mapKeyAnn = property.getAnnotation( org.hibernate.annotations.MapKey.class );
 					if (mapKeyAnn != null && ! BinderHelper.isDefault( mapKeyAnn.type().type() ) ) {
 						elementBinder.setExplicitType( mapKeyAnn.type() );
+					} else {
+						elementBinder.setType( property, elementClass );
 					}
 					mapValue.setIndex( elementBinder.make() );
 				}
Only in hibernate-annotations-3.3.1.GA/src/org/hibernate/cfg/annotations: MapBinder.java~
