Index: src/org/hibernate/tuple/EntityModeToTuplizerMapping.java
===================================================================
--- src/org/hibernate/tuple/EntityModeToTuplizerMapping.java	(revision 14109)
+++ src/org/hibernate/tuple/EntityModeToTuplizerMapping.java	(working copy)
@@ -1,12 +1,11 @@
 package org.hibernate.tuple;
 
-import org.apache.commons.collections.SequencedHashMap;
-
 import org.hibernate.EntityMode;
 import org.hibernate.HibernateException;
 
+import EDU.oswego.cs.dl.util.concurrent.ConcurrentReaderHashMap;
+
 import java.util.Map;
-import java.util.Collections;
 import java.util.Iterator;
 import java.io.Serializable;
 
@@ -18,7 +17,7 @@
 public abstract class EntityModeToTuplizerMapping implements Serializable {
 
 	// map of EntityMode -> Tuplizer
-	private final Map tuplizers = Collections.synchronizedMap( new SequencedHashMap() );
+	private final Map tuplizers = new ConcurrentReaderHashMap();
 
 	protected void addTuplizer(EntityMode entityMode, Tuplizer tuplizer) {
 		tuplizers.put( entityMode, tuplizer );
