Index: EntityModeToTuplizerMapping.java
===================================================================
--- EntityModeToTuplizerMapping.java	(revision 14201)
+++ EntityModeToTuplizerMapping.java	(working copy)
@@ -18,7 +18,7 @@
 public abstract class EntityModeToTuplizerMapping implements Serializable {
 
 	// map of EntityMode -> Tuplizer
-	private final Map tuplizers = Collections.synchronizedMap( new SequencedHashMap() );
+	private final Map tuplizers = new org.hibernate.util.FastHashMap();
 
 	protected void addTuplizer(EntityMode entityMode, Tuplizer tuplizer) {
 		tuplizers.put( entityMode, tuplizer );
