Index: src/main/java/org/hibernate/ejb/metamodel/AttributeFactory.java
===================================================================
--- src/main/java/org/hibernate/ejb/metamodel/AttributeFactory.java	(revision 19571)
+++ src/main/java/org/hibernate/ejb/metamodel/AttributeFactory.java	(working copy)
@@ -834,7 +834,7 @@
 			}
 			else {
 				throw new AssertionFailure("Fail to process type argument in a generic declaration. Type: "
-						+ type.getClass() );
+						+ type.getClass() + " Member: " + getMember());
 			}
 			return javaType;
 		}
