Index: src/main/java/org/hibernate/mapping/Table.java
===================================================================
--- src/main/java/org/hibernate/mapping/Table.java	(revision 19478)
+++ src/main/java/org/hibernate/mapping/Table.java	(working copy)
@@ -277,7 +277,7 @@
 			}
 			else {
 				final boolean typesMatch = col.getSqlType( dialect, mapping ).toLowerCase()
-						.startsWith( columnInfo.getTypeName().toLowerCase() )
+						.equals( columnInfo.getTypeName().toLowerCase() )
 						|| columnInfo.getTypeCode() == col.getSqlTypeCode( mapping );
 				if ( !typesMatch ) {
 					throw new HibernateException(
