Index: DeleteOneToOneOrphansTest.java
===================================================================
--- DeleteOneToOneOrphansTest.java	(revision 20876)
+++ DeleteOneToOneOrphansTest.java	(working copy)
@@ -90,43 +90,17 @@
 		cleanupData();
 	}
 
-	public void testOrphanedWhileDetachedFailureExpected() {
-		createData();
-
+	public void testOrphanedWhileDetachedFailureExpected() {			
 		Session session = openSession();
 		session.beginTransaction();
-		List results = session.createQuery( "from EmployeeInfo" ).list();
-		assertEquals( 1, results.size() );
-		results = session.createQuery( "from Employee" ).list();
-		assertEquals( 1, results.size() );
-		Employee emp = ( Employee ) results.get( 0 );
-		assertNotNull( emp.getInfo() );
-
-		//only fails if the object is detached
-		session.getTransaction().commit();
-		session.close();
-		session = openSession();
-		session.beginTransaction();
-
-		emp.setInfo( null );
-
-		//save using the new session (this used to work prior to 3.5.x)
+		
+		Employee emp = new Employee();
+		session.save(emp);
+		session.evict(emp);
 		session.saveOrUpdate(emp);
 
 		session.getTransaction().commit();
 		session.close();
-
-		session = openSession();
-		session.beginTransaction();
-		emp = ( Employee ) session.get( Employee.class, emp.getId() );
-		assertNull( emp.getInfo() );
-		results = session.createQuery( "from EmployeeInfo" ).list();
-		assertEquals( 0, results.size() );
-		results = session.createQuery( "from Employee" ).list();
-		assertEquals( 1, results.size() );
-		session.getTransaction().commit();
-		session.close();
-
 		cleanupData();
 	}
 }
\ No newline at end of file
