Index: testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/reversed/unidirectional/DeleteOneToOneOrphansTest.java
===================================================================
--- testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/reversed/unidirectional/DeleteOneToOneOrphansTest.java	(revision 19838)
+++ testsuite/src/test/java/org/hibernate/test/orphan/one2one/fk/reversed/unidirectional/DeleteOneToOneOrphansTest.java	(working copy)
@@ -72,7 +72,18 @@
 		assertEquals( 1, results.size() );
 		Employee emp = ( Employee ) results.get( 0 );
 		assertNotNull( emp.getInfo() );
+		
+		//only fails if the object is detached
+		session.getTransaction().commit();
+		session.close();		
+		session = openSession();
+		session.beginTransaction();
+		
 		emp.setInfo( null );
+		
+		//save using the new session (this used to work prior to 3.5.x)
+		session.saveOrUpdate(emp);
+		
 		session.getTransaction().commit();
 		session.close();
 
