Index: DeleteOneToOneOrphansTest.java
===================================================================
--- DeleteOneToOneOrphansTest.java	(revision 19969)
+++ DeleteOneToOneOrphansTest.java	(working copy)
@@ -89,4 +89,34 @@
 
 		cleanupData();
 	}
+	
+	public void testOrphanedAfterMerge() {
+		createData();
+
+		Session session = openSession();
+		session.beginTransaction();
+		
+		//the object has to be new, if we use the Employee saved by createData(); the test will pass
+		Employee emp = new Employee();
+		emp.setInfo( new EmployeeInfo( emp ) );
+				
+		emp = (Employee) session.merge( emp );
+		//remove cascaded entity from relation after the merge
+		emp.setInfo( null );
+		
+		//the commit should delete EmployeeInfo from the database
+		session.getTransaction().commit();
+		session.close();		
+		
+		session = openSession();
+		session.beginTransaction();
+
+		emp = ( Employee ) session.get( Employee.class, emp.getId() );
+		assertNull( emp.getInfo() );
+		
+		session.getTransaction().commit();
+		session.close();
+
+		cleanupData();
+	}	
 }
\ No newline at end of file
