Index: U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3MergeEventListener.java
===================================================================
--- U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3MergeEventListener.java	(revision 14771)
+++ U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3MergeEventListener.java	(working copy)
@@ -5,7 +5,7 @@
 
 import org.hibernate.event.EventSource;
 import org.hibernate.event.def.DefaultMergeEventListener;
-import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.event.def.IDAttachResult;
 
 /**
  * Overrides the LifeCycle OnSave call to call the PrePersist operation
@@ -43,4 +43,12 @@
 		return super.saveWithGeneratedId( entity, entityName, anything, source,
 				requiresImmediateIdAccess );
 	}
+
+	@Override
+	protected Serializable saveWithAttachedID(IDAttachResult attachResult, Object entity, String entityName, Object anything, EventSource source)
+	{
+		callbackHandler.preCreate( entity );
+		return super.saveWithAttachedID(attachResult, entity, entityName,
+				anything, source);
+	}
 }
Index: U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3PersistEventListener.java
===================================================================
--- U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3PersistEventListener.java	(revision 14771)
+++ U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3PersistEventListener.java	(working copy)
@@ -5,6 +5,7 @@
 
 import org.hibernate.event.EventSource;
 import org.hibernate.event.def.DefaultPersistEventListener;
+import org.hibernate.event.def.IDAttachResult;
 import org.hibernate.engine.CascadingAction;
 import org.hibernate.engine.EJB3CascadingAction;
 import org.hibernate.engine.EJB3CascadeStyle;
@@ -51,6 +52,14 @@
 	}
 
 	@Override
+	protected Serializable saveWithAttachedID(IDAttachResult attachResult, Object entity, String entityName, Object anything, EventSource source)
+	{
+		callbackHandler.preCreate( entity );
+		return super.saveWithAttachedID(attachResult, entity, entityName, anything,
+				source);
+	}
+
+	@Override
 	protected CascadingAction getCascadeAction() {
 		return EJB3CascadingAction.PERSIST_SKIPLAZY;
 	}
Index: U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3SaveEventListener.java
===================================================================
--- U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3SaveEventListener.java	(revision 14771)
+++ U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3SaveEventListener.java	(working copy)
@@ -5,6 +5,7 @@
 
 import org.hibernate.event.EventSource;
 import org.hibernate.event.def.DefaultSaveEventListener;
+import org.hibernate.event.def.IDAttachResult;
 import org.hibernate.persister.entity.EntityPersister;
 
 /**
@@ -43,4 +44,13 @@
 		return super.saveWithGeneratedId( entity, entityName, anything, source,
 				requiresImmediateIdAccess );
 	}
+
+	@Override
+	protected Serializable saveWithAttachedID(IDAttachResult attachResult, Object entity, String entityName, Object anything, EventSource source)
+	{
+		callbackHandler.preCreate( entity );
+		return super.saveWithAttachedID(attachResult, entity, entityName, anything,
+				source);
+	}
+
 }
Index: U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java
===================================================================
--- U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java	(revision 14771)
+++ U:/workspaces/atn-Standart/HibernateExt/entitymanager/src/java/org/hibernate/ejb/event/EJB3SaveOrUpdateEventListener.java	(working copy)
@@ -5,7 +5,7 @@
 
 import org.hibernate.event.EventSource;
 import org.hibernate.event.def.DefaultSaveOrUpdateEventListener;
-import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.event.def.IDAttachResult;
 
 /**
  * Overrides the LifeCycle OnSave call to call the PrePersist operation
@@ -43,4 +43,13 @@
 		return super.saveWithGeneratedId( entity, entityName, anything, source,
 				requiresImmediateIdAccess );
 	}
+
+	@Override
+	protected Serializable saveWithAttachedID(IDAttachResult attachResult, Object entity, String entityName, Object anything, EventSource source)
+	{
+		callbackHandler.preCreate( entity );
+		return super.saveWithAttachedID(attachResult, entity, entityName, anything,
+				source);
+	}
+
 }
