diff --git a/hibernate-core/src/main/java/org/hibernate/dialect/Oracle8iDialect.java b/hibernate-core/src/main/java/org/hibernate/dialect/Oracle8iDialect.java
index e549ed5..3949751 100644
--- a/hibernate-core/src/main/java/org/hibernate/dialect/Oracle8iDialect.java
+++ b/hibernate-core/src/main/java/org/hibernate/dialect/Oracle8iDialect.java
@@ -48,6 +48,7 @@ import org.hibernate.type.StandardBasicTypes;
  * A dialect for Oracle 8i.
  *
  * @author Steve Ebersole
+ * @author Lukasz Antoniak (lukasz dot antoniak at gmail dot com)
  */
 public class Oracle8iDialect extends Dialect {
 
@@ -56,6 +57,7 @@ public class Oracle8iDialect extends Dialect {
 		registerCharacterTypeMappings();
 		registerNumericTypeMappings();
 		registerDateTimeTypeMappings();
+		registerBooleanTypeMappings();
 		registerLargeObjectTypeMappings();
 
 		registerReverseHibernateTypeMappings();
@@ -65,6 +67,10 @@ public class Oracle8iDialect extends Dialect {
 		registerDefaultProperties();
 	}
 
+	protected void registerBooleanTypeMappings() {
+		registerColumnType( Types.BOOLEAN, "number(1,0)" );
+	}
+
 	protected void registerCharacterTypeMappings() {
 		registerColumnType( Types.CHAR, "char(1)" );
 		registerColumnType( Types.VARCHAR, 4000, "varchar2($l)" );
