Index: C:/projects/workspace/Hibernate3/src/org/hibernate/event/def/DefaultEvictEventListener.java
===================================================================
--- C:/projects/workspace/Hibernate3/src/org/hibernate/event/def/DefaultEvictEventListener.java	(revision 10177)
+++ C:/projects/workspace/Hibernate3/src/org/hibernate/event/def/DefaultEvictEventListener.java	(working copy)
@@ -54,6 +54,7 @@
 			EntityKey key = new EntityKey( id, persister, source.getEntityMode() );
 			persistenceContext.removeProxy(key);
 			
+			
 			if ( !li.isUninitialized() ) {
 				final Object entity = persistenceContext.removeEntity(key);
 				if ( entity != null ) {
@@ -62,6 +63,7 @@
 				}
 			}
 			
+			li.setSession(null);
 		}
 		else {
 			
