diff --git a/core/src/main/java/org/hibernate/persister/collection/BasicCollectionPersister.java b/core/src/main/java/org/hibernate/persister/collection/BasicCollectionPersister.java
index 100c530..041eb4b 100644
--- a/core/src/main/java/org/hibernate/persister/collection/BasicCollectionPersister.java
+++ b/core/src/main/java/org/hibernate/persister/collection/BasicCollectionPersister.java
@@ -126,16 +126,17 @@ public class BasicCollectionPersister extends AbstractCollectionPersister {
 		Update update = new Update( getDialect() )
 			.setTableName( qualifiedTableName );
 		
-		//if ( !elementIsFormula ) {
-			update.addColumns( elementColumnNames, elementColumnIsSettable, elementColumnWriters );
-		//}
-		
+        if ( hasIndex && !indexContainsFormula ) {
+            update.addColumns( indexColumnNames );
+        } else {
+    		//if ( !elementIsFormula ) {
+    			update.addColumns( elementColumnNames, elementColumnIsSettable, elementColumnWriters );
+    		//}
+        }
+        
 		if ( hasIdentifier ) {
 			update.addPrimaryKeyColumns( new String[]{ identifierColumnName } );
 		}
-		else if ( hasIndex && !indexContainsFormula ) {
-			update.addPrimaryKeyColumns( ArrayHelper.join( keyColumnNames, indexColumnNames ) );
-		}
 		else {
 			update.addPrimaryKeyColumns( keyColumnNames );
 			update.addPrimaryKeyColumns( elementColumnNames, elementColumnIsInPrimaryKey, elementColumnWriters );
@@ -159,9 +160,6 @@ public class BasicCollectionPersister extends AbstractCollectionPersister {
 		if ( hasIdentifier ) {
 			delete.addPrimaryKeyColumns( new String[]{ identifierColumnName } );
 		}
-		else if ( hasIndex && !indexContainsFormula ) {
-			delete.addPrimaryKeyColumns( ArrayHelper.join( keyColumnNames, indexColumnNames ) );
-		}
 		else {
 			delete.addPrimaryKeyColumns( keyColumnNames );
 			delete.addPrimaryKeyColumns( elementColumnNames, elementColumnIsInPrimaryKey, elementColumnWriters );
