Index: core/src/main/java/org/hibernate/dialect/HSQLDialect.java
===================================================================
--- core/src/main/java/org/hibernate/dialect/HSQLDialect.java	(revision )
+++ core/src/main/java/org/hibernate/dialect/HSQLDialect.java	(revision )
@@ -84,7 +84,7 @@
 		}
 
 		registerColumnType( Types.BIGINT, "bigint" );
-		registerColumnType( Types.BINARY, "binary" );
+		registerColumnType( Types.BINARY, "binary($l)" );
 		registerColumnType( Types.BIT, "bit" );
         registerColumnType( Types.BOOLEAN, "boolean" );
 		registerColumnType( Types.CHAR, "char($l)" );
