Index: ./src/org/hibernate/stat/QueryStatistics.java
===================================================================
--- ./src/org/hibernate/stat/QueryStatistics.java	(revision 14265)
+++ ./src/org/hibernate/stat/QueryStatistics.java	(working copy)
@@ -16,7 +16,7 @@
 	/*package*/ long cachePutCount;
 	private long executionCount;
 	private long executionRowCount;
-	private long executionAvgTime;
+	private long executionTotalTime = 0;
 	private long executionMaxTime;
 	private long executionMinTime = Long.MAX_VALUE;
 
@@ -63,7 +63,7 @@
 	 * average time in ms taken by the excution of this query onto the DB
 	 */
 	public long getExecutionAvgTime() {
-		return executionAvgTime;
+		return executionTotalTime / executionCount;
 	}
 
 	/**
@@ -89,7 +89,7 @@
 	void executed(long rows, long time) {
 		if (time < executionMinTime) executionMinTime = time;
 		if (time > executionMaxTime) executionMaxTime = time;
-		executionAvgTime = ( executionAvgTime * executionCount + time ) / ( executionCount + 1 );
+		executionTotalTime += time;
 		executionCount++;
 		executionRowCount += rows;
 	}
@@ -102,7 +102,7 @@
 				.append( ",cachePutCount=" ).append( this.cachePutCount )
 				.append( ",executionCount=" ).append( this.executionCount )
 				.append( ",executionRowCount=" ).append( this.executionRowCount )
-				.append( ",executionAvgTime=" ).append( this.executionAvgTime )
+				.append( ",executionAvgTime=" ).append( this.getExecutionAvgTime() )
 				.append( ",executionMaxTime=" ).append( this.executionMaxTime )
 				.append( ",executionMinTime=" ).append( this.executionMinTime )
 				.append( ']' )
