Index: annotations/src/test/java/org/hibernate/test/annotations/manytomany/ContactInfo.java
===================================================================
--- annotations/src/test/java/org/hibernate/test/annotations/manytomany/ContactInfo.java	(revision 0)
+++ annotations/src/test/java/org/hibernate/test/annotations/manytomany/ContactInfo.java	(revision 0)
@@ -0,0 +1,14 @@
+package org.hibernate.test.annotations.manytomany;
+
+import javax.persistence.Embeddable;
+import javax.persistence.ManyToMany;
+import javax.persistence.ManyToOne;
+import java.util.List;
+
+@Embeddable
+public class ContactInfo {
+//	@ManyToOne
+//	Address address; // Unidirectional
+	@ManyToMany
+List<PhoneNumber> phoneNumbers; // Bidirectional
+}
Index: annotations/src/test/java/org/hibernate/test/annotations/manytomany/Employee.java
===================================================================
--- annotations/src/test/java/org/hibernate/test/annotations/manytomany/Employee.java	(revision 18520)
+++ annotations/src/test/java/org/hibernate/test/annotations/manytomany/Employee.java	(working copy)
@@ -4,6 +4,7 @@
 import java.io.Serializable;
 import java.util.Collection;
 import javax.persistence.CascadeType;
+import javax.persistence.Embedded;
 import javax.persistence.Entity;
 import javax.persistence.GeneratedValue;
 import javax.persistence.Id;
@@ -26,7 +27,17 @@
 	private Integer id;
 	private Collection<Employer> employers;
 	private String name;
+	ContactInfo contactInfo;
 
+	@Embedded
+	public ContactInfo getContactInfo() {
+		return contactInfo;
+	}
+
+	public void setContactInfo(ContactInfo contactInfo) {
+		this.contactInfo = contactInfo;
+	}
+
 	@Column(name="fld_name")
 	public String getName() {
 		return name;
Index: annotations/src/test/java/org/hibernate/test/annotations/manytomany/PhoneNumber.java
===================================================================
--- annotations/src/test/java/org/hibernate/test/annotations/manytomany/PhoneNumber.java	(revision 0)
+++ annotations/src/test/java/org/hibernate/test/annotations/manytomany/PhoneNumber.java	(revision 0)
@@ -0,0 +1,30 @@
+package org.hibernate.test.annotations.manytomany;
+
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.ManyToMany;
+import java.util.Collection;
+
+@Entity
+public class PhoneNumber {
+		@Id
+		int phNumber;
+		@ManyToMany(mappedBy="contactInfo.phoneNumbers")
+		Collection<Employee> employees;
+
+	public int getPhNumber() {
+		return phNumber;
+	}
+
+	public void setPhNumber(int phNumber) {
+		this.phNumber = phNumber;
+	}
+
+	public Collection<Employee> getEmployees() {
+		return employees;
+	}
+
+	public void setEmployees(Collection<Employee> employees) {
+		this.employees = employees;
+	}
+}
Index: annotations/src/test/java/org/hibernate/test/annotations/manytomany/ManyToManyTest.java
===================================================================
--- annotations/src/test/java/org/hibernate/test/annotations/manytomany/ManyToManyTest.java	(revision 18520)
+++ annotations/src/test/java/org/hibernate/test/annotations/manytomany/ManyToManyTest.java	(working copy)
@@ -664,7 +664,8 @@
 				Inspector.class,
 				InspectorPrefixes.class,
 				BuildingCompany.class,
-				Building.class
+				Building.class,
+				PhoneNumber.class
 		};
 	}
 
