Index: tools/src/java/org/hibernate/cfg/JDBCBinder.java
===================================================================
--- tools/src/java/org/hibernate/cfg/JDBCBinder.java	(revision 17381)
+++ tools/src/java/org/hibernate/cfg/JDBCBinder.java	(working copy)
@@ -259,7 +259,7 @@
             ForeignKey fk, Set processedColumns, boolean constrained, boolean inverseProperty) {
 
 
-        OneToOne value = new OneToOne(targetTable, rc);
+        OneToOne value = new OneToOne(fk.getTable(), rc);
         value.setReferencedEntityName(revengStrategy
                 .tableToClassName(TableIdentifier.create(targetTable)));
 
