Index: CodeGenerator.java
===================================================================
RCS file: /cvsroot/hibernate/HibernateExt/tools/src/java/net/sf/hibernate/tool/hbm2java/CodeGenerator.java,v
retrieving revision 1.4
diff -u -r1.4 CodeGenerator.java
--- CodeGenerator.java	4 Dec 2003 21:42:17 -0000	1.4
+++ CodeGenerator.java	11 Jan 2004 16:54:36 -0000
@@ -40,7 +40,7 @@
         try {
             ArrayList mappingFiles = new ArrayList();
     
-            SAXBuilder builder = new SAXBuilder();
+            SAXBuilder builder = new SAXBuilder(true);
             builder.setEntityResolver( new DTDEntityResolver() );
             
             builder.setErrorHandler( new ErrorHandler() {
