Index: net/sf/hibernate/type/TimestampType.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/type/TimestampType.java,v
retrieving revision 1.6.2.6
diff -u -r1.6.2.6 TimestampType.java
--- net/sf/hibernate/type/TimestampType.java	13 Dec 2003 15:43:52 -0000	1.6.2.6
+++ net/sf/hibernate/type/TimestampType.java	13 May 2004 18:24:36 -0000
@@ -41,7 +41,7 @@
 	public String getName() { return "timestamp"; }
 	
 	public String toString(Object val) {
-		return new SimpleDateFormat("dd MMMM yyyy HH:mm:ss").format( (java.util.Date) val );
+		return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format( (java.util.Date) val );
 	}
 	
 	public Object deepCopyNotNull(Object value) {
@@ -102,7 +102,7 @@
 	
 	public Object fromStringValue(String xml) throws HibernateException {
 		try {
-			return new SimpleDateFormat().parse(xml);
+			return new Timestamp(new SimpleDateFormat("yyyy-MM-dd hh:mm:ss").parse(xml).getTime());
 		}
 		catch (ParseException pe) {
 			throw new HibernateException("could not parse XML", pe);
