Index: C:/projects/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java
===================================================================
--- C:/projects/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java	(revision 10973)
+++ C:/projects/Hibernate3/src/org/hibernate/persister/entity/AbstractEntityPersister.java	(working copy)
@@ -2915,7 +2915,9 @@
 		if ( isIdentifierAssignedByInsert() ) {
 			identityDelegate = ( ( PostInsertIdentifierGenerator ) getIdentifierGenerator() )
 					.getInsertGeneratedIdentifierDelegate( this, getFactory().getDialect(), useGetGeneratedKeys() );
-			sqlIdentityInsertString = generateIdentityInsertString( getPropertyInsertability() );
+			sqlIdentityInsertString = customSQLInsert[0] == null ?
+                                    generateIdentityInsertString( getPropertyInsertability() )
+                                    : customSQLInsert[0];
 		}
 		else {
 			sqlIdentityInsertString = null;
