Index: tools/src/java/org/hibernate/cfg/JDBCBinder.java
===================================================================
--- tools/src/java/org/hibernate/cfg/JDBCBinder.java	(revision 15749)
+++ tools/src/java/org/hibernate/cfg/JDBCBinder.java	(working copy)
@@ -245,7 +245,8 @@ public class JDBCBinder {
 					rc.addProperty(property);
 				} else {										
 					Property property = bindOneToMany(rc, foreignKey, processed, mapping);
-					rc.addProperty(property);
+					if (property != null)
+					    rc.addProperty(property);
 				}
 			}
 		}
@@ -358,6 +359,13 @@ public class JDBCBinder {
 			collection.setElement( element );
 			
         } else {
+        	Iterator foreignKeyIterator = foreignKey.getTable().getForeignKeyIterator();
+        	while ( foreignKeyIterator.hasNext() ) {
+				Object next = foreignKeyIterator.next();
+				if (next == foreignKey)
+					return null;
+			}
+
         	String tableToClassName = bindCollection( rc, foreignKey, null, collection );
             
         	OneToMany oneToMany = new OneToMany( collection.getOwner() );
@@ -380,6 +388,7 @@ public class JDBCBinder {
 		}
 
 		SimpleValue keyValue = new DependantValue( collectionTable, referencedKeyValue );
+		keyValue.setForeignKeyName(foreignKey.getName());
 		//keyValue.setForeignKeyName("none"); // Avoid creating the foreignkey
 		//key.setCascadeDeleteEnabled( "cascade".equals( subnode.attributeValue("on-delete") ) );
 		Iterator columnIterator = foreignKey.getColumnIterator();        
