*** C:\hibernate-2.1\src\net\sf\hibernate\impl\AbstractQueryImpl.java	Sun Aug 08 22:38:22 2004
--- C:\workspace\weight\source\java\net\sf\hibernate\impl\AbstractQueryImpl.java	Mon Aug 16 12:06:50 2004
***************
*** 507,515 ****
  			String namedParam = params[i];
  			try {
  				Getter getter = ReflectHelper.getGetter(clazz, namedParam);
  				Class retType = getter.getReturnType(); 
! 			      if (retType.isAssignableFrom(java.util.Collection.class)) { 
  			        setParameterList( namedParam, (Collection) getter.get(bean) ); 
  			      } 
  			      else if (retType.isArray()) {
  			      	setParameterList(namedParam, (Object[])getter.get(bean));
--- 507,515 ----
  			String namedParam = params[i];
  			try {
  				Getter getter = ReflectHelper.getGetter(clazz, namedParam);
  				Class retType = getter.getReturnType(); 
! 			      if (java.util.Collection.class.isAssignableFrom(retType)) { 
  			        setParameterList( namedParam, (Collection) getter.get(bean) ); 
  			      } 
  			      else if (retType.isArray()) {
  			      	setParameterList(namedParam, (Object[])getter.get(bean));
