Index: src/org/hibernate/dialect/MySQLInnoDBDialect.java
===================================================================
--- src/org/hibernate/dialect/MySQLInnoDBDialect.java	(revision 15766)
+++ src/org/hibernate/dialect/MySQLInnoDBDialect.java	(working copy)
@@ -18,4 +18,7 @@ public class MySQLInnoDBDialect extends 
 		return true;
 	}
 	
+	public boolean supportsUnique() {
+		return false;
+	}
 }
