Index: src/org/hibernate/dialect/Dialect.java
===================================================================
--- src/org/hibernate/dialect/Dialect.java	(revision 10348)
+++ src/org/hibernate/dialect/Dialect.java	(working copy)
@@ -1010,5 +1010,11 @@
 		return "current_timestamp";
 	}
 
-	
+	public boolean supportsSchemaNames() {
+		return true;
+	}	
+
+	public boolean supportsCatalogNames() {
+		return true;
+	}
 }
Index: src/org/hibernate/dialect/HSQLDialect.java
===================================================================
--- src/org/hibernate/dialect/HSQLDialect.java	(revision 10348)
+++ src/org/hibernate/dialect/HSQLDialect.java	(working copy)
@@ -273,4 +273,11 @@
 		return false;
 	}
 
+	public boolean supportsSchemaNames() {
+		return false;
+	}
+
+	public boolean supportsCatalogNames() {
+		return false;
+	}
 }
Index: src/org/hibernate/mapping/Table.java
===================================================================
--- src/org/hibernate/mapping/Table.java	(revision 10348)
+++ src/org/hibernate/mapping/Table.java	(working copy)
@@ -90,6 +90,12 @@
 		String quotedName = getQuotedName(dialect);
 		String usedSchema = schema == null ? defaultSchema : getQuotedSchema(dialect);
 		String usedCatalog = catalog == null ? defaultCatalog : catalog;
+		
+		if (dialect.supportsSchemaNames() == false)
+			usedSchema = null;
+		if (dialect.supportsCatalogNames() == false)
+			usedCatalog = null;
+		
 		return qualify( usedCatalog, usedSchema, quotedName );
 	}
 
