Index: src/org/hibernate/criterion/DetachedCriteria.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/criterion/DetachedCriteria.java,v
retrieving revision 1.4
diff -u -r1.4 DetachedCriteria.java
--- src/org/hibernate/criterion/DetachedCriteria.java	6 Dec 2005 03:35:50 -0000	1.4
+++ src/org/hibernate/criterion/DetachedCriteria.java	14 Dec 2005 18:51:03 -0000
@@ -85,7 +85,13 @@
 		return this;
 	}
 
-	public DetachedCriteria createCriteria(String associationPath, String alias)
+	public DetachedCriteria createAlias(String associationPath, String alias, int joinType)
+	throws HibernateException {
+		criteria.createAlias(associationPath, alias, joinType);
+		return this;
+	}
+
+    public DetachedCriteria createCriteria(String associationPath, String alias)
 	throws HibernateException {
 		return new DetachedCriteria( impl, criteria.createCriteria(associationPath) );
 	}
@@ -95,7 +101,17 @@
 		return new DetachedCriteria( impl, criteria.createCriteria(associationPath) );
 	}
 
-	public String getAlias() {
+    public DetachedCriteria createCriteria(String associationPath, String alias, int joinType)
+	throws HibernateException {
+		return new DetachedCriteria( impl, criteria.createCriteria(associationPath, joinType) );
+	}
+
+	public DetachedCriteria createCriteria(String associationPath, int joinType)
+	throws HibernateException {
+		return new DetachedCriteria( impl, criteria.createCriteria(associationPath, joinType) );
+	}
+
+    public String getAlias() {
 		return criteria.getAlias();
 	}
