Index: hibernate-core/src/main/java/org/hibernate/loader/entity/CollectionElementLoader.java
===================================================================
--- hibernate-core/src/main/java/org/hibernate/loader/entity/CollectionElementLoader.java	(revision 1)
+++ hibernate-core/src/main/java/org/hibernate/loader/entity/CollectionElementLoader.java	(revision )
@@ -73,7 +73,7 @@
 				persister, 
 				ArrayHelper.join( 
 						collectionPersister.getKeyColumnNames(), 
-						collectionPersister.getIndexColumnNames()
+						collectionPersister.toColumns("index")
 					),
 				1, 
 				LockMode.NONE, 
