Index: src/main/java/org/hibernate/stat/StatisticsImpl.java
===================================================================
--- src/main/java/org/hibernate/stat/StatisticsImpl.java	(revision 15827)
+++ src/main/java/org/hibernate/stat/StatisticsImpl.java	(working copy)
@@ -293,6 +293,7 @@
 		if (hql!=null) {
 			QueryStatistics qs = getQueryStatistics(hql);
 			qs.executed(rows, time);
+            log.info("HQL: {}, time: {}ms, rows: {}", new Object[]{hql, new Long(time), new Long(rows)});
 		}
 	}
 	
