Index: tools/src/java/org/hibernate/tool/hbm2x/Cfg2HbmTool.java
===================================================================
--- tools/src/java/org/hibernate/tool/hbm2x/Cfg2HbmTool.java	(revision 16240)
+++ tools/src/java/org/hibernate/tool/hbm2x/Cfg2HbmTool.java	(working copy)
@@ -427,4 +427,16 @@
 	public Iterator getProperties(PersistentClass pc) {
 		return new SkipBackRefPropertyIterator(pc.getUnjoinedPropertyIterator());
 	}
+	
+	public boolean hasNaturalIdentifierProperties(PersistentClass pc) {
+		Iterator propertyIterator = pc.getPropertyIterator();
+		while (propertyIterator.hasNext()) {
+			Property property = (Property) propertyIterator.next();
+			
+			if (property.isNaturalIdentifier())
+				return true;
+		}
+		
+		return false;
+	}
 }
Index: tools/src/templates/hbm/natural-id.hbm.ftl
===================================================================
--- tools/src/templates/hbm/natural-id.hbm.ftl	(revision 0)
+++ tools/src/templates/hbm/natural-id.hbm.ftl	(revision 0)
@@ -0,0 +1,7 @@
+<natural-id>
+	<#foreach property in c2h.getProperties(clazz)>
+		<#if property.naturalIdentifier>
+			<#include "${c2h.getTag(property)}.hbm.ftl"/>
+		</#if>
+	</#foreach>
+</natural-id>
\ No newline at end of file
Index: tools/src/templates/hbm/persistentclass.hbm.ftl
===================================================================
--- tools/src/templates/hbm/persistentclass.hbm.ftl	(revision 16240)
+++ tools/src/templates/hbm/persistentclass.hbm.ftl	(working copy)
@@ -102,6 +102,11 @@
  </key>
 </#if>
 
+<#-- natural-id has to be done here (order in DTD) -->
+<#if c2h.hasNaturalIdentifierProperties(clazz)>
+	<#include "natural-id.hbm.ftl"/>
+</#if>
+
 <#-- version has to be done explicitly since Annotation's does not list version first -->
 <#if pojo.hasVersionProperty()>
 <#assign property=clazz.getVersion()/>
@@ -109,7 +114,7 @@
 </#if>
 
 <#foreach property in c2h.getProperties(clazz)>
-<#if c2h.getTag(property)!="version" && c2h.getTag(property)!="timestamp">
+<#if c2h.getTag(property)!="version" && c2h.getTag(property)!="timestamp" && !property.naturalIdentifier>
 <#include "${c2h.getTag(property)}.hbm.ftl"/>
 </#if>
 </#foreach>
