diff --git a/hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java b/hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java
index ca759e1..9fbfef6 100644
--- a/hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java
+++ b/hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java
@@ -91,7 +91,6 @@ import org.hibernate.internal.util.collections.ArrayHelper;
 import org.hibernate.jdbc.Expectation;
 import org.hibernate.jdbc.Expectations;
 import org.hibernate.jdbc.TooManyRowsAffectedException;
-import org.hibernate.loader.entity.BatchingEntityLoader;
 import org.hibernate.loader.entity.BatchingEntityLoaderBuilder;
 import org.hibernate.loader.entity.CascadeEntityLoader;
 import org.hibernate.loader.entity.EntityLoader;
@@ -603,6 +602,7 @@ public abstract class AbstractEntityPersister
 				if ( thing.isFormula() ) {
 					foundFormula = true;
 					formulaTemplates[k] = thing.getTemplate( factory.getDialect(), factory.getSqlFunctionRegistry() );
+					colReaderTemplates[k] = colAliases[k];
 				}
 				else {
 					Column col = (Column)thing;
@@ -699,8 +699,10 @@ public abstract class AbstractEntityPersister
 					formulaTemplates.add( template );
 					forms[l] = template;
 					formulas.add( thing.getText( factory.getDialect() ) );
-					formulaAliases.add( thing.getAlias( factory.getDialect() ) );
+					final String formulaAlias = thing.getAlias( factory.getDialect() );
+					formulaAliases.add( formulaAlias );
 					formulasLazy.add( lazy );
+					readerTemplates[l] = formulaAlias;
 				}
 				else {
 					Column col = (Column)thing;
