Index: hibernate-configuration-2.0.dtd
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/hibernate-configuration-2.0.dtd,v
retrieving revision 1.7.2.1
diff -u -r1.7.2.1 hibernate-configuration-2.0.dtd
--- hibernate-configuration-2.0.dtd	27 Aug 2003 00:59:49 -0000	1.7.2.1
+++ hibernate-configuration-2.0.dtd	14 Oct 2003 05:19:06 -0000
@@ -40,5 +40,5 @@
 <!ATTLIST collection-cache region CDATA #IMPLIED>
 <!ATTLIST collection-cache usage (read-only|read-write|nonstrict-read-write) #REQUIRED>
 
-<!ELEMENT session-factory (property*, mapping+, (class-cache|collection-cache|jcs-class-cache|jcs-collection-cache)*)>
+<!ELEMENT session-factory (property*, mapping*, (class-cache|collection-cache|jcs-class-cache|jcs-collection-cache)*)>
 <!ATTLIST session-factory name CDATA #IMPLIED> <!-- the JNDI name -->
