Index: tools/src/java/org/hibernate/tool/hbm2x/pojo/POJOClass.java
===================================================================
--- tools/src/java/org/hibernate/tool/hbm2x/pojo/POJOClass.java	(revision 18464)
+++ tools/src/java/org/hibernate/tool/hbm2x/pojo/POJOClass.java	(working copy)
@@ -66,6 +66,7 @@
 	public boolean hasIdentifierProperty();
 	
 	public String generateAnnColumnAnnotation(Property property);
+	public String generateAnnTypeAnnotation(Property property);
 	public String generateAnnIdGenerator();
 	public String generateAnnTableUniqueConstraint();
 	public String generateBasicAnnotation(Property property);
Index: tools/src/java/org/hibernate/tool/hbm2x/pojo/BasicPOJOClass.java
===================================================================
--- tools/src/java/org/hibernate/tool/hbm2x/pojo/BasicPOJOClass.java	(revision 18464)
+++ tools/src/java/org/hibernate/tool/hbm2x/pojo/BasicPOJOClass.java	(working copy)
@@ -28,6 +28,8 @@
 import org.hibernate.tool.hbm2x.MetaAttributeConstants;
 import org.hibernate.tool.hbm2x.MetaAttributeHelper;
 import org.hibernate.tool.hbm2x.visitor.DefaultValueVisitor;
+import org.hibernate.type.CustomType;
+import org.hibernate.type.Type;
 import org.hibernate.util.StringHelper;
 
 /**
@@ -495,6 +497,16 @@
 		return properties.iterator();
 	}
 
+    public String generateAnnTypeAnnotation(Property property) {
+        Type type = property.getType();
+        if (type instanceof CustomType) {
+            return "    @" + importType("org.hibernate.annotations.Type") + "(type=\"" +
+                    ((CustomType) type).getName() + "\")";
+        } else {
+            return "";
+        }
+    }
+
 	public Iterator getEqualsHashCodePropertiesIterator() {
 		Iterator iter = getAllPropertiesIterator();
 		return getEqualsHashCodePropertiesIterator(iter);
Index: tools/src/templates/pojo/Ejb3PropertyGetAnnotation.ftl
===================================================================
--- tools/src/templates/pojo/Ejb3PropertyGetAnnotation.ftl	(revision 18464)
+++ tools/src/templates/pojo/Ejb3PropertyGetAnnotation.ftl	(working copy)
@@ -18,5 +18,6 @@
 <#else>
 ${pojo.generateBasicAnnotation(property)}
 ${pojo.generateAnnColumnAnnotation(property)}
+${pojo.generateAnnTypeAnnotation(property)}
 </#if>
 </#if>
\ No newline at end of file
