Index: CGLIBLazyInitializer.java
===================================================================
--- CGLIBLazyInitializer.java	(revision 10107)
+++ CGLIBLazyInitializer.java	(working copy)
@@ -84,29 +84,7 @@
 			final Method getIdentifierMethod, final Method setIdentifierMethod,
 			final AbstractComponentType componentIdType, final Serializable id,
 			final SessionImplementor session) throws HibernateException {
-		
-		final CGLIBLazyInitializer instance = new CGLIBLazyInitializer(
-				entityName,
-				persistentClass,
-				interfaces,
-				id,
-				getIdentifierMethod,
-				setIdentifierMethod,
-				componentIdType,
-				session 
-			);
-		
-		final HibernateProxy proxy;
-		try {
-			proxy = (HibernateProxy) factory.newInstance();
-		}
-		catch (Exception e) {
-			throw new HibernateException( "CGLIB Enhancement failed: " + persistentClass.getName(), e );
-		}
-		( (Factory) proxy ).setCallback( 0, instance );
-		instance.constructed = true;
-
-		return proxy;
+		return getProxy(entityName, persistentClass, interfaces, getIdentifierMethod, setIdentifierMethod, componentIdType, id, session);
 	}
 
 	public static Class getProxyFactory(Class persistentClass, Class[] interfaces)
