--- C:\temp\hibernate-3.2\src\org\hibernate\collection\PersistentMap.java	2006-09-02 07:56:40.000000000 +-1100
+++ C:\temp\hibernate-3.2\src\org\hibernate\collection\PersistentMap-fixes.java	2007-01-10 18:09:28.000000000 +-1100
@@ -382,24 +382,24 @@
 	throws HibernateException {
 		List deletes = new ArrayList();
 		Iterator iter = ( (Map) getSnapshot() ).entrySet().iterator();
 		while ( iter.hasNext() ) {
 			Map.Entry e = (Map.Entry) iter.next();
 			Object key = e.getKey();
-			if ( e.getValue()!=null && map.get(key)==null ) {
+			if ( !map.containsKey(key) ) {
 				deletes.add( indexIsFormula ? e.getValue() : key );
 			}
 		}
 		return deletes.iterator();
 	}
 
 	public boolean needsInserting(Object entry, int i, Type elemType) 
 	throws HibernateException {
 		final Map sn = (Map) getSnapshot();
 		Map.Entry e = (Map.Entry) entry;
-		return e.getValue()!=null && sn.get( e.getKey() )==null;
+		return !sn.containsKey( e.getKey() );
 	}
 
 	public boolean needsUpdating(Object entry, int i, Type elemType) 
 	throws HibernateException {
 		final Map sn = (Map) getSnapshot();
 		Map.Entry e = (Map.Entry) entry;
