Index: core/src/main/java/org/hibernate/type/ComponentType.java
===================================================================
--- core/src/main/java/org/hibernate/type/ComponentType.java	(revision 19372)
+++ core/src/main/java/org/hibernate/type/ComponentType.java	(working copy)
@@ -282,7 +282,7 @@
 			return current != null;
 		}
 		Object[] currentValues = getPropertyValues( current, session );
-		Object[] oldValues = ( Object[] ) old;
+		Object[] oldValues =     getPropertyValues(old, session);
 		int loc = 0;
 		for ( int i = 0; i < currentValues.length; i++ ) {
 			int len = propertyTypes[i].getColumnSpan( session.getFactory() );
