Index: tools/src/java/org/hibernate/tool/hbm2x/pojo/EntityPOJOClass.java
===================================================================
--- tools/src/java/org/hibernate/tool/hbm2x/pojo/EntityPOJOClass.java	(revision 17230)
+++ tools/src/java/org/hibernate/tool/hbm2x/pojo/EntityPOJOClass.java	(working copy)
@@ -59,7 +59,7 @@
 
 		if ( isInterface() ) {
 			if ( clazz.getSuperclass() != null ) {
-				extendz = clazz.getSuperclass().getClassName();
+				extendz = importType(clazz.getSuperclass().getClassName());
 			}
 			if ( clazz.getMetaAttribute( EXTENDS ) != null ) {
 				if ( !"".equals( extendz ) ) {
@@ -73,7 +73,7 @@
 				// class cannot extend it's superclass because the superclass is marked as an interface
 			}
 			else {
-				extendz = clazz.getSuperclass().getClassName();
+				extendz = importType(clazz.getSuperclass().getClassName());
 			}
 		}
 		else if ( clazz.getMetaAttribute( EXTENDS ) != null ) {
