Index: .
===================================================================
--- .	(revision 10039)
+++ .	(working copy)
@@ -318,10 +318,14 @@
 	}
 
 	private static String generateAliasRoot(String description) {
-		final String result = truncate( unqualifyEntityName(description), ALIAS_TRUNCATE_LENGTH )
+		String result = truncate( unqualifyEntityName(description), ALIAS_TRUNCATE_LENGTH )
 				.toLowerCase()
 		        .replace( '/', '_' ) // entityNames may now include slashes for the representations
 				.replace( '$', '_' ); //classname may be an inner class
+		
+		if(result.charAt(0)=='_') // Oracle doesn't like leading dashes
+			result = "x" + result;
+		
 		if ( Character.isDigit( result.charAt(result.length()-1) ) ) {
 			return result + "x"; //ick!
 		}
