diff --git a/hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java b/hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java
index 45eaa63..902770c 100644
--- a/hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java
+++ b/hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java
@@ -1387,14 +1387,12 @@
 		bindFilters( annotatedClass, entityBinder );
 
 		XClass classToProcess = annotatedClass.getSuperclass();
-		while ( classToProcess != null ) {
-			AnnotatedClassType classType = mappings.getClassType( classToProcess );
-			if ( AnnotatedClassType.EMBEDDABLE_SUPERCLASS.equals( classType ) ) {
-				bindFilters( classToProcess, entityBinder );
-			}
+		
+		// HHH-9257: only include filters if superclass is a MapperSuperClasses, anything else is mapped and has their own bound filters.
+		while ( classToProcess != null && AnnotatedClassType.EMBEDDABLE_SUPERCLASS.equals( mappings.getClassType( classToProcess ) )) {
+			bindFilters( classToProcess, entityBinder );
 			classToProcess = classToProcess.getSuperclass();
 		}
-
 	}
 
 	private static void bindFilters(XAnnotatedElement annotatedElement, EntityBinder entityBinder) {