Index: SingleTableEntityPersister.java
===================================================================
--- SingleTableEntityPersister.java	(revision 16993)
+++ SingleTableEntityPersister.java	(working copy)
@@ -658,7 +658,7 @@
 		
 		//figure out which formulas are needed
 		ArrayList formulaNumbers = new ArrayList();
-		final int[] formulaTableNumbers = getSubclassColumnTableNumberClosure();
+		final int[] formulaTableNumbers = getSubclassFormulaTableNumberClosure();
 		for ( int i=0; i<getSubclassFormulaTemplateClosure().length; i++ ) {
 			if ( tableNumbers.contains( new Integer( formulaTableNumbers[i] ) ) ) {
 				formulaNumbers.add( new Integer(i) );
