Index:  core/src/main/java/org/hibernate/dialect/resolver/StandardDialectResolver.java
===================================================================
--- hibernate_firebird/StandardDialectResolver.java	(revision ???)
+++ hibernate_firebird/StandardDialectResolver.java	(working copy)
@@ -39,6 +39,7 @@
 import org.hibernate.dialect.SQLServerDialect;
 import org.hibernate.dialect.InformixDialect;
 import org.hibernate.dialect.DB2Dialect;
+import org.hibernate.dialect.FirebirdDialect;
 import org.hibernate.dialect.Oracle10gDialect;
 import org.hibernate.dialect.Oracle9iDialect;
 import org.hibernate.dialect.Oracle8iDialect;
@@ -85,6 +86,10 @@
 			return new SQLServerDialect();
 		}
 
+		if ( databaseName.startsWith( "Firebird" ) ) {
+			return new FirebirdDialect();
+		}
+
 		if ( "Sybase SQL Server".equals( databaseName ) || "Adaptive Server Enterprise".equals( databaseName ) ) {
 			return new SybaseASE15Dialect();
 		}
