--- src/org/hibernate/persister/collection/AbstractCollectionPersister.java.org	2006-03-14 15:02:46.341973857 +0100
+++ src/org/hibernate/persister/collection/AbstractCollectionPersister.java	2006-03-14 15:04:40.338360114 +0100
@@ -809,11 +809,14 @@
 		String selectValue = isIntegerIndexed ? 
 			"max(" + getIndexColumnNames()[0] + ") + 1": //lists, arrays
 			"count(" + getElementColumnNames()[0] + ")"; //sets, maps, bags
-		return new SimpleSelect(dialect)
+		SimpleSelect simpleSelect = new SimpleSelect(dialect)
 				.setTableName( getTableName() )
 				.addCondition( getKeyColumnNames(), "=?" )
-				.addColumn(selectValue)
-				.toStatementString();
+				.addColumn(selectValue);
+		if  ( hasWhere() ) {
+			simpleSelect.addCondition("", sqlWhereString);
+		}
+		return simpleSelect.toStatementString();
 	}
 
 	protected String generateDetectRowByIndexString() {
