diff -ur hibernate-extensions-2.1.2-clean\tools\src\java\net\sf\hibernate\tool\hbm2java\ClassMapping.java hibernate-extensions-2.1.2\tools\src\java\net\sf\hibernate\tool\hbm2java\ClassMapping.java
--- hibernate-extensions-2.1.2-clean\tools\src\java\net\sf\hibernate\tool\hbm2java\ClassMapping.java	Mon Apr 26 01:30:46 2004
+++ hibernate-extensions-2.1.2\tools\src\java\net\sf\hibernate\tool\hbm2java\ClassMapping.java	Wed Dec 01 09:50:37 2004
@@ -309,6 +309,21 @@
          addFieldProperty(ff);
          components.put( mapping.getFullyQualifiedName(), mapping );
       }
+
+      // dynamic-components
+      for ( Iterator iter = classElement.getChildren("dynamic-component").iterator(); iter.hasNext(); ) {
+         Element cmpe = (Element) iter.next();
+         MultiMap metaForComponent = MetaAttributeHelper.loadAndMergeMetaMap(cmpe, getMetaAttribs());
+         String cmpname = cmpe.getAttributeValue("name");
+         String cmpclass = "java.util.Map";
+         ClassMapping mapping = new ClassMapping(classPackage, cmpe, this, false, getMetaAttribs());
+         
+         ClassName classType = new ClassName(cmpclass);
+         // add an import and field for this property
+         addImport(classType);
+         FieldProperty ff =  new FieldProperty(cmpe, this, cmpname, classType, false, metaForComponent);
+         addFieldProperty(ff);
+      }
     
         
         //    subclasses (done last so they can access this superclass for info)
Only in hibernate-extensions-2.1.2\tools\src\java\net\sf\hibernate\tool\hbm2java: ClassMapping.java~
