Index: Configuration.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate2/src/net/sf/hibernate/cfg/Configuration.java,v
retrieving revision 1.27.2.26
diff -u -r1.27.2.26 Configuration.java
--- Configuration.java	16 Dec 2003 14:07:06 -0000	1.27.2.26
+++ Configuration.java	5 Jan 2004 22:36:14 -0000
@@ -166,8 +166,9 @@
 		log.info("Mapping file: " + xmlFile);
 		try {
 			List errors = new ArrayList();
-			add( XMLHelper.createSAXReader(xmlFile, errors).read( new File(xmlFile) ) );
+            org.dom4j.Document parsedDocument = XMLHelper.createSAXReader(xmlFile, errors).read( new File(xmlFile) ) ;
 			if ( errors.size()!=0 ) throw new MappingException( "invalid mapping", (Throwable) errors.get(0) );
+            add(parsedDocument);
 			return this;
 		}
 		catch (Exception e) {
@@ -200,8 +201,9 @@
 		if ( log.isDebugEnabled() ) log.debug("Mapping XML:\n" + xml);
 		try {
 			List errors = new ArrayList();
-			add( XMLHelper.createSAXReader("XML String", errors).read( new StringReader(xml) ) );
-			if ( errors.size()!=0 ) throw new MappingException( "invalid mapping", (Throwable) errors.get(0) );
+            org.dom4j.Document parsedDocument = XMLHelper.createSAXReader("XML String", errors).read( new StringReader(xml) );
+            if ( errors.size()!=0 ) throw new MappingException( "invalid mapping", (Throwable) errors.get(0) );
+			add(parsedDocument);
 		}
 		catch (Exception e) {
 			log.error("Could not configure datastore from XML", e);
@@ -278,8 +280,9 @@
 	public Configuration addInputStream(InputStream xmlInputStream) throws MappingException {
 		try {
 			List errors = new ArrayList();
-			add( XMLHelper.createSAXReader("XML InputStream", errors).read( new InputSource(xmlInputStream) ) );
+            org.dom4j.Document parsedDocument =  XMLHelper.createSAXReader("XML InputStream", errors).read( new InputSource(xmlInputStream) ) ;
 			if ( errors.size()!=0 ) throw new MappingException( "invalid mapping", (Throwable) errors.get(0) );
+            add(parsedDocument);			
 			return this;
 		}
 		catch (MappingException me) {
