Index: hql-sql.g
===================================================================
RCS file: /usr/local/cvsroot/Hibernate-3.0.1/grammar/hql-sql.g,v
retrieving revision 1.1
diff -u -r1.1 hql-sql.g
--- hql-sql.g	19 Apr 2005 00:02:08 -0000	1.1
+++ hql-sql.g	19 Apr 2005 17:22:40 -0000
@@ -261,7 +261,10 @@
 
 constructor
 	{ String className = null; }
-	: #(CONSTRUCTOR className=path ( selectExpr | aliasedSelectExpr )* )
+	: #(c:CONSTRUCTOR className=path ( selectExpr | aliasedSelectExpr )* CLOSE! (i:identifier!)? {
+			if (#i != null )
+				((org.hibernate.hql.ast.SelectExpression)#c).setAlias(#i.getText());
+		})
 	;
 
 aggregateExpr
Index: hql.g
===================================================================
RCS file: /usr/local/cvsroot/Hibernate-3.0.1/grammar/hql.g,v
retrieving revision 1.1
diff -u -r1.1 hql.g
--- hql.g	19 Apr 2005 00:02:08 -0000	1.1
+++ hql.g	19 Apr 2005 17:22:40 -0000
@@ -269,8 +269,12 @@
 	;
 
 newExpression
-	: (NEW! path) op:OPEN^ {#op.setType(CONSTRUCTOR);} selectedPropertiesList CLOSE!
-	;
+        : (NEW! path) op:OPEN^ {#op.setType(CONSTRUCTOR);} newPropertiesList CLOSE (AS! identifier)?
+        ;
+
+newPropertiesList
+        : (newExpression | aliasedExpression ) (options {greedy=true;}: COMMA! newPropertiesList )*
+        ;
 
 selectObject
    : OBJECT^ OPEN! identifier CLOSE!
