Index: D:/projets/hibernate3/src/org/hibernate/dialect/SybaseDialect.java
===================================================================
--- D:/projets/hibernate3/src/org/hibernate/dialect/SybaseDialect.java	(revision 11246)
+++ D:/projets/hibernate3/src/org/hibernate/dialect/SybaseDialect.java	(working copy)
@@ -227,4 +227,29 @@
 	public boolean supportsExistsInSelect() {
 		return false;
 	}
+	
+	public String getLimitString(String querySelect, int offset, int limit) {
+		if (offset > 0)
+			throw new UnsupportedOperationException("sql server has no offset");
+
+		else
+			return "SET ROWCOUNT " + limit + " " + querySelect + " SET ROWCOUNT 0" ;
+	}
+
+	public boolean supportsLimit() {
+		return true;
+	}
+
+	public boolean useMaxForLimit() {
+		return true;
+	}
+
+	public boolean supportsLimitOffset() {
+		return false;
+	}
+
+	public boolean supportsVariableLimit() {
+		return false;
+	}
+
 }
