Index: src/org/hibernate/cfg/HbmBinder.java
===================================================================
RCS file: /cvsroot/hibernate/Hibernate3/src/org/hibernate/cfg/HbmBinder.java,v
retrieving revision 1.123
diff -u -r1.123 HbmBinder.java
--- src/org/hibernate/cfg/HbmBinder.java	13 Aug 2005 23:03:34 -0000	1.123
+++ src/org/hibernate/cfg/HbmBinder.java	14 Aug 2005 22:23:41 -0000
@@ -1405,7 +1405,8 @@
 		Attribute lazyNode = node.attribute( "lazy" );
 		boolean isLazyTrue = lazyNode == null ? 
 				defaultLazy && fetchable.isLazy() : //fetch="join" overrides default laziness
-				lazyNode.getValue().equals(proxyVal); //fetch="join" overrides default laziness
+				lazyNode.getValue().equals(proxyVal) ||
+				lazyNode.getValue().equals("true"); //fetch="join" overrides default laziness
 		fetchable.setLazy( isLazyTrue );
 	}
 
