Index: org/hibernate/envers/ebstest/entities/Country.java
===================================================================
--- org/hibernate/envers/ebstest/entities/Country.java	(revision 0)
+++ org/hibernate/envers/ebstest/entities/Country.java	(revision 0)
@@ -0,0 +1,91 @@
+package org.hibernate.envers.ebstest.entities;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+@Entity
+@org.hibernate.envers.Audited
+public class Country {
+	@Id
+	@Column(length = 4)
+	private final Integer code;
+
+	@Column(length = 40)
+	private final String name;
+
+	/**
+	 * Default constructor for persistence provider.
+	 */
+	private Country() {
+		this.code = null;
+		this.name = null;
+	}
+
+	private Country(Integer code, String naam) {
+		this.code = code;
+		this.name = naam;
+	}
+
+	/**
+	 * Factory method voor Country instances.
+	 * 
+	 * @param code
+	 * @param name
+	 * @return a Country
+	 */
+	public static Country of(Integer code, String name) {
+		return new Country(code, name);
+	}
+
+	public Integer getCode() {
+		return code;
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder output = new StringBuilder();
+
+		output.append("Country {");
+		output.append(" code = \"").append(this.code).append("\",");
+		output.append(" name = \"").append(this.name).append("\"}");
+
+		return output.toString();
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((code == null) ? 0 : code.hashCode());
+		result = prime * result + ((name == null) ? 0 : name.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		Country other = (Country) obj;
+		if (code == null) {
+			if (other.code != null)
+				return false;
+		} else if (!code.equals(other.code))
+			return false;
+		if (name == null) {
+			if (other.name != null)
+				return false;
+		} else if (!name.equals(other.name))
+			return false;
+		return true;
+	}
+
+}
Index: org/hibernate/envers/ebstest/entities/Country.java
===================================================================
--- org/hibernate/envers/ebstest/entities/Country.java	(revision 0)
+++ org/hibernate/envers/ebstest/entities/Country.java	(revision 0)
@@ -0,0 +1,91 @@
+package org.hibernate.envers.ebstest.entities;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+@Entity
+@org.hibernate.envers.Audited
+public class Country {
+	@Id
+	@Column(length = 4)
+	private final Integer code;
+
+	@Column(length = 40)
+	private final String name;
+
+	/**
+	 * Default constructor for persistence provider.
+	 */
+	private Country() {
+		this.code = null;
+		this.name = null;
+	}
+
+	private Country(Integer code, String naam) {
+		this.code = code;
+		this.name = naam;
+	}
+
+	/**
+	 * Factory method voor Country instances.
+	 * 
+	 * @param code
+	 * @param name
+	 * @return a Country
+	 */
+	public static Country of(Integer code, String name) {
+		return new Country(code, name);
+	}
+
+	public Integer getCode() {
+		return code;
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder output = new StringBuilder();
+
+		output.append("Country {");
+		output.append(" code = \"").append(this.code).append("\",");
+		output.append(" name = \"").append(this.name).append("\"}");
+
+		return output.toString();
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((code == null) ? 0 : code.hashCode());
+		result = prime * result + ((name == null) ? 0 : name.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		Country other = (Country) obj;
+		if (code == null) {
+			if (other.code != null)
+				return false;
+		} else if (!code.equals(other.code))
+			return false;
+		if (name == null) {
+			if (other.name != null)
+				return false;
+		} else if (!name.equals(other.name))
+			return false;
+		return true;
+	}
+
+}
Index: org/hibernate/envers/ebstest/integration/accesstype/ImmutableClassAccessType.java
===================================================================
--- org/hibernate/envers/ebstest/integration/accesstype/ImmutableClassAccessType.java	(revision 0)
+++ org/hibernate/envers/ebstest/integration/accesstype/ImmutableClassAccessType.java	(revision 0)
@@ -0,0 +1,51 @@
+package org.hibernate.envers.ebstest.integration.accesstype;
+
+import java.util.Arrays;
+
+import javax.persistence.EntityManager;
+
+import org.hibernate.ejb.Ejb3Configuration;
+import org.hibernate.envers.ebstest.entities.Country;
+import org.hibernate.envers.test.AbstractEntityTest;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import static org.testng.Assert.assertEquals;
+
+public class ImmutableClassAccessType extends AbstractEntityTest {
+	private Country country;
+
+	public void configure(Ejb3Configuration cfg) {
+		cfg.addAnnotatedClass(Country.class);
+	}
+
+	@BeforeClass(dependsOnMethods = "init")
+	public void initData() {
+		EntityManager em = getEntityManager();
+
+		// Revision 1
+		em.getTransaction().begin();
+		country = Country.of(new Integer(123), "Germany");
+		em.persist(country);
+		em.getTransaction().commit();
+
+	}
+
+	@Test
+	public void testRevisionsCounts() {
+		assert Arrays.asList(1)
+				.equals(
+						getAuditReader().getRevisions(Country.class,
+								country.getCode()));
+	}
+
+	@Test
+	public void testHistoryOfId1() {
+		Country country1 = getEntityManager().find(Country.class,
+				country.getCode());
+		assertEquals(country1, country);
+		
+		Country history = getAuditReader().find(Country.class, country1.getCode(), 1);
+		assertEquals(country, history);
+	}
+
+}
\ No newline at end of file
Index: org/hibernate/envers/ebstest/integration/accesstype/ImmutableClassAccessType.java
===================================================================
--- org/hibernate/envers/ebstest/integration/accesstype/ImmutableClassAccessType.java	(revision 0)
+++ org/hibernate/envers/ebstest/integration/accesstype/ImmutableClassAccessType.java	(revision 0)
@@ -0,0 +1,51 @@
+package org.hibernate.envers.ebstest.integration.accesstype;
+
+import java.util.Arrays;
+
+import javax.persistence.EntityManager;
+
+import org.hibernate.ejb.Ejb3Configuration;
+import org.hibernate.envers.ebstest.entities.Country;
+import org.hibernate.envers.test.AbstractEntityTest;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import static org.testng.Assert.assertEquals;
+
+public class ImmutableClassAccessType extends AbstractEntityTest {
+	private Country country;
+
+	public void configure(Ejb3Configuration cfg) {
+		cfg.addAnnotatedClass(Country.class);
+	}
+
+	@BeforeClass(dependsOnMethods = "init")
+	public void initData() {
+		EntityManager em = getEntityManager();
+
+		// Revision 1
+		em.getTransaction().begin();
+		country = Country.of(new Integer(123), "Germany");
+		em.persist(country);
+		em.getTransaction().commit();
+
+	}
+
+	@Test
+	public void testRevisionsCounts() {
+		assert Arrays.asList(1)
+				.equals(
+						getAuditReader().getRevisions(Country.class,
+								country.getCode()));
+	}
+
+	@Test
+	public void testHistoryOfId1() {
+		Country country1 = getEntityManager().find(Country.class,
+				country.getCode());
+		assertEquals(country1, country);
+		
+		Country history = getAuditReader().find(Country.class, country1.getCode(), 1);
+		assertEquals(country, history);
+	}
+
+}
\ No newline at end of file
