Index: core/src/main/java/org/hibernate/mapping/Table.java
===================================================================
--- core/src/main/java/org/hibernate/mapping/Table.java	(revision 14470)
+++ core/src/main/java/org/hibernate/mapping/Table.java	(working copy)
@@ -266,6 +266,16 @@
 							", expected: " + col.getSqlType( dialect, mapping )
 					);
 				}
+				final boolean nullabilitiesMatch = columnInfo.getNullable().trim().equalsIgnoreCase((col.isNullable()?"yes":"no"));
+				if ( !nullabilitiesMatch ) {
+					throw new HibernateException(
+							"Wrong column nullability in " +
+							Table.qualify( tableInfo.getCatalog(), tableInfo.getSchema(), tableInfo.getName()) +
+							" for column " + col.getName() +
+							". Found nullable: " + columnInfo.getNullable().toLowerCase() +
+							", expected nullable: " + (col.isNullable()?"yes":"no")
+					);
+				}
 			}
 		}
 
