Index: hibernate-core/src/main/java/org/hibernate/query/QueryProducer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- hibernate-core/src/main/java/org/hibernate/query/QueryProducer.java	(date 1527430233000)
+++ hibernate-core/src/main/java/org/hibernate/query/QueryProducer.java	(revision )
@@ -117,7 +117,7 @@
 	 * implicit mapping to the specified Java type.
 	 *
 	 * @param sqlString Native (SQL) query string
-	 * @param resultClass The Java type to map results to
+	 * @param resultClass The entity to map results to
 	 *
 	 * @return The NativeQuery instance for manipulation and execution
 	 *
