diff -c org/hibernate/util/StringHelper.java ../hibernate-core-3.3.2.GA-aliaspatch/org/hibernate/util/StringHelper.java
*** org/hibernate/util/StringHelper.java	2009-06-24 02:54:32.000000000 +0300
--- ../hibernate-core-3.3.2.GA-aliaspatch/org/hibernate/util/StringHelper.java	2013-01-25 10:47:28.808764000 +0200
***************
*** 25,30 ****
--- 25,31 ----
  package org.hibernate.util;
  
  import java.util.Iterator;
+ import java.util.Locale;
  import java.util.StringTokenizer;
  import java.util.ArrayList;
  import java.util.Arrays;
***************
*** 441,447 ****
  	 */
  	private static String generateAliasRoot(String description) {
  		String result = truncate( unqualifyEntityName(description), ALIAS_TRUNCATE_LENGTH )
! 				.toLowerCase()
  		        .replace( '/', '_' ) // entityNames may now include slashes for the representations
  				.replace( '$', '_' ); //classname may be an inner class
  		result = cleanAlias( result );
--- 442,448 ----
  	 */
  	private static String generateAliasRoot(String description) {
  		String result = truncate( unqualifyEntityName(description), ALIAS_TRUNCATE_LENGTH )
! 				.toLowerCase(Locale.ENGLISH)
  		        .replace( '/', '_' ) // entityNames may now include slashes for the representations
  				.replace( '$', '_' ); //classname may be an inner class
  		result = cleanAlias( result );
