Index: core/src/main/java/org/hibernate/cfg/Configuration.java
===================================================================
--- core/src/main/java/org/hibernate/cfg/Configuration.java	(revision 14897)
+++ core/src/main/java/org/hibernate/cfg/Configuration.java	(working copy)
@@ -90,6 +90,7 @@
 import org.hibernate.mapping.RootClass;
 import org.hibernate.mapping.SimpleValue;
 import org.hibernate.mapping.Table;
+import org.hibernate.mapping.TypeDef;
 import org.hibernate.mapping.UniqueKey;
 import org.hibernate.proxy.EntityNotFoundDelegate;
 import org.hibernate.secure.JACCConfiguration;
@@ -2177,4 +2178,14 @@
 	public void setSessionFactoryObserver(SessionFactoryObserver sessionFactoryObserver) {
 		this.sessionFactoryObserver = sessionFactoryObserver;
 	}
+	
+	/**
+	 * Specifies the {@link TypeDef} that should be used for provided class name.
+	 *
+	 * @param className  the class for which the specified {@link TypeDef} should be used.
+	 * @param typeDef  the type definition to used with the provided class name.
+	 */
+	public void addTypeDef(String className, TypeDef typeDef) {
+		typeDefs.put(className, typeDef);
+	}
 }
