Index: src/org/hibernate/type/TypeFactory.java
===================================================================
--- src/org/hibernate/type/TypeFactory.java	(revision 11419)
+++ src/org/hibernate/type/TypeFactory.java	(working copy)
@@ -532,7 +532,7 @@
 				AbstractComponentType componentType = ( AbstractComponentType ) types[i];
 				Type[] subtypes = componentType.getSubtypes();
 				Object[] origComponentValues = original[i] == null ? new Object[subtypes.length] : componentType.getPropertyValues( original[i], session );
-				Object[] targetComponentValues = componentType.getPropertyValues( target[i], session );
+				Object[] targetComponentValues = target[i] == null ? new Object[subtypes.length] : componentType.getPropertyValues( target[i], session ); 
 				replaceAssociations( origComponentValues, targetComponentValues, subtypes, session, null, copyCache, foreignKeyDirection );
 				copied[i] = target[i];
 			}
