Index: C:/dev/hibernate/Branch_3_2/Hibernate3/src/org/hibernate/cache/OptimisticTreeCache.java
===================================================================
--- C:/dev/hibernate/Branch_3_2/Hibernate3/src/org/hibernate/cache/OptimisticTreeCache.java	(revision 11253)
+++ C:/dev/hibernate/Branch_3_2/Hibernate3/src/org/hibernate/cache/OptimisticTreeCache.java	(working copy)
@@ -294,6 +294,14 @@
 					return false;
 				}
 			}
+            
+            if (currentVersion == null) {
+               // If the workspace node has null as well, OK; if not we've
+               // been modified in a non-comparable manner, which we have to
+               // treat as us being newer 
+               return (other.previousVersion != null);
+            }
+            
 			return versionComparator.compare( currentVersion, other.previousVersion ) >= 1;
 		}
 
