--- BlobType.class	2008-07-30 11:23:07.000000000 -0600
+++ BlobType.java	2008-07-30 11:19:54.000000000 -0600
@@ -11,6 +11,7 @@
 
 import org.dom4j.Node;
 import org.hibernate.EntityMode;
+import org.hibernate.Hibernate;
 import org.hibernate.HibernateException;
 import org.hibernate.MappingException;
 import org.hibernate.engine.Mapping;
@@ -125,10 +126,12 @@
 	public Object replace(Object original, Object target,
 			SessionImplementor session, Object owner, Map copyCache)
 			throws HibernateException {
-		//Blobs are ignored by merge()
-		return target;
+		try {
+			return Hibernate.createBlob(((Blob)original).getBinaryStream());
+		} catch (Exception e) {
+			throw new HibernateException("Error copying the blob during the merge", e);
+		}
 	}
-
 	public int[] sqlTypes(Mapping mapping) throws MappingException {
 		return new int[] { Types.BLOB };
 	}
