diff --git a/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/GenerationTimingConverter.java b/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/GenerationTimingConverter.java
index 55b7d93..1aa6e9f 100644
--- a/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/GenerationTimingConverter.java
+++ b/hibernate-core/src/main/java/org/hibernate/boot/jaxb/hbm/internal/GenerationTimingConverter.java
@@ -19,6 +19,6 @@ public class GenerationTimingConverter {
 	}
 
 	public static String toXml(GenerationTiming generationTiming) {
-		return generationTiming.name().toLowerCase( Locale.ENGLISH );
+		return (null==generationTiming) ? null : generationTiming.name().toLowerCase( Locale.ENGLISH );
 	}
 }
