Index: /home/jean/workspace/Hibernate3/src/org/hibernate/dialect/Oracle9Dialect.java
===================================================================
--- /home/jean/workspace/Hibernate3/src/org/hibernate/dialect/Oracle9Dialect.java	(revision 10975)
+++ /home/jean/workspace/Hibernate3/src/org/hibernate/dialect/Oracle9Dialect.java	(working copy)
@@ -299,7 +299,7 @@
 
 	public String generateTemporaryTableName(String baseTableName) {
 		String name = super.generateTemporaryTableName(baseTableName);
-		return name.length() > 30 ? name.substring( 1, 30 ) : name;
+		return name.length() > 30 ? name.substring( 0, 30 ) : name;
 	}
 
 	public String getCreateTemporaryTableString() {
